
<!doctype html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
      
      
        <link rel="canonical" href="https://seqscope.github.io/cartloader/vignettes/subregion_tutorials/visiumhd/">
      
      
        <link rel="prev" href="../stereoseq/">
      
      
        <link rel="next" href="../pixelseq/">
      
      
      <link rel="icon" href="../../../assets/images/favicon.png">
      <meta name="generator" content="mkdocs-1.6.1, mkdocs-material-9.6.2">
    
    
      
        <title>Visium HD Starter Tutorial - cartloader</title>
      
    
    
      <link rel="stylesheet" href="../../../assets/stylesheets/main.d7758b05.min.css">
      
        
        <link rel="stylesheet" href="../../../assets/stylesheets/palette.06af60db.min.css">
      
      

  
  
  
  
  <style>:root{--md-annotation-icon:url('data:image/svg+xml;charset=utf-8,<svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M22 12a10 10 0 0 1-10 10A10 10 0 0 1 2 12 10 10 0 0 1 12 2a10 10 0 0 1 10 10M6 13h8l-3.5 3.5 1.42 1.42L17.84 12l-5.92-5.92L10.5 7.5 14 11H6z"/></svg>');}</style>


    
    
      
    
    
      
        
        
        <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&display=fallback">
        <style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style>
      
    
    
      <link rel="stylesheet" href="../../../css/extra.css">
    
    <script>__md_scope=new URL("../../..",location),__md_hash=e=>[...e].reduce(((e,_)=>(e<<5)-e+_.charCodeAt(0)),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script>
    
      

    
    
    
  </head>
  
  
    
    
      
    
    
    
    
    <body dir="ltr" data-md-color-scheme="default" data-md-color-primary="indigo" data-md-color-accent="indigo">
  
    
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
        
        <a href="#10x-visiumhd-starter-tutorial" class="md-skip">
          Skip to content
        </a>
      
    </div>
    <div data-md-component="announce">
      
    </div>
    
    
      

<header class="md-header" data-md-component="header">
  <nav class="md-header__inner md-grid" aria-label="Header">
    <a href="../../.." title="cartloader" class="md-header__button md-logo" aria-label="cartloader" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54"/></svg>

    </a>
    <label class="md-header__button md-icon" for="__drawer">
      
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3zm0 5h18v2H3zm0 5h18v2H3z"/></svg>
    </label>
    <div class="md-header__title" data-md-component="header-title">
      <div class="md-header__ellipsis">
        <div class="md-header__topic">
          <span class="md-ellipsis">
            cartloader
          </span>
        </div>
        <div class="md-header__topic" data-md-component="header-topic">
          <span class="md-ellipsis">
            
              Visium HD Starter Tutorial
            
          </span>
        </div>
      </div>
    </div>
    
      
        <form class="md-header__option" data-md-component="palette">
  
    
    
    
    <input class="md-option" data-md-color-media="(prefers-color-scheme)" data-md-color-scheme="default" data-md-color-primary="indigo" data-md-color-accent="indigo"  aria-label="Switch to light mode"  type="radio" name="__palette" id="__palette_0">
    
      <label class="md-header__button md-icon" title="Switch to light mode" for="__palette_1" hidden>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3.9 12c0-1.71 1.39-3.1 3.1-3.1h4V7H7a5 5 0 0 0-5 5 5 5 0 0 0 5 5h4v-1.9H7c-1.71 0-3.1-1.39-3.1-3.1M8 13h8v-2H8zm9-6h-4v1.9h4c1.71 0 3.1 1.39 3.1 3.1s-1.39 3.1-3.1 3.1h-4V17h4a5 5 0 0 0 5-5 5 5 0 0 0-5-5"/></svg>
      </label>
    
  
    
    
    
    <input class="md-option" data-md-color-media="(prefers-color-scheme: light)" data-md-color-scheme="default" data-md-color-primary="indigo" data-md-color-accent="indigo"  aria-label="Switch to dark mode"  type="radio" name="__palette" id="__palette_1">
    
      <label class="md-header__button md-icon" title="Switch to dark mode" for="__palette_2" hidden>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M17 7H7a5 5 0 0 0-5 5 5 5 0 0 0 5 5h10a5 5 0 0 0 5-5 5 5 0 0 0-5-5m0 8a3 3 0 0 1-3-3 3 3 0 0 1 3-3 3 3 0 0 1 3 3 3 3 0 0 1-3 3"/></svg>
      </label>
    
  
    
    
    
    <input class="md-option" data-md-color-media="(prefers-color-scheme: dark)" data-md-color-scheme="slate" data-md-color-primary="black" data-md-color-accent="indigo"  aria-label="Switch to system preference"  type="radio" name="__palette" id="__palette_2">
    
      <label class="md-header__button md-icon" title="Switch to system preference" for="__palette_0" hidden>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M17 7H7a5 5 0 0 0-5 5 5 5 0 0 0 5 5h10a5 5 0 0 0 5-5 5 5 0 0 0-5-5M7 15a3 3 0 0 1-3-3 3 3 0 0 1 3-3 3 3 0 0 1 3 3 3 3 0 0 1-3 3"/></svg>
      </label>
    
  
</form>
      
    
    
      <script>var palette=__md_get("__palette");if(palette&&palette.color){if("(prefers-color-scheme)"===palette.color.media){var media=matchMedia("(prefers-color-scheme: light)"),input=document.querySelector(media.matches?"[data-md-color-media='(prefers-color-scheme: light)']":"[data-md-color-media='(prefers-color-scheme: dark)']");palette.color.media=input.getAttribute("data-md-color-media"),palette.color.scheme=input.getAttribute("data-md-color-scheme"),palette.color.primary=input.getAttribute("data-md-color-primary"),palette.color.accent=input.getAttribute("data-md-color-accent")}for(var[key,value]of Object.entries(palette.color))document.body.setAttribute("data-md-color-"+key,value)}</script>
    
    
    
      <label class="md-header__button md-icon" for="__search">
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg>
      </label>
      <div class="md-search" data-md-component="search" role="dialog">
  <label class="md-search__overlay" for="__search"></label>
  <div class="md-search__inner" role="search">
    <form class="md-search__form" name="search">
      <input type="text" class="md-search__input" name="query" aria-label="Search" placeholder="Search" autocapitalize="off" autocorrect="off" autocomplete="off" spellcheck="false" data-md-component="search-query" required>
      <label class="md-search__icon md-icon" for="__search">
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg>
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11z"/></svg>
      </label>
      <nav class="md-search__options" aria-label="Search">
        
        <button type="reset" class="md-search__icon md-icon" title="Clear" aria-label="Clear" tabindex="-1">
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"/></svg>
        </button>
      </nav>
      
    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" tabindex="0" data-md-scrollfix>
        <div class="md-search-result" data-md-component="search-result">
          <div class="md-search-result__meta">
            Initializing search
          </div>
          <ol class="md-search-result__list" role="presentation"></ol>
        </div>
      </div>
    </div>
  </div>
</div>
    
    
      <div class="md-header__source">
        <a href="https://github.com/seqscope/cartloader" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 6.7.2 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2024 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81"/></svg>
  </div>
  <div class="md-source__repository">
    seqscope/cartloader
  </div>
</a>
      </div>
    
  </nav>
  
</header>
    
    <div class="md-container" data-md-component="container">
      
      
        
          
            
<nav class="md-tabs" aria-label="Tabs" data-md-component="tabs">
  <div class="md-grid">
    <ul class="md-tabs__list">
      
        
  
  
  
    
    
      <li class="md-tabs__item">
        <a href="../../.." class="md-tabs__link">
          
  
    
  
  Home

        </a>
      </li>
    
  

      
        
  
  
  
    <li class="md-tabs__item">
      <a href="../../../installation.md" class="md-tabs__link">
        
  
    
  
  Installation

      </a>
    </li>
  

      
        
  
  
  
    <li class="md-tabs__item">
      <a href="../../quickstart/quickstart/" class="md-tabs__link">
        
  
    
  
  Quick Start

      </a>
    </li>
  

      
        
  
  
    
  
  
    
    
      <li class="md-tabs__item md-tabs__item--active">
        <a href="../../intro/" class="md-tabs__link">
          
  
    
  
  Vignettes

        </a>
      </li>
    
  

      
        
  
  
  
    
    
      <li class="md-tabs__item">
        <a href="../../../reference/sge_convert/" class="md-tabs__link">
          
  
    
  
  References

        </a>
      </li>
    
  

      
    </ul>
  </div>
</nav>
          
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              
              <div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    


  


<nav class="md-nav md-nav--primary md-nav--lifted" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href="../../.." title="cartloader" class="md-nav__button md-logo" aria-label="cartloader" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54"/></svg>

    </a>
    cartloader
  </label>
  
    <div class="md-nav__source">
      <a href="https://github.com/seqscope/cartloader" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 6.7.2 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2024 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81"/></svg>
  </div>
  <div class="md-source__repository">
    seqscope/cartloader
  </div>
</a>
    </div>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
  
  
  
  
    
    
      
        
          
        
      
    
    
      
      
        
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
          
        
        <input class="md-nav__toggle md-toggle md-toggle--indeterminate" type="checkbox" id="__nav_1" >
        
          
          
          <div class="md-nav__link md-nav__container">
            <a href="../../.." class="md-nav__link ">
              
  
  <span class="md-ellipsis">
    Home
    
  </span>
  

            </a>
            
          </div>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_1_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_1">
            <span class="md-nav__icon md-icon"></span>
            Home
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../installation.md" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Installation
    
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../../quickstart/quickstart/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Quick Start
    
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
    
  
  
  
    
    
      
        
      
        
      
    
    
      
        
        
      
      
        
      
    
    
    <li class="md-nav__item md-nav__item--active md-nav__item--section md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_4" checked>
        
          
          <label class="md-nav__link" for="__nav_4" id="__nav_4_label" tabindex="">
            
  
  <span class="md-ellipsis">
    Vignettes
    
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_4_label" aria-expanded="true">
          <label class="md-nav__title" for="__nav_4">
            <span class="md-nav__icon md-icon"></span>
            Vignettes
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../intro/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Overview
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
    
  
  
  
    
    
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
    
    
      
      
        
          
          
        
      
    
    
    <li class="md-nav__item md-nav__item--active md-nav__item--section md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_4_2" checked>
        
          
          <label class="md-nav__link" for="__nav_4_2" id="__nav_4_2_label" tabindex="">
            
  
  <span class="md-ellipsis">
    Getting Started with `cartloader`
    
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="2" aria-labelledby="__nav_4_2_label" aria-expanded="true">
          <label class="md-nav__title" for="__nav_4_2">
            <span class="md-nav__icon md-icon"></span>
            Getting Started with `cartloader`
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../seqscope/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    SeqScope Starter Tutorial
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../stereoseq/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    StereoSeq Starter Tutorial
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
    
  
  
  
    <li class="md-nav__item md-nav__item--active">
      
      <input class="md-nav__toggle md-toggle" type="checkbox" id="__toc">
      
      
        
      
      
        <label class="md-nav__link md-nav__link--active" for="__toc">
          
  
  <span class="md-ellipsis">
    Visium HD Starter Tutorial
    
  </span>
  

          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <a href="./" class="md-nav__link md-nav__link--active">
        
  
  <span class="md-ellipsis">
    Visium HD Starter Tutorial
    
  </span>
  

      </a>
      
        

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#input-data" class="md-nav__link">
    <span class="md-ellipsis">
      Input Data
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#set-up-the-environment" class="md-nav__link">
    <span class="md-ellipsis">
      Set Up the Environment
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#sge-format-conversion" class="md-nav__link">
    <span class="md-ellipsis">
      SGE Format Conversion
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#ficture-analysis" class="md-nav__link">
    <span class="md-ellipsis">
      FICTURE analysis
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#cartloader-compilation" class="md-nav__link">
    <span class="md-ellipsis">
      cartloader Compilation
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#upload-to-data-repository" class="md-nav__link">
    <span class="md-ellipsis">
      Upload to Data Repository
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#output-data" class="md-nav__link">
    <span class="md-ellipsis">
      Output Data
    </span>
  </a>
  
</li>
      
    </ul>
  
</nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../pixelseq/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Pixel-Seq Starter Tutorial
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../xenium/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Xenium Starter Tutorial
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../merscope/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    MERSCOPE Starter Tutorial
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../cosmxsmi/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    CoxMX SMI Starter Tutorial
    
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
  
  
    
    
      
        
      
        
      
        
      
        
      
    
    
      
      
        
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
          
        
        <input class="md-nav__toggle md-toggle md-toggle--indeterminate" type="checkbox" id="__nav_5" >
        
          
          <label class="md-nav__link" for="__nav_5" id="__nav_5_label" tabindex="0">
            
  
  <span class="md-ellipsis">
    References
    
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_5_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_5">
            <span class="md-nav__icon md-icon"></span>
            References
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../reference/sge_convert/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    SGE Format Conversion
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../reference/run_ficture2/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    FICTURE Analysis
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../reference/run_cartload2/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    cartloader Compilation
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../reference/upload_repository/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Data Repository Upload
    
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
              
              <div class="md-sidebar md-sidebar--secondary" data-md-component="sidebar" data-md-type="toc" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#input-data" class="md-nav__link">
    <span class="md-ellipsis">
      Input Data
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#set-up-the-environment" class="md-nav__link">
    <span class="md-ellipsis">
      Set Up the Environment
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#sge-format-conversion" class="md-nav__link">
    <span class="md-ellipsis">
      SGE Format Conversion
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#ficture-analysis" class="md-nav__link">
    <span class="md-ellipsis">
      FICTURE analysis
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#cartloader-compilation" class="md-nav__link">
    <span class="md-ellipsis">
      cartloader Compilation
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#upload-to-data-repository" class="md-nav__link">
    <span class="md-ellipsis">
      Upload to Data Repository
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#output-data" class="md-nav__link">
    <span class="md-ellipsis">
      Output Data
    </span>
  </a>
  
</li>
      
    </ul>
  
</nav>
                  </div>
                </div>
              </div>
            
          
          
            <div class="md-content" data-md-component="content">
              <article class="md-content__inner md-typeset">
                
                  


  
  


<h1 id="10x-visiumhd-starter-tutorial">10X VisiumHD Starter Tutorial<a class="headerlink" href="#10x-visiumhd-starter-tutorial" title="Permanent link">&para;</a></h1>
<h2 id="input-data">Input Data<a class="headerlink" href="#input-data" title="Permanent link">&para;</a></h2>
<p>The input data originates from the mouse hippocampus and is extracted from an official release of mouse brain spatial gene expression (<abbr title="Spatial Digital Gene Expression">SGE</abbr>) data.</p>
<p><strong>File Format</strong></p>
<p>Visium HD slides use a 2×2 <abbr title="micrometer">µm</abbr> grid of barcoded squares (<code>square_002um</code>) for high-resolution spatial gene mapping. The <abbr title="Spatial Digital Gene Expression">SGE</abbr> file format comprises several key files as below: </p>
<details>
<summary><code>barcodes.tsv.gz</code> – spatial barcode for tissue locations </summary>
<div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span>
<span class="normal">2</span>
<span class="normal">3</span></pre></div></td><td class="code"><div><pre><span></span><code>s_002um_00639_00600-1
s_002um_00923_01639-1
s_002um_01050_01530-1
</code></pre></div></td></tr></table></div>
<ul>
<li>Column 1: spatial barcodes corresponding to specific locations on the tissue section. </li>
</ul>
</details>
<details>
<summary><code>features.tsv.gz</code> – feature metadata </summary>
<div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span>
<span class="normal">2</span>
<span class="normal">3</span></pre></div></td><td class="code"><div><pre><span></span><code>ENSMUSG00000051951  Xkr4    Gene Expression
ENSMUSG00000025900  Rp1     Gene Expression
ENSMUSG00000025902  Sox17   Gene Expression
</code></pre></div></td></tr></table></div>
<ul>
<li>Column 1: Feature ID</li>
<li>Column 2: Feature symbol</li>
<li>Column 3: Features types</li>
</ul>
</details>
<details>
<summary><code>matrix.mtx.gz</code> – expression count matrix </summary>
<div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span>
<span class="normal">2</span>
<span class="normal">3</span>
<span class="normal">4</span>
<span class="normal">5</span>
<span class="normal">6</span></pre></div></td><td class="code"><div><pre><span></span><code>%%MatrixMarket matrix coordinate integer general
%
19059 869411 11376563
3606 1 1
8957 1 1
9733 1 1
</code></pre></div></td></tr></table></div>
<ul>
<li><code>Header</code>: Initial lines form the header, declaring the matrix's adherence to the <a href="https://math.nist.gov/MatrixMarket/formats.html">Market Matrix (MTX) format</a>, outlining its traits. This may include comments (lines beginning with <code>%</code>) for extra metadata, all marked by a “%”.</li>
<li><code>Dimensions</code>: Following the header, the first line details the matrix dimensions: the count of rows (features), columns (barcodes), and non-zero entries.</li>
<li><code>Data Entries</code>: Post-dimensions, subsequent lines enumerate non-zero entries in seven columns: row index (feature index), column index (barcode index), and one value presenting the expression count per barcode per feature.</li>
</ul>
</details>
<details>
<summary><code>tissue_positions.parquet</code> – spatial barcode metadata </summary>
<div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span>
<span class="normal">2</span></pre></div></td><td class="code"><div><pre><span></span><code>barcode                 in_tissue   array_row   array_col   pxl_row_in_fullres  pxl_col_in_fullres
s_002um_00434_01637-1   1           434         1637        3396.371014         9125.919898
</code></pre></div></td></tr></table></div>
<ul>
<li><code>barcode</code>: Unique spatial barcode associated with each capture spot.</li>
<li><code>in_tissue</code>: Binary flag (1 = in tissue, 0 = background) indicating whether the spot falls within the tissue boundary.</li>
<li><code>array_row</code>, <code>array_col</code>: Integer indices representing the position of the spot on the capture array grid.</li>
<li><code>pxl_row_in_fullres</code>, <code>pxl_col_in_fullres</code>: Floating point coordinates locating the spot in full-resolution tissue image pixels.</li>
</ul>
</details>
<details>
<summary><code>scalefactors_json.json</code> – pixel-to-micrometer scaling factors </summary>
<div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span>
<span class="normal">2</span>
<span class="normal">3</span>
<span class="normal">4</span>
<span class="normal">5</span>
<span class="normal">6</span>
<span class="normal">7</span>
<span class="normal">8</span>
<span class="normal">9</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="p">{</span>
<span class="w">    </span><span class="nt">&quot;spot_diameter_fullres&quot;</span><span class="p">:</span><span class="w"> </span><span class="mf">7.303953797779634</span><span class="p">,</span>
<span class="w">    </span><span class="nt">&quot;bin_size_um&quot;</span><span class="p">:</span><span class="w"> </span><span class="mf">2.0</span><span class="p">,</span>
<span class="w">    </span><span class="nt">&quot;microns_per_pixel&quot;</span><span class="p">:</span><span class="w"> </span><span class="mf">0.2738242950835738</span><span class="p">,</span>
<span class="w">    </span><span class="nt">&quot;regist_target_img_scalef&quot;</span><span class="p">:</span><span class="w"> </span><span class="mf">0.2505533</span><span class="p">,</span>
<span class="w">    </span><span class="nt">&quot;tissue_lowres_scalef&quot;</span><span class="p">:</span><span class="w"> </span><span class="mf">0.02505533</span><span class="p">,</span>
<span class="w">    </span><span class="nt">&quot;fiducial_diameter_fullres&quot;</span><span class="p">:</span><span class="w"> </span><span class="mf">1205.1523766336395</span><span class="p">,</span>
<span class="w">    </span><span class="nt">&quot;tissue_hires_scalef&quot;</span><span class="p">:</span><span class="w"> </span><span class="mf">0.2505533</span>
<span class="p">}</span>
</code></pre></div></td></tr></table></div>
<ul>
<li><code>spot_diameter_fullres</code>: Estimated diameter of a barcoded spot in full-resolution pixels.</li>
<li><code>bin_size_um</code>: Physical size (in micrometers) of the smallest bin, typically 2.0 <abbr title="micrometer">µm</abbr> for Visium HD.</li>
<li><code>microns_per_pixel</code>: Resolution of the full-res image, used to convert pixel distances to micrometers.</li>
<li><code>regist_target_img_scalef</code>: Scaling factor applied during image registration to the target image.</li>
<li><code>tissue_lowres_scalef</code>: Downscaling factor from full-res to low-resolution tissue image.</li>
<li><code>fiducial_diameter_fullres</code>: Diameter of fiducial markers in full-resolution pixels, useful for alignment.</li>
<li><code>tissue_hires_scalef</code>: Downscaling factor from full-res to high-resolution tissue image.</li>
</ul>
</details>
<p><strong>Data Access</strong></p>
<p>The example data is hosted on Zenedo ().</p>
<p>Follow the commands below to download the example data.</p>
<div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span>
<span class="normal">2</span>
<span class="normal">3</span>
<span class="normal">4</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="nv">work_dir</span><span class="o">=</span>/path/to/work/directory
<span class="nb">cd</span><span class="w"> </span><span class="nv">$work_dir</span>
wget<span class="w">  </span>https://zenodo.org/records/15786632/files/visiumhd_starter.raw.tar.gz<span class="w"> </span>
tar<span class="w"> </span>-zxvf<span class="w"> </span>visiumhd_starter.raw.tar.gz<span class="w">  </span>
</code></pre></div></td></tr></table></div>
<h2 id="set-up-the-environment">Set Up the Environment<a class="headerlink" href="#set-up-the-environment" title="Permanent link">&para;</a></h2>
<p>Define paths to all required binaries and resources, and target AWS S3 bucket. Optionally, specify a fixed color map for consistent rendering.</p>
<div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"> 1</span>
<span class="normal"> 2</span>
<span class="normal"> 3</span>
<span class="normal"> 4</span>
<span class="normal"> 5</span>
<span class="normal"> 6</span>
<span class="normal"> 7</span>
<span class="normal"> 8</span>
<span class="normal"> 9</span>
<span class="normal">10</span>
<span class="normal">11</span>
<span class="normal">12</span>
<span class="normal">13</span>
<span class="normal">14</span>
<span class="normal">15</span>
<span class="normal">16</span>
<span class="normal">17</span>
<span class="normal">18</span>
<span class="normal">19</span>
<span class="normal">20</span>
<span class="normal">21</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="c1"># ====</span>
<span class="c1"># Replace each placeholder with the actual path on your system.  </span>
<span class="c1"># ====</span>
<span class="nv">work_dir</span><span class="o">=</span>/path/to/work/directory<span class="w">        </span><span class="c1"># path to work directory that contains the downloaded input data</span>
<span class="nb">cd</span><span class="w"> </span><span class="nv">$work_dir</span>

<span class="c1"># Define paths to required binaries and resources</span>
<span class="nv">spatula</span><span class="o">=</span>/path/to/spatula/binary<span class="w">         </span><span class="c1"># path to spatula executable</span>
<span class="nv">punkst</span><span class="o">=</span>/path/to/punkst/binary<span class="w">           </span><span class="c1"># path to FICTURE2/punkst executable</span>
<span class="nv">tippecanoe</span><span class="o">=</span>/path/to/tippecanoe/binary<span class="w">   </span><span class="c1"># path to tippecanoe executable</span>
<span class="nv">pmtiles</span><span class="o">=</span>/path/to/pmtiles/binary<span class="w">         </span><span class="c1"># path to pmtiles executable</span>
<span class="nv">aws</span><span class="o">=</span>/path/to/aws/cli/binary<span class="w">             </span><span class="c1"># path to AWS CLI binary</span>

<span class="c1"># (Optional) Define path to color map. </span>
<span class="nv">cmap</span><span class="o">=</span>/path/to/color/map<span class="w">                 </span><span class="c1"># Path to the fixed color map for rendering. cartloader provides a fixed color map at cartloader/assets/fixed_color_map_256.tsv.</span>

<span class="c1"># AWS S3 target location for cartostore</span>
<span class="nv">AWS_BUCKET</span><span class="o">=</span><span class="s2">&quot;EXAMPLE_AWS_BUCKET&quot;</span><span class="w">         </span><span class="c1"># replace EXAMPLE_AWS_BUCKET with your actual S3 bucket name</span>

<span class="c1"># Activate the bioconda environment</span>
conda<span class="w"> </span>activate<span class="w"> </span>ENV_NAME<span class="w">              </span><span class="c1"># replace BIOENV_NAME with your bioconda environment name</span>
</code></pre></div></td></tr></table></div>
<p>Define data ID and analysis parameters:</p>
<div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span>
<span class="normal">2</span>
<span class="normal">3</span>
<span class="normal">4</span>
<span class="normal">5</span>
<span class="normal">6</span>
<span class="normal">7</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="c1"># Unique identifier for your dataset</span>
<span class="nv">DATA_ID</span><span class="o">=</span><span class="s2">&quot;visiumhd_hippo&quot;</span><span class="w">                 </span><span class="c1"># change this to reflect your dataset name</span>
<span class="nv">PLATFORM</span><span class="o">=</span><span class="s2">&quot;10x_visium_hd&quot;</span><span class="w">                 </span><span class="c1"># platform information</span>

<span class="c1"># LDA parameters</span>
<span class="nv">train_width</span><span class="o">=</span><span class="m">18</span><span class="w">                           </span><span class="c1"># define LDA training hexagon width (comma-separated if multiple widths are applied)</span>
<span class="nv">n_factor</span><span class="o">=</span><span class="m">6</span>,12<span class="w">                            </span><span class="c1"># define number of factors in LDA training (comma-separated if multiple n-factor are applied)</span>
</code></pre></div></td></tr></table></div>
<div class="admonition info">
<p class="admonition-title">How to Define Scaling Factors for Visium HD</p>
<p>10x Visium HD includes a <code>scalefactors_json.json</code> file that provides pixel-to-micrometer scaling information. <code>cartloader</code> can directly accept this file via the <code>--scale-json</code> option and will automatically compute the appropriate scaling factor, omitting manually calculate and specify <code>--units-per-um</code>. </p>
<p>Alternatively, users may bypass the JSON file by directly providing a value through the <code>--units-per-um</code> option.</p>
</div>
<h2 id="sge-format-conversion"><abbr title="Spatial Digital Gene Expression">SGE</abbr> Format Conversion<a class="headerlink" href="#sge-format-conversion" title="Permanent link">&para;</a></h2>
<p>Convert the raw input to the unified <abbr title="Spatial Digital Gene Expression">SGE</abbr> format. See more details in <a href="../../../reference/sge_convert/">Reference page</a>.</p>
<div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"> 1</span>
<span class="normal"> 2</span>
<span class="normal"> 3</span>
<span class="normal"> 4</span>
<span class="normal"> 5</span>
<span class="normal"> 6</span>
<span class="normal"> 7</span>
<span class="normal"> 8</span>
<span class="normal"> 9</span>
<span class="normal">10</span>
<span class="normal">11</span></pre></div></td><td class="code"><div><pre><span></span><code>cartloader<span class="w"> </span>sge_convert<span class="w"> </span><span class="se">\</span>
<span class="w">  </span>--makefn<span class="w"> </span>sge_convert.mk<span class="w"> </span><span class="se">\</span>
<span class="w">  </span>--platform<span class="w"> </span><span class="si">${</span><span class="nv">PLATFORM</span><span class="si">}</span><span class="w"> </span><span class="se">\</span>
<span class="w">  </span>--in-mex<span class="w"> </span>./raw<span class="w"> </span><span class="se">\</span>
<span class="w">  </span>--in-parquet<span class="w"> </span>./raw/tissue_positions.parquet<span class="w"> </span><span class="se">\</span>
<span class="w">  </span>--scale-json<span class="w"> </span>./raw/scalefactors_json.json<span class="w"> </span><span class="se">\</span>
<span class="w">  </span>--out-dir<span class="w"> </span>./sge<span class="w"> </span><span class="se">\</span>
<span class="w">  </span>--exclude-feature-regex<span class="w"> </span><span class="s1">&#39;^(BLANK|NegCon|NegPrb)&#39;</span><span class="w"> </span><span class="se">\</span>
<span class="w">  </span>--sge-visual<span class="w"> </span><span class="se">\</span>
<span class="w">  </span>--spatula<span class="w"> </span><span class="si">${</span><span class="nv">spatula</span><span class="si">}</span><span class="w"> </span><span class="se">\</span>
<span class="w">  </span>--n-jobs<span class="w"> </span><span class="m">10</span>
</code></pre></div></td></tr></table></div>
<table>
<thead>
<tr>
<th>Parameter</th>
<th>Required</th>
<th>Type</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr>
<td><code>--platform</code></td>
<td>required</td>
<td>string</td>
<td>Platform (options: "<code>10x_visium_hd</code>", "<code>seqscope</code>", "<code>10x_xenium</code>", "<code>bgi_stereoseq</code>", "<code>cosmx_smi</code>", "<code>vizgen_merscope</code>", "<code>pixel_seq</code>", "<code>generic</code>")</td>
</tr>
<tr>
<td><code>--in-mex</code></td>
<td>required</td>
<td>string</td>
<td>Path to the input <abbr title="Market Exchange">MEX</abbr> directory containing gene × barcode matrix</td>
</tr>
<tr>
<td><code>--in-parquet</code></td>
<td>required</td>
<td>string</td>
<td>Path to the <code>tissue_positions.parquet</code> file with spatial barcode metadata</td>
</tr>
<tr>
<td><code>--scale-json</code></td>
<td>required <sup>1</sup></td>
<td>string</td>
<td>Path to the <code>scalefactors_json.json</code> file for coordinate scaling (or use <code>--units-per-um</code> to specify directly)</td>
</tr>
<tr>
<td><code>--out-dir</code></td>
<td>required</td>
<td>string</td>
<td>Output directory for the converted <abbr title="Spatial Digital Gene Expression">SGE</abbr> files</td>
</tr>
<tr>
<td><code>--makefn</code></td>
<td></td>
<td>string</td>
<td>File name for the generated Makefile (default: <code>sge_convert.mk</code>)</td>
</tr>
<tr>
<td><code>--exclude-feature-regex</code></td>
<td></td>
<td>regex</td>
<td>Pattern to exclude control features</td>
</tr>
<tr>
<td><code>--sge-visual</code></td>
<td></td>
<td>flag</td>
<td>Enable <abbr title="Spatial Digital Gene Expression">SGE</abbr> visualization step (generates diagnostic image) (default: <code>FALSE</code>)</td>
</tr>
<tr>
<td><code>--spatula</code></td>
<td></td>
<td>string</td>
<td>Path to the spatula binary (default: <code>spatula</code>)</td>
</tr>
<tr>
<td><code>--n-jobs</code></td>
<td></td>
<td>int</td>
<td>Number of parallel jobs for processing (default: <code>1</code>)</td>
</tr>
</tbody>
</table>
<p><sub><sup>1</sup>: To define the scaling factor, <code>cartloader</code> requires either a JSON file (via <code>--scale-json</code>) or a direct scale value using <code>--units-per-um</code>. When using <code>--scale-json</code>, make sure the JSON file has <code>microns_per_pixel</code> information.</sub></p>
<h2 id="ficture-analysis"><code>FICTURE</code> analysis<a class="headerlink" href="#ficture-analysis" title="Permanent link">&para;</a></h2>
<!-- !! Visium HD uses --decode-scale this should be updated. -->
<!-- 
<div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"> 1</span>
<span class="normal"> 2</span>
<span class="normal"> 3</span>
<span class="normal"> 4</span>
<span class="normal"> 5</span>
<span class="normal"> 6</span>
<span class="normal"> 7</span>
<span class="normal"> 8</span>
<span class="normal"> 9</span>
<span class="normal">10</span>
<span class="normal">11</span>
<span class="normal">12</span>
<span class="normal">13</span>
<span class="normal">14</span>
<span class="normal">15</span>
<span class="normal">16</span>
<span class="normal">17</span>
<span class="normal">18</span>
<span class="normal">19</span>
<span class="normal">20</span>
<span class="normal">21</span>
<span class="normal">22</span>
<span class="normal">23</span>
<span class="normal">24</span>
<span class="normal">25</span>
<span class="normal">26</span>
<span class="normal">27</span>
<span class="normal">28</span>
<span class="normal">29</span>
<span class="normal">30</span>
<span class="normal">31</span>
<span class="normal">32</span>
<span class="normal">33</span>
<span class="normal">34</span>
<span class="normal">35</span>
<span class="normal">36</span>
<span class="normal">37</span>
<span class="normal">38</span>
<span class="normal">39</span>
<span class="normal">40</span>
<span class="normal">41</span>
<span class="normal">42</span>
<span class="normal">43</span>
<span class="normal">44</span>
<span class="normal">45</span>
<span class="normal">46</span>
<span class="normal">47</span>
<span class="normal">48</span>
<span class="normal">49</span>
<span class="normal">50</span>
<span class="normal">51</span>
<span class="normal">52</span>
<span class="normal">53</span>
<span class="normal">54</span>
<span class="normal">55</span></pre></div></td><td class="code"><div><pre><span></span><code>cartloader<span class="w"> </span>run_ficture2<span class="w"> </span><span class="se">\</span>
<span class="w">  </span>--makefn<span class="w"> </span>run_ficture2.mk<span class="w"> </span><span class="se">\</span>
<span class="w">  </span>--main<span class="w"> </span><span class="se">\</span>
<span class="w">  </span>--in-transcript<span class="w"> </span>./sge/transcripts.unsorted.tsv.gz<span class="w"> </span><span class="se">\</span>
<span class="w">  </span>--in-feature<span class="w"> </span>./sge/feature.clean.tsv.gz<span class="w"> </span><span class="se">\</span>
<span class="w">  </span>--in-minmax<span class="w"> </span>./sge/coordinate_minmax.tsv<span class="w"> </span><span class="se">\</span>
<span class="w">  </span>--cmap-file<span class="w"> </span><span class="si">${</span><span class="nv">cmap</span><span class="si">}</span><span class="w"> </span><span class="se">\</span>
<span class="w">  </span>--exclude-feature-regex<span class="w"> </span><span class="s1">&#39;^(mt-.*$|Gm\d+$)&#39;</span><span class="w"> </span><span class="se">\</span>
<span class="w">  </span>--decode-scale<span class="w"> </span><span class="m">2</span><span class="w"> </span><span class="se">\</span>
<span class="w">  </span>--out-dir<span class="w"> </span>./ficture2<span class="w"> </span><span class="se">\</span>
<span class="w">  </span>--width<span class="w"> </span><span class="si">${</span><span class="nv">train_width</span><span class="si">}</span><span class="w"> </span><span class="se">\</span>
<span class="w">  </span>--n-factor<span class="w"> </span><span class="si">${</span><span class="nv">n_factor</span><span class="si">}</span><span class="w"> </span><span class="se">\</span>
<span class="w">  </span>--spatula<span class="w"> </span><span class="si">${</span><span class="nv">spatula</span><span class="si">}</span><span class="w"> </span><span class="se">\</span>
<span class="w">  </span>--ficture2<span class="w"> </span><span class="si">${</span><span class="nv">punkst</span><span class="si">}</span><span class="w"> </span><span class="se">\</span>
<span class="w">  </span>--n-jobs<span class="w"> </span><span class="m">10</span><span class="w"> </span><span class="se">\</span>
<span class="w">  </span>--threads<span class="w"> </span><span class="m">10</span>
<span class="sb">```</span><span class="w"> </span>--&gt;

Compute<span class="w"> </span>spatial<span class="w"> </span>factors<span class="w"> </span>using<span class="w"> </span><span class="sb">`</span>punkst<span class="sb">`</span><span class="w"> </span><span class="o">(</span>FICTURE2<span class="w"> </span>mode<span class="o">)</span>.<span class="w"> </span>See<span class="w"> </span>more<span class="w"> </span>details<span class="w"> </span><span class="k">in</span><span class="w"> </span><span class="o">[</span>Reference<span class="w"> </span>page<span class="o">](</span>../../reference/run_ficture2.md<span class="o">)</span>.

&lt;!--<span class="w"> </span><span class="sb">```</span>bash
cartloader<span class="w"> </span>run_ficture2<span class="w"> </span><span class="se">\</span>
<span class="w">    </span>--makefn<span class="w"> </span>run_ficture2.mk<span class="w"> </span><span class="se">\ </span><span class="w">                         </span><span class="c1"># (optional) file name of the output make file</span>
<span class="w">    </span>--main<span class="w"> </span><span class="se">\ </span><span class="w">                                           </span><span class="c1"># run all five steps in `run_ficture2`</span>
<span class="w">    </span>--in-transcript<span class="w"> </span>./sge/transcripts.unsorted.tsv.gz<span class="w"> </span><span class="se">\ </span><span class="c1"># path to input transcript-level SGE file</span>
<span class="w">    </span>--in-feature<span class="w"> </span>./sge/feature.clean.tsv.gz<span class="w"> </span><span class="se">\ </span><span class="w">          </span><span class="c1"># (optional) path to input feature file</span>
<span class="w">    </span>--in-minmax<span class="w"> </span>./sge/coordinate_minmax.tsv<span class="w"> </span><span class="se">\ </span><span class="w">          </span><span class="c1"># (optional) path to input minmax file</span>
<span class="w">    </span>--cmap-file<span class="w"> </span><span class="si">${</span><span class="nv">cmap</span><span class="si">}</span><span class="w"> </span><span class="se">\ </span><span class="w">                              </span><span class="c1"># (optional) path to input color map file</span>
<span class="w">    </span>--exclude-feature-regex<span class="w"> </span><span class="s1">&#39;^(mt-.*$|Gm\d+$)&#39;</span><span class="w"> </span><span class="se">\ </span><span class="w">       </span><span class="c1"># regex pattern to exclude features (removing mitochondrial and predicted genes in the example analysis)</span>
<span class="w">    </span>--out-dir<span class="w"> </span>./ficture2<span class="w"> </span><span class="se">\ </span><span class="w">                             </span><span class="c1"># path to output directory</span>
<span class="w">    </span>--width<span class="w"> </span><span class="si">${</span><span class="nv">train_width</span><span class="si">}</span><span class="w"> </span><span class="se">\ </span><span class="w">                           </span><span class="c1"># LDA training hexagon width (comma-separated if multiple widths are applied)</span>
<span class="w">    </span>--n-factor<span class="w"> </span><span class="si">${</span><span class="nv">n_factor</span><span class="si">}</span><span class="w"> </span><span class="se">\ </span><span class="w">                           </span><span class="c1"># number of factors in LDA training (comma-separated if multiple n-factor are applied)</span>
<span class="w">    </span>--spatula<span class="w"> </span><span class="si">${</span><span class="nv">spatula</span><span class="si">}</span><span class="w"> </span><span class="se">\ </span><span class="w">                             </span><span class="c1"># (optional) path to the spatula binary</span>
<span class="w">    </span>--ficture2<span class="w"> </span><span class="si">${</span><span class="nv">punkst</span><span class="si">}</span><span class="w"> </span><span class="se">\ </span><span class="w">                             </span><span class="c1"># (optional) path to the punkst directory</span>
<span class="w">    </span>--n-jobs<span class="w"> </span><span class="m">10</span><span class="w">  </span><span class="se">\ </span><span class="w">                                     </span><span class="c1"># (optional) number of parallel jobs </span>
<span class="w">    </span>--threads<span class="w"> </span><span class="m">10</span><span class="w">                                        </span><span class="c1"># (optional) number of threads per job</span>
<span class="sb">```</span><span class="w"> </span>--&gt;


<span class="sb">```</span>bash
cartloader<span class="w"> </span>run_ficture2<span class="w"> </span><span class="se">\</span>
<span class="w">  </span>--makefn<span class="w"> </span>run_ficture2.mk<span class="w"> </span><span class="se">\</span>
<span class="w">  </span>--main<span class="w"> </span><span class="se">\</span>
<span class="w">  </span>--in-transcript<span class="w"> </span>./sge/transcripts.unsorted.tsv.gz<span class="w"> </span><span class="se">\</span>
<span class="w">  </span>--in-feature<span class="w"> </span>./sge/feature.clean.tsv.gz<span class="w"> </span><span class="se">\</span>
<span class="w">  </span>--in-minmax<span class="w"> </span>./sge/coordinate_minmax.tsv<span class="w"> </span><span class="se">\</span>
<span class="w">  </span>--cmap-file<span class="w"> </span><span class="si">${</span><span class="nv">cmap</span><span class="si">}</span><span class="w"> </span><span class="se">\</span>
<span class="w">  </span>--exclude-feature-regex<span class="w"> </span><span class="s1">&#39;^(mt-.*$|Gm\d+$)&#39;</span><span class="w"> </span><span class="se">\</span>
<span class="w">  </span>--out-dir<span class="w"> </span>./ficture2<span class="w"> </span><span class="se">\</span>
<span class="w">  </span>--width<span class="w"> </span><span class="si">${</span><span class="nv">train_width</span><span class="si">}</span><span class="w"> </span><span class="se">\</span>
<span class="w">  </span>--n-factor<span class="w"> </span><span class="si">${</span><span class="nv">n_factor</span><span class="si">}</span><span class="w"> </span><span class="se">\</span>
<span class="w">  </span>--spatula<span class="w"> </span><span class="si">${</span><span class="nv">spatula</span><span class="si">}</span><span class="w"> </span><span class="se">\</span>
<span class="w">  </span>--ficture2<span class="w"> </span><span class="si">${</span><span class="nv">punkst</span><span class="si">}</span><span class="w"> </span><span class="se">\</span>
<span class="w">  </span>--n-jobs<span class="w"> </span><span class="m">10</span><span class="w"> </span><span class="se">\</span>
<span class="w">  </span>--threads<span class="w"> </span><span class="m">10</span>
</code></pre></div></td></tr></table></div>

<!--parameter-start-->
<table>
<thead>
<tr>
<th>Parameter</th>
<th>Required</th>
<th>Type</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr>
<td><code>--main</code></td>
<td>required <sup>1</sup></td>
<td>flag</td>
<td>Enable <code>cartloader</code> to run <a href="../../../reference/run_ficture2/#actions">all five steps</a></td>
</tr>
<tr>
<td><code>--in-transcript</code></td>
<td>required</td>
<td>string</td>
<td>Path to input transcript-level <abbr title="Spatial Digital Gene Expression">SGE</abbr> file</td>
</tr>
<tr>
<td><code>--out-dir</code></td>
<td>required</td>
<td>string</td>
<td>Path to output directory</td>
</tr>
<tr>
<td><code>--width</code></td>
<td>required</td>
<td>int or comma-separated list</td>
<td><abbr title="Latent Dirichlet Allocation">LDA</abbr> training hexagon width(s)</td>
</tr>
<tr>
<td><code>--n-factor</code></td>
<td>required</td>
<td>int or comma-separated list</td>
<td>Number of <abbr title="Latent Dirichlet Allocation">LDA</abbr> factors</td>
</tr>
<tr>
<td><code>--makefn</code></td>
<td></td>
<td>string</td>
<td>File name for the generated Makefile (default: <code>run_ficture2.mk</code> )</td>
</tr>
<tr>
<td><code>--in-feature</code></td>
<td></td>
<td>string</td>
<td>Path to input feature file</td>
</tr>
<tr>
<td><code>--in-minmax</code></td>
<td></td>
<td>string</td>
<td>Path to input coordinate min/max file</td>
</tr>
<tr>
<td><code>--cmap-file</code></td>
<td></td>
<td>string</td>
<td>Path to color map file</td>
</tr>
<tr>
<td><code>--exclude-feature-regex</code></td>
<td></td>
<td>regex</td>
<td>Pattern to exclude features</td>
</tr>
<tr>
<td><code>--spatula</code></td>
<td></td>
<td>string</td>
<td>Path to the <code>spatula</code> binary (default: <code>spatula</code>)</td>
</tr>
<tr>
<td><code>--ficture2</code></td>
<td></td>
<td>string</td>
<td>Path to the <code>punkst</code> directory (defaults to <code>punkst</code> repository within <code>submodules</code> directory of  <code>cartloader</code>)</td>
</tr>
<tr>
<td><code>--n-jobs</code></td>
<td></td>
<td>int</td>
<td>Number of parallel jobs (default: <code>1</code>)</td>
</tr>
<tr>
<td><code>--threads</code></td>
<td></td>
<td>int</td>
<td>Number of threads per job (default: <code>1</code>)</td>
</tr>
</tbody>
</table>
<p><sub><sup>1</sup>: <code>cartloader</code> requires the user to specify <strong>at least one action</strong>. Available actions includes: <code>--tile</code> to run tiling step; <code>--segment</code> to run segmentation step; <code>--init-lda</code> to run <abbr title="Latent Dirichlet Allocation">LDA</abbr> training step; <code>--decode</code> to run decoding step; <code>--summary</code> to run summarization step; <code>--main</code> to run all above five actions.</sub></p>
<!--parameter-end-->

<h2 id="cartloader-compilation"><code>cartloader</code> Compilation<a class="headerlink" href="#cartloader-compilation" title="Permanent link">&para;</a></h2>
<p>Generate pmtiles and web-compatible tile directories. See more details in <a href="../../../reference/run_cartload2/">Reference page</a>.</p>
<!-- ```bash
cartloader run_cartload2 \
    --makefn run_cartload2.mk \         # (optional) file name of the output make file
    --fic-dir ./ficture2 \              # path to input directory containing FICTURE2 output
    --out-dir ./cartload2 \             # path to output directory for PMTiles and web tiles
    --id ${DATA_ID} \                   # dataset ID used for naming outputs and metadata
    --spatula ${spatula} \              # (optional) path to the spatula binary
    --pmtiles ${pmtiles} \              # (optional) path to the pmtiles binary
    --tippecanoe ${tippecanoe} \        # (optional) path to the tippecanoe binary
    --n-jobs 10 \                       # (optional) number of parallel jobs
    --threads 10                        # (optional) number of threads per job
``` -->

<div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"> 1</span>
<span class="normal"> 2</span>
<span class="normal"> 3</span>
<span class="normal"> 4</span>
<span class="normal"> 5</span>
<span class="normal"> 6</span>
<span class="normal"> 7</span>
<span class="normal"> 8</span>
<span class="normal"> 9</span>
<span class="normal">10</span></pre></div></td><td class="code"><div><pre><span></span><code>cartloader<span class="w"> </span>run_cartload2<span class="w"> </span><span class="se">\</span>
<span class="w">  </span>--makefn<span class="w"> </span>run_cartload2.mk<span class="w"> </span><span class="se">\</span>
<span class="w">  </span>--fic-dir<span class="w"> </span>./ficture2<span class="w"> </span><span class="se">\</span>
<span class="w">  </span>--out-dir<span class="w"> </span>./cartload2<span class="w"> </span><span class="se">\</span>
<span class="w">  </span>--id<span class="w"> </span><span class="si">${</span><span class="nv">DATA_ID</span><span class="si">}</span><span class="w"> </span><span class="se">\</span>
<span class="w">  </span>--spatula<span class="w"> </span><span class="si">${</span><span class="nv">spatula</span><span class="si">}</span><span class="w"> </span><span class="se">\</span>
<span class="w">  </span>--pmtiles<span class="w"> </span><span class="si">${</span><span class="nv">pmtiles</span><span class="si">}</span><span class="w"> </span><span class="se">\</span>
<span class="w">  </span>--tippecanoe<span class="w"> </span><span class="si">${</span><span class="nv">tippecanoe</span><span class="si">}</span><span class="w"> </span><span class="se">\</span>
<span class="w">  </span>--n-jobs<span class="w"> </span><span class="m">10</span><span class="w"> </span><span class="se">\</span>
<span class="w">  </span>--threads<span class="w"> </span><span class="m">10</span>
</code></pre></div></td></tr></table></div>
<table>
<thead>
<tr>
<th>Parameter</th>
<th>Required</th>
<th>Type</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr>
<td><code>--fic-dir</code></td>
<td>required</td>
<td>string</td>
<td>Path to the input directory containing FICTURE2 output</td>
</tr>
<tr>
<td><code>--out-dir</code></td>
<td>required</td>
<td>string</td>
<td>Path to the output directory for PMTiles and web tiles</td>
</tr>
<tr>
<td><code>--id</code></td>
<td>required</td>
<td>string</td>
<td>Dataset ID used for naming outputs and metadata</td>
</tr>
<tr>
<td><code>--makefn</code></td>
<td></td>
<td>string</td>
<td>File name for the generated Makefile (default: <code>run_cartload2.mk</code>)</td>
</tr>
<tr>
<td><code>--spatula</code></td>
<td></td>
<td>string</td>
<td>Path to the <code>spatula</code> binary (default: <code>spatula</code>)</td>
</tr>
<tr>
<td><code>--pmtiles</code></td>
<td></td>
<td>string</td>
<td>Path to the <code>pmtiles</code> binary (default: <code>pmtiles</code>)</td>
</tr>
<tr>
<td><code>--tippecanoe</code></td>
<td></td>
<td>string</td>
<td>Path to the <code>tippecanoe</code> binary (default: <code>tippecanoe</code>)</td>
</tr>
<tr>
<td><code>--n-jobs</code></td>
<td></td>
<td>int</td>
<td>Number of parallel jobs (default: <code>1</code>)</td>
</tr>
<tr>
<td><code>--threads</code></td>
<td></td>
<td>int</td>
<td>Number of threads per job (default: <code>1</code>)</td>
</tr>
</tbody>
</table>
<h2 id="upload-to-data-repository">Upload to Data Repository<a class="headerlink" href="#upload-to-data-repository" title="Permanent link">&para;</a></h2>
<div class="admonition info">
<p class="admonition-title">Choose a data repository to host/share your output</p>
<p><code>cartloader</code> supports two upload options (<code>AWS</code> and <code>Zenodo</code>) for storing PMTiles of <abbr title="Spatial Digital Gene Expression">SGE</abbr> and spatial factors in a data repository.</p>
<p><strong>Choose the one that best suits your needs.</strong></p>
</div>
<h3 id="aws-uploads">AWS Uploads<a class="headerlink" href="#aws-uploads" title="Permanent link">&para;</a></h3>
<p>Upload the generated cartloader outputs to your designated AWS S3 directory:</p>
<div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span>
<span class="normal">2</span>
<span class="normal">3</span>
<span class="normal">4</span>
<span class="normal">5</span></pre></div></td><td class="code"><div><pre><span></span><code>cartloader<span class="w"> </span>upload_aws<span class="w"> </span><span class="se">\</span>
<span class="w">  </span>--in-dir<span class="w"> </span>./cartload2<span class="w"> </span><span class="se">\</span>
<span class="w">  </span>--s3-dir<span class="w"> </span><span class="s2">&quot;s3://</span><span class="si">${</span><span class="nv">AWS_BUCKET</span><span class="si">}</span><span class="s2">/</span><span class="si">${</span><span class="nv">DATA_ID</span><span class="si">}</span><span class="s2">&quot;</span><span class="w"> </span><span class="se">\</span>
<span class="w">  </span>--aws<span class="w"> </span><span class="si">${</span><span class="nv">aws</span><span class="si">}</span><span class="w"> </span><span class="se">\</span>
<span class="w">  </span>--n-jobs<span class="w"> </span><span class="m">10</span>
</code></pre></div></td></tr></table></div>
<table>
<thead>
<tr>
<th>Parameter</th>
<th>Required</th>
<th>Type</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr>
<td><code>--in-dir</code></td>
<td>required</td>
<td>string</td>
<td>Path to the input directory containing the cartloader compilation output</td>
</tr>
<tr>
<td><code>--s3-dir</code></td>
<td>required</td>
<td>string</td>
<td>Path to the target S3 directory for uploading</td>
</tr>
<tr>
<td><code>--aws</code></td>
<td></td>
<td>string</td>
<td>Path to the AWS CLI binary</td>
</tr>
<tr>
<td><code>--n-jobs</code></td>
<td></td>
<td>int</td>
<td>Number of parallel jobs</td>
</tr>
</tbody>
</table>
<h3 id="zenodo-uploads">Zenodo Uploads<a class="headerlink" href="#zenodo-uploads" title="Permanent link">&para;</a></h3>
<p>Upload the generated cartloader outputs to your designated Zenodo deposition or a new deposition.</p>
<div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span>
<span class="normal">2</span>
<span class="normal">3</span>
<span class="normal">4</span>
<span class="normal">5</span>
<span class="normal">6</span>
<span class="normal">7</span>
<span class="normal">8</span></pre></div></td><td class="code"><div><pre><span></span><code>cartloader<span class="w"> </span>upload_zenodo<span class="w"> </span><span class="se">\</span>
<span class="w">  </span>--in-dir<span class="w"> </span>./cartload2<span class="w"> </span><span class="se">\</span>
<span class="w">  </span>--upload-method<span class="w"> </span>catalog<span class="w"> </span><span class="se">\</span>
<span class="w">  </span>--zenodo-token<span class="w"> </span>/path/to/zenodo/token/file<span class="w"> </span><span class="se">\</span>
<span class="w">  </span>--create-new-deposition<span class="w"> </span><span class="se">\</span>
<span class="w">  </span>--title<span class="w">  </span><span class="s2">&quot;Yur Title&quot;</span><span class="w"> </span><span class="se">\</span>
<span class="w">  </span>--creators<span class="w"> </span><span class="s2">&quot;Your Name&quot;</span><span class="w"> </span><span class="se">\</span>
<span class="w">  </span>--description<span class="w"> </span><span class="s2">&quot;This is an example description&quot;</span>
</code></pre></div></td></tr></table></div>
<table>
<thead>
<tr>
<th>Parameter</th>
<th>Required</th>
<th>Type</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr>
<td><code>--in-dir</code></td>
<td>required</td>
<td>string</td>
<td>Path to the input directory containing the cartloader compilation output</td>
</tr>
<tr>
<td><code>--upload-method</code></td>
<td>required</td>
<td>string</td>
<td>Method to determine which files to upload. Options: <code>all</code> to upload all files in <code>--in-dir</code>; <code>catalog</code> to upload files listed in a catalog YAML file, <code>user_list</code> to upload files explicitly listed via <code>--in-list</code></td>
</tr>
<tr>
<td><code>--catalog-yaml</code></td>
<td></td>
<td>string</td>
<td>Required if <code>--upload-method catalog</code>.  Path to the catalog.yaml file generated in <code>run_cartload2</code>. If absent, will use the catalog.yaml in the input directory specified by <code>--in-dir</code>.</td>
</tr>
<tr>
<td><code>--zenodo-token</code></td>
<td>required</td>
<td>string</td>
<td>Path to your Zenodo access file</td>
</tr>
<tr>
<td><code>--create-new-deposition</code></td>
<td></td>
<td>flag</td>
<td>a new Zenodo deposition will be created.</td>
</tr>
<tr>
<td><code>--title</code></td>
<td>required</td>
<td>string</td>
<td>Required if <code>--create-new-deposition</code>. Title for the new Zenodo deposition.</td>
</tr>
<tr>
<td><code>--creators</code></td>
<td>required</td>
<td>list of str</td>
<td>List of creators in "Lastname, Firstname" format.</td>
</tr>
</tbody>
</table>
<h2 id="output-data">Output Data<a class="headerlink" href="#output-data" title="Permanent link">&para;</a></h2>
<p>See more details of output at the Reference pages for <a href="../../../reference/run_ficture2/">run_ficture2</a> and <a href="../../../reference/run_cartload2/">run_cartload2</a>.</p>
<h3 id="spatial-factor-inference-from-ficture">Spatial Factor Inference from <code>FICTURE</code><a class="headerlink" href="#spatial-factor-inference-from-ficture" title="Permanent link">&para;</a></h3>
<p>Below is an example of spatial factor inference results produced by <code>FICTURE</code> using a training width of 18, 12 factors, a fit width of 18, and an anchor resolution of 6.</p>
<p><img alt="FICTURE" src="../../../images/starter_vignettes/visiumhd.t18_f12_p18_a6.png" title="Spatial factor map" />
<img alt="cmap" src="../../../images/starter_vignettes/visiumhd.t18-f12-rgb.png" /></p>
<!-- 
<div id="color-legend-container"></div>
<script>
  document.addEventListener("DOMContentLoaded", function() {
    loadColorLegend("../../../tabs/colormap/visiumhd_starter.t18-f12-rgb.tsv", "color-legend-container");
  });
</script> -->

<table>
<thead>
<tr>
<th style="text-align: right;">Factor</th>
<th style="text-align: left;">RGB</th>
<th style="text-align: right;">Weight</th>
<th style="text-align: right;">PostUMI</th>
<th style="text-align: left;">TopGene_pval</th>
<th style="text-align: left;">TopGene_fc</th>
<th style="text-align: left;">TopGene_weight</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align: right;">0</td>
<td style="text-align: left;">255,101,101</td>
<td style="text-align: right;">0.20931</td>
<td style="text-align: right;">2920184</td>
<td style="text-align: left;">Atp2b1,Tmsb4x,Hpca,Wfs1,Cck,Plk2,Ptk2b,Rasgrp1,Scn3b,Cacng8,Ppp3r1,Kcnab1,Man1a,Epha6,Ppp3ca,Khdrbs3,Fibcd1,Prkcg,Rasd1,Itpka</td>
<td style="text-align: left;">Lefty1,Stac,Grem1,Wnt11,Spink8,Wfs1,Fibcd1,Gpr101,Wnt2,Htr5b,Chrm5,Egfl6,Gpr161,B4galnt3,Myo1h,Col11a1,Htr1a,Ankrd63,Has3,Sult5a1</td>
<td style="text-align: left;">Tmsb4x,Ncdn,Calm3,App,Selenow,Ptk2b,Hpca,Atp2b1,Nrgn,Atp1b1,Itm2b,Cck,Olfm1,Ppp3ca,Rasgrp1,Atp2a2,Gpm6a,Camk2a,Calm2,Ppp3r1</td>
</tr>
<tr>
<td style="text-align: right;">1</td>
<td style="text-align: left;">237,238,0</td>
<td style="text-align: right;">0.1423</td>
<td style="text-align: right;">1985225</td>
<td style="text-align: left;">Ncdn,Olfm1,Prox1,C1ql2,Stxbp6,Ptk2b,Fam163b,Sema5a,Dsp,Rfx3,Calb1,C1ql3,Tiam1,Ccdc85a,Synpr,Bhlhe22,Rogdi,Rasgrf2,Dgkh,Ppfia2</td>
<td style="text-align: left;">Dsp,Il27ra,Arr3,Rtl3,Krt2,Wnt16,Pdyn,Ppl,Pla2g2f,Tdo2,C1ql2,Arsj,Plk5,Ntf3,Mylk3,Slc26a10,Olfr539,Gmnc,Cdh15,Htra4</td>
<td style="text-align: left;">Ncdn,Olfm1,Tmsb4x,Ptk2b,Calm3,Nrgn,Camk2n1,Camk2a,Selenow,Ppp3ca,Atp1b1,Snap25,Map1b,Gpm6a,Arf3,App,Chn1,Ywhah,Ptms,Gria2</td>
</tr>
<tr>
<td style="text-align: right;">2</td>
<td style="text-align: left;">101,255,101</td>
<td style="text-align: right;">0.13453</td>
<td style="text-align: right;">1876906</td>
<td style="text-align: left;">Sst,Npy,Gad2,Rab3b,Cnr1,Gad1,Pvalb,Nap1l5,Slc6a1,Serpini1,Slc32a1,Sncb,Cplx1,Snhg11,Kcnab3,Atp1b1,Zfp385a,Mdh1,Tshz2,Coro6</td>
<td style="text-align: left;">Rxfp3,Vax1,Lrrc38,Prph,Mybpc1,Dmrt2,Dlx5,Sst,Igf1,Cntnap3,Cox6a2,Slc17a8,Nhs,Crh,Pnoc,Ecel1,Slc32a1,Npy,Htr3a,Npas1</td>
<td style="text-align: left;">App,Snap25,Atp1b1,Calm3,Camk2n1,Atp2a2,Selenow,Uchl1,Snrpn,Snhg11,Eef1a2,Ncdn,Ndfip1,Tmsb4x,Sncb,Dnm1,Atp1a3,Atp6v1b2,Stxbp1,Scn1b</td>
</tr>
<tr>
<td style="text-align: right;">4</td>
<td style="text-align: left;">101,101,255</td>
<td style="text-align: right;">0.1124</td>
<td style="text-align: right;">1568195</td>
<td style="text-align: left;">Camk2a,Slc1a2,Aldoc,Ppp1r9b,Apoe,Cst3,Ddn,Mt1,Glul,Map2,Selenow,Atp1a2,Mt2,Psd,Nsmf,Agap2,Ndrg2,Clu,Fam107a,Ncs1</td>
<td style="text-align: left;">Mettl11b,Calr4,Car5b,Mlph,Aqp9,Sox21,Ppp1r3c,Cyp4f14,Mctp2,Olfr287,Ikzf1,Umodl1,Dmp1,Ttpa,Acot11,Gli2,Apoc1,Gldc,Papss2,Dio2</td>
<td style="text-align: left;">Camk2a,Apoe,Selenow,Slc1a2,Aldoc,Camk2n1,Ppp1r9b,Cst3,Fth1,Ddn,Mt1,Cpe,Calm3,Mt3,Ncdn,Map2,Tmsb4x,Olfm1,Tspan7,Arf3</td>
</tr>
<tr>
<td style="text-align: right;">3</td>
<td style="text-align: left;">101,254,255</td>
<td style="text-align: right;">0.10862</td>
<td style="text-align: right;">1515463</td>
<td style="text-align: left;">Nptxr,Chgb,Snap25,Uchl1,Ywhah,Hpca,Slc17a7,Dnm1,Cadm3,Olfm1,Syn2,Bok,Stmn2,Ak5,Ywhag,Rasgrp1,Neurod6,Nectin3,Ak4,Cpne4</td>
<td style="text-align: left;">Iyd,Olfr550,Tmem114,Cldn22,Ttn,Syna,Cyp26b1,Slc9a4,Cd109,Prss35,Itgbl1,Rerg,Adamts8,Nkd2,Bok,Tmem215,Cpne4,Ccbe1,Socs2,Mcub</td>
<td style="text-align: left;">Snap25,Selenow,Olfm1,Calm3,Ncdn,App,Uchl1,Ywhah,Hpca,Ptk2b,Tmsb4x,Atp1b1,Dnm1,Chgb,Ywhag,Nptxr,Camk2n1,Slc17a7,Cpe,Nrgn</td>
</tr>
<tr>
<td style="text-align: right;">5</td>
<td style="text-align: left;">255,101,254</td>
<td style="text-align: right;">0.10013</td>
<td style="text-align: right;">1396880</td>
<td style="text-align: left;">Snap25,Vxn,Camk2n1,Pcp4,Ncald,Vsnl1,Stmn1,Ttc9b,Atp2b4,Myl4,Stxbp1,Basp1,Tbr1,Slc17a7,Cacnb4,Ngef,Stx1a,Efhd2,Egr1,Cabp1</td>
<td style="text-align: left;">Ntsr1,Krt80,Myl4,Neu2,Lipg,Trbc2,Gpr150,Palm3,Hs3st2,Rasgef1c,Arhgap25,Ovol2,Igfn1,Foxp2,Pou6f2,Ipcef1,Grp,Col5a1,Tpbg,Pdzph1</td>
<td style="text-align: left;">Snap25,Camk2n1,Olfm1,Calm3,Selenow,App,Nrgn,Tmsb4x,Uchl1,Atp2a2,Snrpn,Gpm6a,Dnm1,Eef1a2,Scn1b,Slc1a2,Ywhag,Slc17a7,Camk2a,Stxbp1</td>
</tr>
<tr>
<td style="text-align: right;">6</td>
<td style="text-align: left;">255,178,101</td>
<td style="text-align: right;">0.08468</td>
<td style="text-align: right;">1181439</td>
<td style="text-align: left;">Mbp,Plp1,Mobp,Mal,Cldn11,Trf,Fth1,Mag,Ugt8a,Cnp,Car2,Cryab,Apod,Tspan2,Sgk1,Gatm,Qdpr,Sept4,Plekhb1,Ptgds</td>
<td style="text-align: left;">Cyp3a13,Cyp2j12,Atp10b,Tmem125,Nipal4,Fa2h,Mboat1,Galnt6,Sec14l5,Mag,Mal,Cldn11,Nkx6-2,Tmem88b,S1pr5,Ugt8a,Dapk2,Ada,Gjc2,Aspa</td>
<td style="text-align: left;">Mbp,Plp1,Fth1,Mobp,Camk2n1,Apoe,App,Mal,Trf,Cldn11,Selenow,Tmsb4x,Car2,Qk,Apod,Sgk1,Ptgds,Aplp1,Cnp,Cryab</td>
</tr>
<tr>
<td style="text-align: right;">7</td>
<td style="text-align: left;">178,255,101</td>
<td style="text-align: right;">0.05769</td>
<td style="text-align: right;">804793</td>
<td style="text-align: left;">Apoe,Ptgds,Vim,Pltp,Tm4sf1,Acta2,Apod,Clu,Igf2,Igfbp7,Rarres2,Ahnak,Mt1,Mgp,Vtn,Cldn5,Atp1a2,Cst3,Fn1,Ptprb</td>
<td style="text-align: left;">Ttc21a,Crocc2,Cfap73,Mapk15,Smim5,Syt15,Col4a6,Iqca,Cfap206,Crygn,Cfap65,Pln,Dnaaf1,Col3a1,Ccdc153,Serpina3g,Tmem212,Lrrc23,Dnah1,Slc22a6</td>
<td style="text-align: left;">Apoe,Aldoc,Cst3,Mt1,Clu,Slc1a2,Itm2b,Ptgds,Atp1a2,Tmsb4x,Mt3,Apod,Sparcl1,Cpe,Mt2,App,Fth1,Itm2c,Selenow,Atp1b2</td>
</tr>
<tr>
<td style="text-align: right;">8</td>
<td style="text-align: left;">0,223,95</td>
<td style="text-align: right;">0.05026</td>
<td style="text-align: right;">701127</td>
<td style="text-align: left;">Tmem181a,Lrp1b,Fam163b,Kalrn,Mef2c,Khdrbs2,Syne1,Ptk2,Grm2,Rnf112,Epb41,Stxbp2,Ryr2,Leng8,Atp2b4,Atp1a1,Rsrp1,Cacna1d,Lyst,Adcy1</td>
<td style="text-align: left;">Efhc2,Zfp853,Col22a1,Igsf9,Clec18a,Podn,Xkr7,Cyp2r1,Gucy2g,Rnf207,Dsg2,Ccdc155,Kcnh4,Psg16,Cbx2,Ddx11,Myadml2,Igfbpl1,Mtbp,Eid3</td>
<td style="text-align: left;">Ncdn,Olfm1,Nrgn,Snap25,Camk2n1,Ptk2b,Calm3,Tmsb4x,App,Camk2a,Gpm6a,Selenow,Eef1a2,Atp1b1,Kalrn,Ptms,Ppp3ca,Gria2,Snhg11,Map1b</td>
</tr>
<tr>
<td style="text-align: right;">10</td>
<td style="text-align: left;">178,101,255</td>
<td style="text-align: right;">8e-05</td>
<td style="text-align: right;">1124</td>
<td style="text-align: left;">Hbb-bs,Hba-a2,S100a9,S100a8,Hbb-bt,Anxa1,Atg9b,Paqr5,Rbl1,Tent5c,Lcn2,Ikzf1,Ifit3b,Cdh5,Tcf19,Klf5,Lrrc32,Rrm2,Mob3c,Ifi30</td>
<td style="text-align: left;">Hbb-bs,Hba-a2,S100a8,S100a9,Hbb-bt,Anxa1,Atg9b,Rbl1,Tent5c,Paqr5,Lcn2,Ifit3b,Tcf19,Klf5,Lrrc32,Mob3c,Rrm2,Ifi30,Cdh5,Rassf9</td>
<td style="text-align: left;">Hbb-bs,Hba-a2,S100a9,S100a8,Hbb-bt,Mbp,Ppp1r9b,Sparcl1,Fth1,Mt3,Atp1a2,Rgs5,Tmsb4x,Cpe,Atp1b1,Selenow,Ly6e,Calm1,Apoe,Atp6v0d1</td>
</tr>
<tr>
<td style="text-align: right;">9</td>
<td style="text-align: left;">101,178,255</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">10</td>
<td style="text-align: left;">Mcm3,Bche,Hmgcs2,Cped1,Vwc2l,Spata6,Sh3tc2,Sh2b2,Slc26a2,Hhip,Cd83,Mdk,Zfp524,Unc13c,Nfil3,Bmp6,Dkkl1,Gmppb,Rhbdd1,Fam189a2</td>
<td style="text-align: left;">Mcm3,Bche,Selenop,Kcnab2,Sptbn2,Rnf168,Fa2h,Sacs,Hid1,Cd83,Rhob,Nmral1,Hmgcs2,Fam163b,Gna12,Gfap,Akirin2,Cand1,Phactr1,Sema7a</td>
<td style="text-align: left;">Mcm3,Sptbn2,Selenop,Kcnab2,Rhob,Gnao1,Aldoa_v1,Atp1b1,Gpm6a,Sept7,Calm1,Uqcrc1,Slc17a7,Vxn,Chchd2,Fam163b,Akirin2,Phactr1,Gfap,Cand1</td>
</tr>
</tbody>
</table>
<h3 id="packed-sge-and-spatial-factor-outputs-from-run_cartload2">Packed <abbr title="Spatial Digital Gene Expression">SGE</abbr> and Spatial Factor Outputs from <code>run_cartload2</code><a class="headerlink" href="#packed-sge-and-spatial-factor-outputs-from-run_cartload2" title="Permanent link">&para;</a></h3>
<p>The packed <abbr title="Spatial Digital Gene Expression">SGE</abbr> data and spatial factor inferences generated by <code>FICTURE</code> are available in PMTile format on Zenodo: <a href="https://zenodo.org/records/15824955">DOI:10.5281/zenodo.15824955</a>.</p>
<p>These datasets can also be loaded directly using the following catalog YAML file:<br />
<code>https://zenodo.org/records/15824956/catalog.yaml</code></p>












                
              </article>
            </div>
          
          
<script>var target=document.getElementById(location.hash.slice(1));target&&target.name&&(target.checked=target.name.startsWith("__tabbed_"))</script>
        </div>
        
      </main>
      
        <footer class="md-footer">
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-copyright">
  
  
    Made with
    <a href="https://squidfunk.github.io/mkdocs-material/" target="_blank" rel="noopener">
      Material for MkDocs
    </a>
  
</div>
      
    </div>
  </div>
</footer>
      
    </div>
    <div class="md-dialog" data-md-component="dialog">
      <div class="md-dialog__inner md-typeset"></div>
    </div>
    
    
    <script id="__config" type="application/json">{"base": "../../..", "features": ["navigation.tabs", "navigation.indexes", "navigation.sections", "navigation.expand", "navigation.path", "content.code.annotate", "content.grid", "content.cards", "content.footnote.tooltips"], "search": "../../../assets/javascripts/workers/search.f8cc74c7.min.js", "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version": "Select version"}}</script>
    
    
      <script src="../../../assets/javascripts/bundle.f13b1293.min.js"></script>
      
        <script src="../../../js/load_tsv.js"></script>
      
        <script src="../../../js/open_in_new_tab.js"></script>
      
    
  </body>
</html>