
<!doctype html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
      
      
        <link rel="canonical" href="https://seqscope.github.io/cartloader/reference/run_ficture2/">
      
      
        <link rel="prev" href="../sge_convert/">
      
      
        <link rel="next" href="../run_cartload2/">
      
      
      <link rel="icon" href="../../assets/images/favicon.png">
      <meta name="generator" content="mkdocs-1.6.1, mkdocs-material-9.6.2">
    
    
      
        <title>FICTURE Analysis - cartloader</title>
      
    
    
      <link rel="stylesheet" href="../../assets/stylesheets/main.d7758b05.min.css">
      
        
        <link rel="stylesheet" href="../../assets/stylesheets/palette.06af60db.min.css">
      
      

  
  
  
  
  <style>:root{--md-annotation-icon:url('data:image/svg+xml;charset=utf-8,<svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M22 12a10 10 0 0 1-10 10A10 10 0 0 1 2 12 10 10 0 0 1 12 2a10 10 0 0 1 10 10M6 13h8l-3.5 3.5 1.42 1.42L17.84 12l-5.92-5.92L10.5 7.5 14 11H6z"/></svg>');}</style>


    
    
      
    
    
      
        
        
        <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&display=fallback">
        <style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style>
      
    
    
      <link rel="stylesheet" href="../../css/extra.css">
    
    <script>__md_scope=new URL("../..",location),__md_hash=e=>[...e].reduce(((e,_)=>(e<<5)-e+_.charCodeAt(0)),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script>
    
      

    
    
    
  </head>
  
  
    
    
      
    
    
    
    
    <body dir="ltr" data-md-color-scheme="default" data-md-color-primary="indigo" data-md-color-accent="indigo">
  
    
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
        
        <a href="#spatial-factor-inference-analysis-using-ficture" class="md-skip">
          Skip to content
        </a>
      
    </div>
    <div data-md-component="announce">
      
    </div>
    
    
      

<header class="md-header" data-md-component="header">
  <nav class="md-header__inner md-grid" aria-label="Header">
    <a href="../.." title="cartloader" class="md-header__button md-logo" aria-label="cartloader" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54"/></svg>

    </a>
    <label class="md-header__button md-icon" for="__drawer">
      
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3zm0 5h18v2H3zm0 5h18v2H3z"/></svg>
    </label>
    <div class="md-header__title" data-md-component="header-title">
      <div class="md-header__ellipsis">
        <div class="md-header__topic">
          <span class="md-ellipsis">
            cartloader
          </span>
        </div>
        <div class="md-header__topic" data-md-component="header-topic">
          <span class="md-ellipsis">
            
              FICTURE Analysis
            
          </span>
        </div>
      </div>
    </div>
    
      
        <form class="md-header__option" data-md-component="palette">
  
    
    
    
    <input class="md-option" data-md-color-media="(prefers-color-scheme)" data-md-color-scheme="default" data-md-color-primary="indigo" data-md-color-accent="indigo"  aria-label="Switch to light mode"  type="radio" name="__palette" id="__palette_0">
    
      <label class="md-header__button md-icon" title="Switch to light mode" for="__palette_1" hidden>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3.9 12c0-1.71 1.39-3.1 3.1-3.1h4V7H7a5 5 0 0 0-5 5 5 5 0 0 0 5 5h4v-1.9H7c-1.71 0-3.1-1.39-3.1-3.1M8 13h8v-2H8zm9-6h-4v1.9h4c1.71 0 3.1 1.39 3.1 3.1s-1.39 3.1-3.1 3.1h-4V17h4a5 5 0 0 0 5-5 5 5 0 0 0-5-5"/></svg>
      </label>
    
  
    
    
    
    <input class="md-option" data-md-color-media="(prefers-color-scheme: light)" data-md-color-scheme="default" data-md-color-primary="indigo" data-md-color-accent="indigo"  aria-label="Switch to dark mode"  type="radio" name="__palette" id="__palette_1">
    
      <label class="md-header__button md-icon" title="Switch to dark mode" for="__palette_2" hidden>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M17 7H7a5 5 0 0 0-5 5 5 5 0 0 0 5 5h10a5 5 0 0 0 5-5 5 5 0 0 0-5-5m0 8a3 3 0 0 1-3-3 3 3 0 0 1 3-3 3 3 0 0 1 3 3 3 3 0 0 1-3 3"/></svg>
      </label>
    
  
    
    
    
    <input class="md-option" data-md-color-media="(prefers-color-scheme: dark)" data-md-color-scheme="slate" data-md-color-primary="black" data-md-color-accent="indigo"  aria-label="Switch to system preference"  type="radio" name="__palette" id="__palette_2">
    
      <label class="md-header__button md-icon" title="Switch to system preference" for="__palette_0" hidden>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M17 7H7a5 5 0 0 0-5 5 5 5 0 0 0 5 5h10a5 5 0 0 0 5-5 5 5 0 0 0-5-5M7 15a3 3 0 0 1-3-3 3 3 0 0 1 3-3 3 3 0 0 1 3 3 3 3 0 0 1-3 3"/></svg>
      </label>
    
  
</form>
      
    
    
      <script>var palette=__md_get("__palette");if(palette&&palette.color){if("(prefers-color-scheme)"===palette.color.media){var media=matchMedia("(prefers-color-scheme: light)"),input=document.querySelector(media.matches?"[data-md-color-media='(prefers-color-scheme: light)']":"[data-md-color-media='(prefers-color-scheme: dark)']");palette.color.media=input.getAttribute("data-md-color-media"),palette.color.scheme=input.getAttribute("data-md-color-scheme"),palette.color.primary=input.getAttribute("data-md-color-primary"),palette.color.accent=input.getAttribute("data-md-color-accent")}for(var[key,value]of Object.entries(palette.color))document.body.setAttribute("data-md-color-"+key,value)}</script>
    
    
    
      <label class="md-header__button md-icon" for="__search">
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg>
      </label>
      <div class="md-search" data-md-component="search" role="dialog">
  <label class="md-search__overlay" for="__search"></label>
  <div class="md-search__inner" role="search">
    <form class="md-search__form" name="search">
      <input type="text" class="md-search__input" name="query" aria-label="Search" placeholder="Search" autocapitalize="off" autocorrect="off" autocomplete="off" spellcheck="false" data-md-component="search-query" required>
      <label class="md-search__icon md-icon" for="__search">
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg>
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11z"/></svg>
      </label>
      <nav class="md-search__options" aria-label="Search">
        
        <button type="reset" class="md-search__icon md-icon" title="Clear" aria-label="Clear" tabindex="-1">
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"/></svg>
        </button>
      </nav>
      
    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" tabindex="0" data-md-scrollfix>
        <div class="md-search-result" data-md-component="search-result">
          <div class="md-search-result__meta">
            Initializing search
          </div>
          <ol class="md-search-result__list" role="presentation"></ol>
        </div>
      </div>
    </div>
  </div>
</div>
    
    
      <div class="md-header__source">
        <a href="https://github.com/seqscope/cartloader" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 6.7.2 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2024 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81"/></svg>
  </div>
  <div class="md-source__repository">
    seqscope/cartloader
  </div>
</a>
      </div>
    
  </nav>
  
</header>
    
    <div class="md-container" data-md-component="container">
      
      
        
          
            
<nav class="md-tabs" aria-label="Tabs" data-md-component="tabs">
  <div class="md-grid">
    <ul class="md-tabs__list">
      
        
  
  
  
    
    
      <li class="md-tabs__item">
        <a href="../.." class="md-tabs__link">
          
  
    
  
  Home

        </a>
      </li>
    
  

      
        
  
  
  
    <li class="md-tabs__item">
      <a href="../../installation/" class="md-tabs__link">
        
  
    
  
  Installation

      </a>
    </li>
  

      
        
  
  
  
    <li class="md-tabs__item">
      <a href="../../vignettes/quickstart/quickstart/" class="md-tabs__link">
        
  
    
  
  Quick Start

      </a>
    </li>
  

      
        
  
  
  
    
    
      <li class="md-tabs__item">
        <a href="../../vignettes/intro/" class="md-tabs__link">
          
  
    
  
  Vignettes

        </a>
      </li>
    
  

      
        
  
  
    
  
  
    
    
      <li class="md-tabs__item md-tabs__item--active">
        <a href="../sge_convert/" class="md-tabs__link">
          
  
    
  
  References

        </a>
      </li>
    
  

      
    </ul>
  </div>
</nav>
          
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              
              <div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    


  


<nav class="md-nav md-nav--primary md-nav--lifted" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href="../.." title="cartloader" class="md-nav__button md-logo" aria-label="cartloader" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54"/></svg>

    </a>
    cartloader
  </label>
  
    <div class="md-nav__source">
      <a href="https://github.com/seqscope/cartloader" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 6.7.2 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2024 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81"/></svg>
  </div>
  <div class="md-source__repository">
    seqscope/cartloader
  </div>
</a>
    </div>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
  
  
  
  
    
    
      
        
          
        
      
    
    
      
      
        
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
          
        
        <input class="md-nav__toggle md-toggle md-toggle--indeterminate" type="checkbox" id="__nav_1" >
        
          
          
          <div class="md-nav__link md-nav__container">
            <a href="../.." class="md-nav__link ">
              
  
  <span class="md-ellipsis">
    Home
    
  </span>
  

            </a>
            
          </div>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_1_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_1">
            <span class="md-nav__icon md-icon"></span>
            Home
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../../installation/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Installation
    
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../../vignettes/quickstart/quickstart/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Quick Start
    
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    
    
      
        
      
        
      
    
    
      
      
        
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
          
        
        <input class="md-nav__toggle md-toggle md-toggle--indeterminate" type="checkbox" id="__nav_4" >
        
          
          <label class="md-nav__link" for="__nav_4" id="__nav_4_label" tabindex="0">
            
  
  <span class="md-ellipsis">
    Vignettes
    
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_4_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_4">
            <span class="md-nav__icon md-icon"></span>
            Vignettes
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../vignettes/intro/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Overview
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    
    
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
    
    
      
      
        
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
          
        
        <input class="md-nav__toggle md-toggle md-toggle--indeterminate" type="checkbox" id="__nav_4_2" >
        
          
          <label class="md-nav__link" for="__nav_4_2" id="__nav_4_2_label" tabindex="0">
            
  
  <span class="md-ellipsis">
    Getting Started with `cartloader`
    
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="2" aria-labelledby="__nav_4_2_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_4_2">
            <span class="md-nav__icon md-icon"></span>
            Getting Started with `cartloader`
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../vignettes/subregion_tutorials/seqscope/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    SeqScope Starter Tutorial
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../vignettes/subregion_tutorials/stereoseq/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    StereoSeq Starter Tutorial
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../vignettes/subregion_tutorials/visiumhd/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Visium HD Starter Tutorial
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../vignettes/subregion_tutorials/pixelseq/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Pixel-Seq Starter Tutorial
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../vignettes/subregion_tutorials/xenium/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Xenium Starter Tutorial
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../vignettes/subregion_tutorials/merscope/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    MERSCOPE Starter Tutorial
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../vignettes/subregion_tutorials/cosmx_smi/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    CoxMX SMI Starter Tutorial
    
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
    
  
  
  
    
    
      
        
      
        
      
        
      
        
      
    
    
      
        
        
      
      
        
      
    
    
    <li class="md-nav__item md-nav__item--active md-nav__item--section md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_5" checked>
        
          
          <label class="md-nav__link" for="__nav_5" id="__nav_5_label" tabindex="">
            
  
  <span class="md-ellipsis">
    References
    
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_5_label" aria-expanded="true">
          <label class="md-nav__title" for="__nav_5">
            <span class="md-nav__icon md-icon"></span>
            References
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../sge_convert/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    SGE Format Conversion
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
    
  
  
  
    <li class="md-nav__item md-nav__item--active">
      
      <input class="md-nav__toggle md-toggle" type="checkbox" id="__toc">
      
      
        
      
      
        <label class="md-nav__link md-nav__link--active" for="__toc">
          
  
  <span class="md-ellipsis">
    FICTURE Analysis
    
  </span>
  

          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <a href="./" class="md-nav__link md-nav__link--active">
        
  
  <span class="md-ellipsis">
    FICTURE Analysis
    
  </span>
  

      </a>
      
        

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#overview" class="md-nav__link">
    <span class="md-ellipsis">
      Overview
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#requirements" class="md-nav__link">
    <span class="md-ellipsis">
      Requirements
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#example-usage" class="md-nav__link">
    <span class="md-ellipsis">
      Example Usage
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#actions" class="md-nav__link">
    <span class="md-ellipsis">
      Actions
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#parameters" class="md-nav__link">
    <span class="md-ellipsis">
      Parameters
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#output" class="md-nav__link">
    <span class="md-ellipsis">
      Output
    </span>
  </a>
  
</li>
      
    </ul>
  
</nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../run_cartload2/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    cartloader Compilation
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../upload_repository/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Data Repository Upload
    
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
              
              <div class="md-sidebar md-sidebar--secondary" data-md-component="sidebar" data-md-type="toc" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#overview" class="md-nav__link">
    <span class="md-ellipsis">
      Overview
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#requirements" class="md-nav__link">
    <span class="md-ellipsis">
      Requirements
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#example-usage" class="md-nav__link">
    <span class="md-ellipsis">
      Example Usage
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#actions" class="md-nav__link">
    <span class="md-ellipsis">
      Actions
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#parameters" class="md-nav__link">
    <span class="md-ellipsis">
      Parameters
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#output" class="md-nav__link">
    <span class="md-ellipsis">
      Output
    </span>
  </a>
  
</li>
      
    </ul>
  
</nav>
                  </div>
                </div>
              </div>
            
          
          
            <div class="md-content" data-md-component="content">
              <article class="md-content__inner md-typeset">
                
                  


  
  


<h1 id="spatial-factor-inference-analysis-using-ficture">Spatial Factor Inference Analysis using FICTURE<a class="headerlink" href="#spatial-factor-inference-analysis-using-ficture" title="Permanent link">&para;</a></h1>
<h2 id="overview">Overview<a class="headerlink" href="#overview" title="Permanent link">&para;</a></h2>
<p>Following format conversion, cartloader provides <code>run_ficture2</code> module to run spatial factor inference using <a href="https://www.nature.com/articles/s41592-024-02415-2"><strong><code>FICTURE</code></strong></a> (Si et al., <em>Nature Methods</em>, 2024). This method infers spatial factors directly at the pixel level with submicron resolution, eliminating the need for segmentation.</p>
<div class="admonition info">
<p class="admonition-title"><strong>What is <code>FICTURE</code>?</strong></p>
<p><a href="https://www.nature.com/articles/s41592-024-02415-2"><code>FICTURE</code></a> reconstructs the fine-scale tissue structure by first decomposing gene expression patterns across the tissue section into spatial factors and then assigns each pixel to these factors using local context. Biologically, these inferred factors may correspond to specific cell types, functional or physiological states, subcellular domains, or extracellular transcriptomic signatures.</p>
<p>By default, <a href="https://www.nature.com/articles/s41592-024-02415-2"><code>FICTURE</code></a> learns the spatial factors by implementing a standard latent Dirichlet allocation (<abbr title="Latent Dirichlet Allocation">LDA</abbr>) model on a hexagonal grid overlay of the spatial coordinates. Optionally, spatial factors can also be derived from external sources, such as single-cell or single-nucleus RNA-seq reference datasets, or from spatially agnostic factor learning methods (e.g., Seurat, Scanpy).</p>
<hr />
</div>
<div class="admonition info">
<p class="admonition-title"><strong>The <code>punkst</code> version of <code>FICTURE</code></strong></p>
<p>To efficiently run FICTURE-based inference, cartloader integrates <a href="https://github.com/Yichen-Si/punkst"><code>punkst</code></a>, an optimized implementation of FICTURE that maintains output equivalence while enhancing computational scalability and performance. </p>
<p>Currently, <code>run_ficture2</code> is using <code>punkst</code> version of <code>FICTURE</code>.</p>
</div>
<h2 id="requirements">Requirements<a class="headerlink" href="#requirements" title="Permanent link">&para;</a></h2>
<ul>
<li>An <abbr title="Spatial Digital Gene Expression">SGE</abbr> in <a href="">the unified format</a></li>
<li>Pre-installed tools: <code>spatula</code>, <code>punkst</code>, <code>gzip</code>, <code>sort</code>, <code>python</code></li>
</ul>
<h2 id="example-usage">Example Usage<a class="headerlink" href="#example-usage" title="Permanent link">&para;</a></h2>
<div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"> 1</span>
<span class="normal"> 2</span>
<span class="normal"> 3</span>
<span class="normal"> 4</span>
<span class="normal"> 5</span>
<span class="normal"> 6</span>
<span class="normal"> 7</span>
<span class="normal"> 8</span>
<span class="normal"> 9</span>
<span class="normal">10</span>
<span class="normal">11</span>
<span class="normal">12</span>
<span class="normal">13</span>
<span class="normal">14</span>
<span class="normal">15</span></pre></div></td><td class="code"><div><pre><span></span><code>cartloader<span class="w"> </span>run_ficture2<span class="w"> </span><span class="se">\</span>
<span class="w">    </span>--main<span class="w"> </span><span class="se">\</span>
<span class="w">    </span>--in-transcript<span class="w"> </span>/path/to/input/harmonized/transcripts/tsv/file<span class="w"> </span><span class="se">\</span>
<span class="w">    </span>--in-feature<span class="w"> </span>/path/to/input/harmonized/feature/tsv/file<span class="w"> </span><span class="se">\</span>
<span class="w">    </span>--in-minmax<span class="w"> </span>/path/to/input/harmonized/coordinates/minmax/tsv/file<span class="w"> </span><span class="se">\</span>
<span class="w">    </span>--cmap-file<span class="w"> </span>/path/to/cartloader/assets/fixed_color_map_256.tsv<span class="w"> </span><span class="se">\</span>
<span class="w">    </span>--colname-count<span class="w"> </span>count<span class="w"> </span><span class="se">\</span>
<span class="w">    </span>--out-dir<span class="w"> </span>/path/to/output/directory<span class="w"> </span><span class="se">\</span>
<span class="w">    </span>--width<span class="w"> </span><span class="m">18</span><span class="w"> </span><span class="se">\</span>
<span class="w">    </span>--n-factor<span class="w"> </span><span class="m">24</span><span class="w"> </span><span class="se">\</span>
<span class="w">    </span>--spatula<span class="w"> </span>/path/to/spatula/binary<span class="w"> </span><span class="se">\</span>
<span class="w">    </span>--ficture2<span class="w"> </span>/path/to/punkst/directory<span class="w"> </span><span class="se">\ </span><span class="w"> </span>
<span class="w">    </span>--exclude-feature-regex<span class="w"> </span><span class="s1">&#39;^(mt-|Gm\d+$)&#39;</span><span class="w"> </span><span class="se">\</span>
<span class="w">    </span>--n-jobs<span class="w"> </span><span class="m">20</span><span class="w">  </span><span class="se">\</span>
<span class="w">    </span>--threads<span class="w"> </span><span class="m">20</span><span class="w"> </span>
</code></pre></div></td></tr></table></div>
<h2 id="actions">Actions<a class="headerlink" href="#actions" title="Permanent link">&para;</a></h2>
<h3 id="tiling-step">Tiling step<a class="headerlink" href="#tiling-step" title="Permanent link">&para;</a></h3>
<p>The tiling step takes the standarized <abbr title="Spatial Digital Gene Expression">SGE</abbr> (from <a href="../sge_convert/"><abbr title="Spatial Digital Gene Expression">SGE</abbr> format conversion step</a>) as input. It aims to reorganizes input coordinate data into non-overlapping square tiles in a plain TSV format and generates an index file with tile offsets to enable efficient random access.</p>
<h3 id="segmentation-step">Segmentation Step<a class="headerlink" href="#segmentation-step" title="Permanent link">&para;</a></h3>
<p>The segmentation step starts from the tiled <abbr title="Spatial Digital Gene Expression">SGE</abbr>, using the plain TSV file from <a href="#tiling-step">tiling step</a> as input. It aggregates tiled pixel data into non-overlapping hexagons in a TSV file for spot-level analysis, outputting a tab-delimited file of hexagon records and associated metadata in JSON format.</p>
<h3 id="lda-training-step"><abbr title="Latent Dirichlet Allocation">LDA</abbr> Training Step<a class="headerlink" href="#lda-training-step" title="Permanent link">&para;</a></h3>
<p>The <abbr title="Latent Dirichlet Allocation">LDA</abbr> training step uses the hexagon TSV and JSON file from <a href="#segmentation-step">segmentation step</a> as input, trains a Latent Dirichlet Allocation (<abbr title="Latent Dirichlet Allocation">LDA</abbr>) model on sparse gene count data from hexagon units, using metadata to interpret input structure and optionally filter or weight features, producing a factorized topic model in TSV.</p>
<h3 id="decoding-step">Decoding Step<a class="headerlink" href="#decoding-step" title="Permanent link">&para;</a></h3>
<p>The decoding step applies a trained <abbr title="Latent Dirichlet Allocation">LDA</abbr> model from <a href="#lda-training-step"><abbr title="Latent Dirichlet Allocation">LDA</abbr> training step</a> to tiled pixel-level transcript data from <a href="#tiling-step">tiling step</a> to infer the top spatial factors and their posterior probabilities for each pixel, enabling fine-grained spatial mapping of gene expression. It outputs a pixel-level annotation file in TSV format with coordinates and factor assignments, along with a pseudobulk gene-by-factor matrix in TSV format.</p>
<h3 id="summarization-step">Summarization Step<a class="headerlink" href="#summarization-step" title="Permanent link">&para;</a></h3>
<p>The summarization step generate a JSON file to include all details of the FICTURE analysis, including the input files, output files, and parameters.</p>
<h2 id="parameters">Parameters<a class="headerlink" href="#parameters" title="Permanent link">&para;</a></h2>
<p>The following outlines the <strong>minimum required parameters</strong>. </p>
<p>For auxiliary parameters, we recommend using the default values unless you possess a thorough understanding of FICTURE. For further details, refer to the collapsible sections below or run:</p>
<div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span></pre></div></td><td class="code"><div><pre><span></span><code>cartloader<span class="w"> </span>run_ficture2<span class="w"> </span>--help
</code></pre></div></td></tr></table></div>
<h4 id="action-parameters">Action Parameters<a class="headerlink" href="#action-parameters" title="Permanent link">&para;</a></h4>
<div class="admonition warning">
<p class="admonition-title">Action Specifications</p>
<p>At least one of the actions (<code>--main</code>, <code>--tile</code>, <code>--segment</code>, <code>--init-lda</code>, <code>--decode</code>, <code>--summary</code>) should be enabled.</p>
</div>
<ul>
<li><code>--main</code>: Run all of the following five actions.</li>
<li><code>--tile</code>: Run <a href="#tiling-step">tiling step</a>.</li>
<li><code>--segment</code>: Run <a href="#segmentation-step">segmentation step</a>.</li>
<li><code>--init-lda</code>: Run <a href="#lda-training-step"><abbr title="Latent Dirichlet Allocation">LDA</abbr> training step</a>.</li>
<li><code>--decode</code>: Run <a href="#decoding-step">decoding step</a>.</li>
<li><code>--summary</code>: Run <a href="#summarization-step">summarization step</a>.</li>
</ul>
<h4 id="inputoutput-parameters">Input/Output Parameters<a class="headerlink" href="#inputoutput-parameters" title="Permanent link">&para;</a></h4>
<ul>
<li><code>--out-dir</code> (str): Output directory to store all result files.</li>
<li><code>--out-json</code> (str): Output JSON file summarizing FICTURE parameters (Default: <code>&lt;out-dir&gt;/ficture.params.json</code>).</li>
<li><code>--in-transcript</code> (str): Input transcript-indexed <abbr title="Spatial Digital Gene Expression">SGE</abbr> file in TSV format.</li>
<li><code>--in-minmax</code> (str): Optional input coordinate min-max file.</li>
<li><code>--in-feature</code> (str): Optional input <abbr title="Unique Molecular Identifier">UMI</abbr> count per gene TSV file.</li>
</ul>
<h4 id="key-parameters">Key Parameters<a class="headerlink" href="#key-parameters" title="Permanent link">&para;</a></h4>
<ul>
<li><code>--width</code> (str): Comma-separated hexagon flat-to-flat widths (in <abbr title="micrometer">µm</abbr>) for <abbr title="Latent Dirichlet Allocation">LDA</abbr> training.</li>
<li><code>--n-factor</code> (str): Comma-separated list of factor counts for <abbr title="Latent Dirichlet Allocation">LDA</abbr> training.</li>
<li><code>--include-feature-regex</code> (str): (Optional) Regex pattern for including features/genes.</li>
<li><code>--exclude-feature-regex</code> (str): (Optional) Regex pattern for excluding features/genes.</li>
<li><code>--cmap-file</code> (str): (Optional) Path to fixed color map TSV file. If not provided, FICTURE will generate a color map.</li>
</ul>
<details class="note">
<summary>Auxiliary <code>run_ficture2</code> Paramaters</summary>
<p><strong>Auxiliary Input Parameters</strong></p>
<ul>
<li><code>--in-feature-ficture</code> (str): (Optional) A separate feature file applied to FICTURE analysis. Alternative to customizing via auxiliary parameters.</li>
<li><code>--colidx-x</code> (int): Column index of X in the transcript file (Default: 1).</li>
<li><code>--colidx-y</code> (int): Column index of Y in the transcript file (Default: 2).</li>
<li><code>--colname-count</code> (str): Column name to use as count value (Default: count).</li>
<li><code>--colname-feature</code> (str): Column name for gene/feature name (Default: gene).</li>
</ul>
<p><strong>Auxiliary FICTURE Parameters</strong>:</p>
<ul>
<li>Tiling-specific parameters:<ul>
<li><code>--tile-size</code> (int): Size of tiles for processing (Default: 500).</li>
<li><code>--tile-buffer</code> (int): Buffer zone around each tile (Default: 1000).</li>
</ul>
</li>
<li>Segmentation-specific parameters:<ul>
<li><code>--min-ct-per-unit-hexagon</code> (int): Minimum count per hexagon (Default: 50).</li>
<li><code>--minibatch-size</code> (int): Minibatch size for preprocessing (Default: 500).</li>
</ul>
</li>
<li><abbr title="Latent Dirichlet Allocation">LDA</abbr> training-specific parameters:<ul>
<li><code>--min-ct-per-unit-train</code> (int): Minimum count for training (Default: 50).</li>
<li><code>--train-epoch</code> (int): Number of epochs to train <abbr title="Latent Dirichlet Allocation">LDA</abbr> model (Default: 2).</li>
</ul>
</li>
<li>
<p>Decoding-specific paramaters:</p>
<ul>
<li><code>--fit-width</code> (int): Hexagon width (in <abbr title="micrometer">µm</abbr>) for model fitting (Default: same as train width).</li>
<li><code>--min-ct-per-unit-fit</code> (int): Minimum count per unit during model fitting (Default: 50).</li>
<li><code>--anchor-res</code> (int): Anchor resolution used in decoding (Default: 6).</li>
<li><code>--radius-buffer</code> (int): Buffer added to anchor resolution for decoding (Default: 1).</li>
<li><code>--fit-plot-um-per-pixel</code> (int): Image resolution for fit coarse plots (Default: 1).</li>
<li><code>--decode-scale</code> (int): Scales input coordinates to pixels in the output image (Default: 1)')</li>
</ul>
</li>
<li>
<p>Shared paramaters across steps:</p>
<ul>
<li><code>--seed</code> (int): Random seed for reproducibility (Default: 1).</li>
<li><code>--min-ct-per-feature</code> (int): Minimum count per feature for <abbr title="Latent Dirichlet Allocation">LDA</abbr> and decoding (Default: 20).</li>
<li><code>--de-max-pval</code> (float): p-value cutoff for differential expression (Default: 1e-3).</li>
<li><code>--de-min-fold</code> (float): Fold-change threshold for differential expression (Default: 1.5).</li>
</ul>
</li>
</ul>
<p><strong>Auxiliary Environment Parameters</strong>:
For tools that require specifying the path to their executable binaries, you may omit the path if the binary is already included in your system's <code>PATH</code>.</p>
<ul>
<li><code>--gzip</code> (str): Path to <code>gzip</code> binary; consider <code>pigz -p 4</code> for speed (Default: <code>gzip</code>).</li>
<li><code>--sort</code> (str): Path to <code>sort</code> binary (Default: <code>sort</code>).</li>
<li><code>--sort-mem</code> (str): Memory allocated per <code>sort</code> process (Default: 1G).</li>
<li><code>--spatula</code> (str): Path to <code>spatula</code> binary (Default: <code>spatula</code>).</li>
<li><code>--ficture2</code> (str): Path to the <code>punkst</code> repository (Default to <code>punkst</code> directory in <code>submodules</code>)</li>
<li><code>--python</code> (str): Path to Python 3 binary (Default: <code>python3</code>).</li>
</ul>
</details>
<h2 id="output">Output<a class="headerlink" href="#output" title="Permanent link">&para;</a></h2>
<h3 id="tiling-output">Tiling Output<a class="headerlink" href="#tiling-output" title="Permanent link">&para;</a></h3>
<ul>
<li><code>transcripts.tiled.tsv</code>: A tab-delimited TSV file where each line records a tile with X and Y coordinates, a feature name (e.g., gene), and its associated count.
    <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span></pre></div></td><td class="code"><div><pre><span></span><code>3815.69 491.9   Arfgef1 1
</code></pre></div></td></tr></table></div></li>
<li><code>transcripts.tiled.index</code>: An index file that records the byte offsets of each tile in the tiled TSV file, enabling fast random access to tile-specific data. It includes comment lines with metadata (tile size, pixel count, min/max X and Y coordinates) and data lines listing the offset positions for each tile.
    <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span>
<span class="normal">2</span>
<span class="normal">3</span>
<span class="normal">4</span>
<span class="normal">5</span>
<span class="normal">6</span>
<span class="normal">7</span>
<span class="normal">8</span></pre></div></td><td class="code"><div><pre><span></span><code># tilesize  500
# npixels   333021294
# xmin  0.28
# xmax  19851.1
# ymin  0.5
# ymax  13476.4
0   7   0           13103550    569474
0   8   13103550    40259340    1180529
</code></pre></div></td></tr></table></div></li>
<li><code>transcripts.tiled.coord_range.tsv</code>: A tab-delimited TSV file provides xmin, xmax, ymin, and ymax as key–value pairs.
    <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span>
<span class="normal">2</span>
<span class="normal">3</span>
<span class="normal">4</span></pre></div></td><td class="code"><div><pre><span></span><code>xmin    0.28
xmax    19851.1
ymin    0.5
ymax    13476.4
</code></pre></div></td></tr></table></div></li>
<li><code>transcripts.tiled.features.tsv</code>: A tab-delimited TSV file listing each feature name and its expression count per line.
    <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span>
<span class="normal">2</span></pre></div></td><td class="code"><div><pre><span></span><code>1810009J06Rik   1
Gm48545         1
</code></pre></div></td></tr></table></div></li>
</ul>
<h3 id="segmentation-output">Segmentation Output<a class="headerlink" href="#segmentation-output" title="Permanent link">&para;</a></h3>
<ul>
<li>
<p><code>hexagon_d_{width}.randomized.tsv</code>: A plain-text TSV file that stores sparse feature representations for hexagon units, with each line corresponding to one hexagon.</p>
<div class="admonition info">
<p class="admonition-title">File Format</p>
<p>This is not a regular table—the number of columns varies across lines depending on the number of features present in each hexagon.</p>
</div>
<div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span></pre></div></td><td class="code"><div><pre><span></span><code>00001c64  18999.0000  9773.9627  227  244   2417 1  10469 2  2448 1  543 1   10935 1  6460 1  1797 1  6517 1  1094 1  9872 1  8587 1  7137 1  8786 1  10564 1  2127 1  439 1   1529 1  8432 1   1304 1  1739 1   3012 1  6434 1  5297 1  10017 1  2567 1  4771 1  5644 1  6415 1  5258 1   258 1   2314 1  6378 1  8843 1  7386 1  3322 1  2041 1  4735 1  9755 1  8614 1   4665 1  5693 1  2194 1  1680 1  10406 1  2696 1  7250 1  2367 1  7189 1  1616 1  10721 1  441 1   9981 1   8696 1  1012 1  8545 1  835 1   6550 1  4409 1  1860 1  575 1   139 1   5473 1  10172 1  7316 1  5260 1  180 1   2068 1  10806 1  8817 1  1621 1  4542 1  6193 1  440 1   6608 1  6989 1  4782 1  3258 1  5205 1  1093 1  9060 1  1921 1  6333 1   846 1   10575 1  2608 1  10186 1  932 1   2217 1  5392 1  2275 1  4845 1   8401 1  341 1   2810 1   754 1   4501 1  1160 1  5354 1  5097 1  3713 1  1619 1  9843 1  2749 1  1207 1  7384 1  2758 1  445 1   9040 1  45 1    2517 1   11255 1  8161 1  393 1   2963 1  3457 1  4481 1  10649 1  6954 2  2842 1  7982 1  2034 1  2849 2  5393 1  5257 1  773 1   1030 1  5455 1  2380 1  7079 1  1491 1  353 1    5343 1  1231 1  178 1   8659 1  10994 1  6168 1  267 1    6770 1   9083 1   2691 1  2423 2  3058 3  6921 1  1778 1  236 1   7175 1  7683 1   47 1    9094 2  1898 1  6524 1  5640 1  6668 1   6411 1  3536 1  10475 2  6538 1  1443 1  2728 1  7717 1  4633 1  6634 1   5349 1  11215 2  9159 1  2129 1  5984 1  8584 2  7042 1  3819 2  7931 1  2343 1  1182 2  2724 1  1871 1  716 1   1744 1  263 1   7459 1  6737 1  4654 1   9023 1  775 2    1546 1  1512 1  6395 1  5624 1  1560 2  8919 2  4689 1  2403 1  4163 1  5705 1  1260 1  8068 1  1900 1  1701 1  1187 1  8172 1  5955 1  494 1   428 1   6082 1  479 1   5362 1   11273 1  1764 1   9546 1  8261 1  11088 1  6262 1  2154 1  612 1   6777 2  866 1    10850 1  4168 1  9334 1  9077 2   3818 1  732 1   6386 1  989 1   1277 1  319 1   6230 1
</code></pre></div></td></tr></table></div>
<ul>
<li>1st column (str): Random identifier for the hexagon</li>
<li>2nd and 3rd columns (float): Axial coordinates in the hexagonal coordinate system</li>
<li>4th and 5th columns (int): Number of unique features and total counts, repeated per modality (for K modalities)</li>
<li>6th column onward (integer integer): Index–count pairs for non-zero features, using 0-based indices from --feature-dict or as provided in the input file</li>
</ul>
</li>
<li>
<p><code>hexagon_d_{width}.json</code>: The meta data for the hexagons. such as, , hexagon size, x and y column indices (), number of features, number of modalities, and number of hexagons (<code>n_units</code>),
    <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"> 1</span>
<span class="normal"> 2</span>
<span class="normal"> 3</span>
<span class="normal"> 4</span>
<span class="normal"> 5</span>
<span class="normal"> 6</span>
<span class="normal"> 7</span>
<span class="normal"> 8</span>
<span class="normal"> 9</span>
<span class="normal">10</span>
<span class="normal">11</span>
<span class="normal">12</span>
<span class="normal">13</span>
<span class="normal">14</span>
<span class="normal">15</span>
<span class="normal">16</span>
<span class="normal">17</span>
<span class="normal">18</span>
<span class="normal">19</span>
<span class="normal">20</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="p">{</span>
<span class="nt">&quot;dictionary&quot;</span><span class="p">:</span><span class="w"> </span><span class="p">{</span>
<span class="w">    </span><span class="nt">&quot;A2m&quot;</span><span class="p">:</span><span class="w"> </span><span class="mi">9636</span><span class="p">,</span>
<span class="w">    </span><span class="nt">&quot;AA986860&quot;</span><span class="p">:</span><span class="w"> </span><span class="mi">6486</span><span class="p">,</span>
<span class="w">    </span><span class="err">...</span>
<span class="w">    </span><span class="p">},</span>
<span class="w">    </span><span class="nt">&quot;header_info&quot;</span><span class="p">:</span><span class="w"> </span><span class="p">[</span>
<span class="w">        </span><span class="s2">&quot;random_key&quot;</span><span class="p">,</span>
<span class="w">        </span><span class="s2">&quot;x&quot;</span><span class="p">,</span>
<span class="w">        </span><span class="s2">&quot;y&quot;</span>
<span class="w">    </span><span class="p">],</span>
<span class="w">    </span><span class="nt">&quot;hex_size&quot;</span><span class="p">:</span><span class="w"> </span><span class="mf">10.392304845413264</span><span class="p">,</span>
<span class="w">    </span><span class="nt">&quot;icol_x&quot;</span><span class="p">:</span><span class="w"> </span><span class="mi">1</span><span class="p">,</span>
<span class="w">    </span><span class="nt">&quot;icol_y&quot;</span><span class="p">:</span><span class="w"> </span><span class="mi">2</span><span class="p">,</span>
<span class="w">    </span><span class="nt">&quot;n_features&quot;</span><span class="p">:</span><span class="w"> </span><span class="mi">11319</span><span class="p">,</span>
<span class="w">    </span><span class="nt">&quot;n_modalities&quot;</span><span class="p">:</span><span class="w"> </span><span class="mi">1</span><span class="p">,</span>
<span class="w">    </span><span class="nt">&quot;n_units&quot;</span><span class="p">:</span><span class="w"> </span><span class="mi">498019</span><span class="p">,</span>
<span class="w">    </span><span class="nt">&quot;offset_data&quot;</span><span class="p">:</span><span class="w"> </span><span class="mi">3</span><span class="p">,</span>
<span class="w">    </span><span class="nt">&quot;random_key&quot;</span><span class="p">:</span><span class="w"> </span><span class="mi">0</span>
<span class="p">}</span>
</code></pre></div></td></tr></table></div></p>
<ul>
<li><code>dictionary</code>: The feature name-idex pairs</li>
<li><code>header_info</code>: Header information</li>
<li><code>hex_size</code>: Size of hexagons</li>
<li><code>icol_x</code> and <code>icol_y</code>: Indices of X and Y coordinates</li>
<li><code>n_features</code>: Number of features</li>
<li><code>n_modalities</code>: Number of modalities</li>
<li><code>n_units</code>: Number of hexagons</li>
<li><code>offset_data</code>: ?</li>
<li><code>random_key</code>: the random seed for reproduction</li>
</ul>
</li>
</ul>
<h3 id="lda-training-output"><abbr title="Latent Dirichlet Allocation">LDA</abbr> Training Output<a class="headerlink" href="#lda-training-output" title="Permanent link">&para;</a></h3>
<ul>
<li><code>t{width}_f{n_factor}.model.tsv</code>: A matrix to store the learned topic-word distribution.
    <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span>
<span class="normal">2</span></pre></div></td><td class="code"><div><pre><span></span><code>Feature  0        1      2      3      4      5      6         7      8      9      10        11       12     13     14        15     16     17     18      19     20      21     22     23
Neu4     481.168  0.042  0.042  0.042  0.042  0.042  1828.293  0.042  0.042  0.042  2290.443  836.722  0.042  0.042  1704.589  0.042  0.042  0.042  29.738  0.042  24.854  0.042  0.042  0.042
</code></pre></div></td></tr></table></div><ul>
<li>Rows (str): <abbr title="Latent Dirichlet Allocation">LDA</abbr> topics or spatial factors</li>
<li>Columns (int): Feature identifiers (e.g., gene names or indices)</li>
<li>Values (float): Probability or weight of a feature under each topic <code>(P(feature | topic))</code></li>
</ul>
</li>
<li><code>t{width}_f{n_factor}.results.tsv.gz</code>: A tab-delimited file containing the posterior topic distributions for each hexagon.<br />
    <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span>
<span class="normal">2</span></pre></div></td><td class="code"><div><pre><span></span><code>x           y          0       1       2       3       4       5       6       7       8       9       10      11      12      13      14      15      16      17      18      19      20      21      22      23      topK  topP
14202.0000  5393.6062  0.0000  0.5087  0.0860  0.0000  0.0269  0.0000  0.0136  0.0000  0.0000  0.0179  0.0153  0.0000  0.0490  0.0000  0.0982  0.0000  0.0911  0.0199  0.0143  0.0189  0.0373  0.0000  0.0030  0.0000  1     0.5087
</code></pre></div></td></tr></table></div><ul>
<li><code>x</code> and <code>y</code> (float): X Y coordinates of the hexagon.</li>
<li>Columns <code>0</code> to <code>{n_factor-1}</code> (float): Posterior probabilities for each latent factor <code>(P(topic | hexagon))</code></li>
<li><code>topK</code> (int): Index of the most probable factor</li>
<li><code>topP</code> (float): Posterior probability of the top factor</li>
</ul>
</li>
<li>
<p><code>t{width}_f{n_factor}.bulk_chisq.tsv</code>
    <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span>
<span class="normal">2</span></pre></div></td><td class="code"><div><pre><span></span><code>gene  factor  Chi2      pval      FoldChange  gene_total  log10pval
App   0       864182.5  0.00e+00  4.41        1422353     187657.91
</code></pre></div></td></tr></table></div></p>
<ul>
<li><code>gene</code> (str): Gene names.</li>
<li><code>factor</code> (int): Factor IDs.</li>
<li><code>Chi2</code> (float): Chi-squared test statistic comparing the expression in the target factor and the rest.</li>
<li><code>pval</code> (float): P-value associated with the chi-squared test.</li>
<li><code>FoldChange</code> (float): Ratio of gene expression inside versus outside the factor’s high-loading region.</li>
<li><code>gene_total</code> (int): Total count of the gene in the dataset.</li>
<li><code>log10pval</code> (float): Base-10 logarithm of the inverse p-value (i.e., -log10(pval)), useful for ranking significant genes.</li>
</ul>
</li>
<li>
<p><code>t{width}_f{n_factor}.factor.info.tsv</code> and <code>t{width}_f{n_factor}.factor.info.html</code>: A TSV and HTML file providing the information for each factor per line.
    <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span>
<span class="normal">2</span></pre></div></td><td class="code"><div><pre><span></span><code>Factor  RGB          Weight   PostUMI   TopGene_pval                                                                                                                                       TopGene_fc                                                                                                                                           TopGene_weight
0       255,101,101  0.11279  82949360  App, Snrpn, Calm3, Eef1a2, Dctn1, Ptprn2, Akt3, Tmem181a, Sgtb, Ncdn, Selenow, Atp6v0a1, Cacng8, Atp2a2, Stub1, Akap5, Cmip, Cap1, Rundc3a, Ptprn  Sgtb, Map3k9, Cmip, Vps51, Ptprn2, Zscan2, Ppp2cb, Cacng8, Als2, Eif4e3, Ubxn2a, Gabrb2, Pms1, Rnpc3, Prepl, Zfp418, Tmem181a, Akap5, Dctn1, Atp2b3  App, Calm3, Selenow, Camk2n1, Snrpn, Ncdn, Atp2a2, Eef1a2, Uchl1, Ndfip1, Snap25, Stxbp1, Ywhag, Arf3, Serinc1, Atp6v1b2, Maged1, Atp6v0a1, Dnm1, Rab6b
</code></pre></div></td></tr></table></div></p>
<ul>
<li><code>Factor</code> (int): Factor IDs</li>
<li><code>RGB</code> (str): Comma-separated RGB values</li>
<li><code>Weight</code> (float): Proportion of the total factor signal explained by this factor</li>
<li><code>PostUMI</code> (int): Sum of posterior <abbr title="Unique Molecular Identifier">UMI</abbr> counts across all spatial locations for this factor</li>
<li><code>TopGene_pval</code>, <code>TopGene_fc</code>, <code>TopGene_weight</code> (str): Top marker genes per factor ranked by significance (p-value), fold change, or weight</li>
</ul>
</li>
<li><code>t{width}_f{n_factor}.cmap.tsv</code>: a color map.
    <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span>
<span class="normal">2</span></pre></div></td><td class="code"><div><pre><span></span><code>R    G    B    Color_hex  Name
255  101  101  #ff6565    0
</code></pre></div></td></tr></table></div><ul>
<li><code>Name</code> (int): Factor IDs</li>
<li><code>Color_hex</code> (str): Color HEX code</li>
<li><code>R</code>, <code>G</code>, <code>B</code> (float): Red, Green, and Blue channel values (range: 0.0 to 1.0)</li>
</ul>
</li>
</ul>
<h3 id="decoding-output">Decoding Output<a class="headerlink" href="#decoding-output" title="Permanent link">&para;</a></h3>
<ul>
<li>
<p><code>t{width}_f{n_factor}_p{width}_a{anchor_res}.tsv.gz</code>: A tab-delimited file where each line represents a pixel–feature pair, recording the pixel’s coordinates, the expressed feature and its count, along with the top three most probable latent factors (<code>K1</code>–<code>K3</code>) and their corresponding probabilities (<code>P1</code>–<code>P3</code>).
    <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span>
<span class="normal">2</span></pre></div></td><td class="code"><div><pre><span></span><code>#x         y         feature  ct  K1  K2  K3  P1          P2          P3
5159.4800  450.1500  Retreg2  1   0   19  9   9.5073e-01  4.8011e-02  5.0298e-04
</code></pre></div></td></tr></table></div></p>
<ul>
<li><code>x</code> and <code>y</code> (float): X Y coordinates</li>
<li><code>feature</code> (str): feature names.</li>
<li><code>ct</code> (int): count</li>
<li><code>K1</code> (int) and <code>P1</code> (float): The most probable factor and its probability</li>
<li><code>K2</code> (int) and <code>P2</code> (float): The 2nd most probable factor and its probability</li>
<li><code>K3</code> (int) and <code>P3</code> (float): The 3rd most probable factor and its probability</li>
</ul>
</li>
<li>
<p><code>t{width}_f{n_factor}_p{width}_a{anchor_res}.index</code>: An index file for <code>t{width}_f{n_factor}_p{width}_a{anchor_res}.tsv.gz</code>.</p>
</li>
<li><code>t{width}_f{n_factor}_p{width}_a{anchor_res}.json</code>: A JSON file to provide header information for <code>t{width}_f{n_factor}_p{width}_a{anchor_res}.tsv.gz</code>.
    <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"> 1</span>
<span class="normal"> 2</span>
<span class="normal"> 3</span>
<span class="normal"> 4</span>
<span class="normal"> 5</span>
<span class="normal"> 6</span>
<span class="normal"> 7</span>
<span class="normal"> 8</span>
<span class="normal"> 9</span>
<span class="normal">10</span>
<span class="normal">11</span>
<span class="normal">12</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="p">{</span>
<span class="w">    </span><span class="nt">&quot;K1&quot;</span><span class="p">:</span><span class="w"> </span><span class="mi">4</span><span class="p">,</span>
<span class="w">    </span><span class="nt">&quot;K2&quot;</span><span class="p">:</span><span class="w"> </span><span class="mi">5</span><span class="p">,</span>
<span class="w">    </span><span class="nt">&quot;K3&quot;</span><span class="p">:</span><span class="w"> </span><span class="mi">6</span><span class="p">,</span>
<span class="w">    </span><span class="nt">&quot;P1&quot;</span><span class="p">:</span><span class="w"> </span><span class="mi">7</span><span class="p">,</span>
<span class="w">    </span><span class="nt">&quot;P2&quot;</span><span class="p">:</span><span class="w"> </span><span class="mi">8</span><span class="p">,</span>
<span class="w">    </span><span class="nt">&quot;P3&quot;</span><span class="p">:</span><span class="w"> </span><span class="mi">9</span><span class="p">,</span>
<span class="w">    </span><span class="nt">&quot;ct&quot;</span><span class="p">:</span><span class="w"> </span><span class="mi">3</span><span class="p">,</span>
<span class="w">    </span><span class="nt">&quot;feature&quot;</span><span class="p">:</span><span class="w"> </span><span class="mi">2</span><span class="p">,</span>
<span class="w">    </span><span class="nt">&quot;x&quot;</span><span class="p">:</span><span class="w"> </span><span class="mi">0</span><span class="p">,</span>
<span class="w">    </span><span class="nt">&quot;y&quot;</span><span class="p">:</span><span class="w"> </span><span class="mi">1</span>
<span class="p">}</span>
</code></pre></div></td></tr></table></div></li>
<li><code>t{width}_f{n_factor}_p{width}_a{anchor_res}.png</code>: A PNG file to visualize the spatial factors distribution at pixel level.</li>
<li>
<p><code>t{width}_f{n_factor}_p{width}_a{anchor_res}.pseudobulk.tsv</code>: A feature-by-factor matrix showing feature distribution across topics.
    <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span>
<span class="normal">2</span></pre></div></td><td class="code"><div><pre><span></span><code>Feature  0         1       2       3       4       5       6          7       8       9       10         11        12      13      14        15      16      17      18       19      20       21      22      23
Neu4     648.5156  0.0000  0.9999  0.0000  0.0000  2.1425  1316.2651  0.0000  0.0000  0.0000  1114.2935  527.5519  0.0000  0.0000  820.3735  0.0000  0.0000  3.3504  30.4680  0.0000  38.0400  0.0000  0.0000  0.0000
</code></pre></div></td></tr></table></div></p>
<ul>
<li>Rows (str): <abbr title="Latent Dirichlet Allocation">LDA</abbr> topics or spatial factors</li>
<li>Columns (int): Feature identifiers (e.g., gene names or indices)</li>
<li>Values (float): Posterior probability of a feature in a factor.</li>
</ul>
</li>
<li>
<p><code>t{width}_f{n_factor}_p{width}_a{anchor_res}.bulk_chisq.tsv</code>: Same format as the <code>t{width}_f{n_factor}.bulk_chisq.tsv</code></p>
</li>
<li><code>t{width}_f{n_factor}_p{width}_a{anchor_res}.factor.info.tsv</code>:  Same format as the <code>t{width}_f{n_factor}.factor.info.tsv</code></li>
<li><code>t{width}_f{n_factor}_p{width}_a{anchor_res}.factor.info.html</code>: Same format as the <code>t{width}_f{n_factor}.factor.info.html</code></li>
</ul>
<h3 id="summarization-output">Summarization Output<a class="headerlink" href="#summarization-output" title="Permanent link">&para;</a></h3>
<ul>
<li><code>ficture.params.json</code>: A JSON file to summarize the path to input, output, and paramaters.
    <div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"> 1</span>
<span class="normal"> 2</span>
<span class="normal"> 3</span>
<span class="normal"> 4</span>
<span class="normal"> 5</span>
<span class="normal"> 6</span>
<span class="normal"> 7</span>
<span class="normal"> 8</span>
<span class="normal"> 9</span>
<span class="normal">10</span>
<span class="normal">11</span>
<span class="normal">12</span>
<span class="normal">13</span>
<span class="normal">14</span>
<span class="normal">15</span>
<span class="normal">16</span>
<span class="normal">17</span>
<span class="normal">18</span>
<span class="normal">19</span>
<span class="normal">20</span>
<span class="normal">21</span>
<span class="normal">22</span>
<span class="normal">23</span>
<span class="normal">24</span>
<span class="normal">25</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="p">{</span>
<span class="nt">&quot;in_sge&quot;</span><span class="p">:</span><span class="w"> </span><span class="p">{</span>
<span class="w">    </span><span class="nt">&quot;in_transcript&quot;</span><span class="p">:</span><span class="w"> </span><span class="s2">&quot;/path/to/transcripts.unsorted.tsv.gz&quot;</span><span class="p">,</span>
<span class="w">    </span><span class="nt">&quot;in_feature&quot;</span><span class="p">:</span><span class="w"> </span><span class="s2">&quot;/path/to/transcripts.tiled.features.hdr.tsv&quot;</span><span class="p">,</span>
<span class="w">    </span><span class="nt">&quot;in_minmax&quot;</span><span class="p">:</span><span class="w"> </span><span class="s2">&quot;/path/to/coordinate_minmax.tsv&quot;</span>
<span class="p">},</span>
<span class="nt">&quot;in_feature_ficture&quot;</span><span class="p">:</span><span class="w"> </span><span class="s2">&quot;/path/to/transcripts.tiled.overlapping_features.min100.hdr.tsv&quot;</span><span class="p">,</span><span class="w"> </span><span class="err">#</span><span class="w"> </span><span class="err">i</span><span class="kc">f</span><span class="w"> </span><span class="err">available</span>
<span class="nt">&quot;train_params&quot;</span><span class="p">:</span><span class="w"> </span><span class="p">[</span>
<span class="w">    </span><span class="p">{</span>
<span class="w">        </span><span class="nt">&quot;model_type&quot;</span><span class="p">:</span><span class="w"> </span><span class="s2">&quot;lda&quot;</span><span class="p">,</span>
<span class="w">        </span><span class="nt">&quot;model_id&quot;</span><span class="p">:</span><span class="w"> </span><span class="s2">&quot;t{width}_f{n_factor}&quot;</span><span class="p">,</span>
<span class="w">        </span><span class="nt">&quot;model_path&quot;</span><span class="p">:</span><span class="w"> </span><span class="s2">&quot;/path/to/t{width}_f{n_factor}.model.tsv&quot;</span><span class="p">,</span>
<span class="w">        </span><span class="nt">&quot;train_width&quot;</span><span class="p">:</span><span class="w"> </span><span class="err">wid</span><span class="kc">t</span><span class="err">h</span><span class="p">,</span>
<span class="w">        </span><span class="nt">&quot;n_factor&quot;</span><span class="p">:</span><span class="w"> </span><span class="kc">n</span><span class="err">_</span><span class="kc">fa</span><span class="err">c</span><span class="kc">t</span><span class="err">or</span><span class="p">,</span>
<span class="w">        </span><span class="nt">&quot;cmap&quot;</span><span class="p">:</span><span class="w"> </span><span class="s2">&quot;/path/to/t{width}_f{n_factor}.cmap.tsv&quot;</span><span class="p">,</span>
<span class="w">        </span><span class="nt">&quot;decode_params&quot;</span><span class="p">:</span><span class="w"> </span><span class="p">[</span>
<span class="w">            </span><span class="p">{</span>
<span class="w">                </span><span class="nt">&quot;decode_id&quot;</span><span class="p">:</span><span class="w"> </span><span class="s2">&quot;t{width}_f{n_factor}_p{width}_a{anchor_res}&quot;</span><span class="p">,</span>
<span class="w">                </span><span class="nt">&quot;fit_width&quot;</span><span class="p">:</span><span class="w"> </span><span class="err">wid</span><span class="kc">t</span><span class="err">h</span><span class="p">,</span>
<span class="w">                </span><span class="nt">&quot;anchor_res&quot;</span><span class="p">:</span><span class="w"> </span><span class="err">a</span><span class="kc">n</span><span class="err">chor_res</span>
<span class="w">            </span><span class="p">}</span>
<span class="w">        </span><span class="p">]</span>
<span class="w">    </span><span class="p">}</span>
<span class="p">]</span>
<span class="p">}</span>
</code></pre></div></td></tr></table></div></li>
</ul>












                
              </article>
            </div>
          
          
<script>var target=document.getElementById(location.hash.slice(1));target&&target.name&&(target.checked=target.name.startsWith("__tabbed_"))</script>
        </div>
        
      </main>
      
        <footer class="md-footer">
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-copyright">
  
  
    Made with
    <a href="https://squidfunk.github.io/mkdocs-material/" target="_blank" rel="noopener">
      Material for MkDocs
    </a>
  
</div>
      
    </div>
  </div>
</footer>
      
    </div>
    <div class="md-dialog" data-md-component="dialog">
      <div class="md-dialog__inner md-typeset"></div>
    </div>
    
    
    <script id="__config" type="application/json">{"base": "../..", "features": ["navigation.tabs", "navigation.indexes", "navigation.sections", "navigation.expand", "navigation.path", "content.code.annotate", "content.grid", "content.cards", "content.footnote.tooltips"], "search": "../../assets/javascripts/workers/search.f8cc74c7.min.js", "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version": "Select version"}}</script>
    
    
      <script src="../../assets/javascripts/bundle.f13b1293.min.js"></script>
      
        <script src="../../js/open_in_new_tab.js"></script>
      
    
  </body>
</html>