
<!doctype html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
      
      
        <link rel="canonical" href="https://seqscope.github.io/cartloader/vignettes/subregion_tutorials/cosmxsmi/">
      
      
        <link rel="prev" href="../merscope/">
      
      
        <link rel="next" href="../../../reference/sge_convert/">
      
      
      <link rel="icon" href="../../../assets/images/favicon.png">
      <meta name="generator" content="mkdocs-1.6.1, mkdocs-material-9.6.2">
    
    
      
        <title>CoxMX SMI Starter Tutorial - cartloader</title>
      
    
    
      <link rel="stylesheet" href="../../../assets/stylesheets/main.d7758b05.min.css">
      
        
        <link rel="stylesheet" href="../../../assets/stylesheets/palette.06af60db.min.css">
      
      

  
  
  
  
  <style>:root{--md-annotation-icon:url('data:image/svg+xml;charset=utf-8,<svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M22 12a10 10 0 0 1-10 10A10 10 0 0 1 2 12 10 10 0 0 1 12 2a10 10 0 0 1 10 10M6 13h8l-3.5 3.5 1.42 1.42L17.84 12l-5.92-5.92L10.5 7.5 14 11H6z"/></svg>');}</style>


    
    
      
    
    
      
        
        
        <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&display=fallback">
        <style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style>
      
    
    
      <link rel="stylesheet" href="../../../css/extra.css">
    
    <script>__md_scope=new URL("../../..",location),__md_hash=e=>[...e].reduce(((e,_)=>(e<<5)-e+_.charCodeAt(0)),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script>
    
      

    
    
    
  </head>
  
  
    
    
      
    
    
    
    
    <body dir="ltr" data-md-color-scheme="default" data-md-color-primary="indigo" data-md-color-accent="indigo">
  
    
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
        
        <a href="#cosmx-smi-starter-tutorial" class="md-skip">
          Skip to content
        </a>
      
    </div>
    <div data-md-component="announce">
      
    </div>
    
    
      

<header class="md-header" data-md-component="header">
  <nav class="md-header__inner md-grid" aria-label="Header">
    <a href="../../.." title="cartloader" class="md-header__button md-logo" aria-label="cartloader" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54"/></svg>

    </a>
    <label class="md-header__button md-icon" for="__drawer">
      
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3zm0 5h18v2H3zm0 5h18v2H3z"/></svg>
    </label>
    <div class="md-header__title" data-md-component="header-title">
      <div class="md-header__ellipsis">
        <div class="md-header__topic">
          <span class="md-ellipsis">
            cartloader
          </span>
        </div>
        <div class="md-header__topic" data-md-component="header-topic">
          <span class="md-ellipsis">
            
              CoxMX SMI Starter Tutorial
            
          </span>
        </div>
      </div>
    </div>
    
      
        <form class="md-header__option" data-md-component="palette">
  
    
    
    
    <input class="md-option" data-md-color-media="(prefers-color-scheme)" data-md-color-scheme="default" data-md-color-primary="indigo" data-md-color-accent="indigo"  aria-label="Switch to light mode"  type="radio" name="__palette" id="__palette_0">
    
      <label class="md-header__button md-icon" title="Switch to light mode" for="__palette_1" hidden>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3.9 12c0-1.71 1.39-3.1 3.1-3.1h4V7H7a5 5 0 0 0-5 5 5 5 0 0 0 5 5h4v-1.9H7c-1.71 0-3.1-1.39-3.1-3.1M8 13h8v-2H8zm9-6h-4v1.9h4c1.71 0 3.1 1.39 3.1 3.1s-1.39 3.1-3.1 3.1h-4V17h4a5 5 0 0 0 5-5 5 5 0 0 0-5-5"/></svg>
      </label>
    
  
    
    
    
    <input class="md-option" data-md-color-media="(prefers-color-scheme: light)" data-md-color-scheme="default" data-md-color-primary="indigo" data-md-color-accent="indigo"  aria-label="Switch to dark mode"  type="radio" name="__palette" id="__palette_1">
    
      <label class="md-header__button md-icon" title="Switch to dark mode" for="__palette_2" hidden>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M17 7H7a5 5 0 0 0-5 5 5 5 0 0 0 5 5h10a5 5 0 0 0 5-5 5 5 0 0 0-5-5m0 8a3 3 0 0 1-3-3 3 3 0 0 1 3-3 3 3 0 0 1 3 3 3 3 0 0 1-3 3"/></svg>
      </label>
    
  
    
    
    
    <input class="md-option" data-md-color-media="(prefers-color-scheme: dark)" data-md-color-scheme="slate" data-md-color-primary="black" data-md-color-accent="indigo"  aria-label="Switch to system preference"  type="radio" name="__palette" id="__palette_2">
    
      <label class="md-header__button md-icon" title="Switch to system preference" for="__palette_0" hidden>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M17 7H7a5 5 0 0 0-5 5 5 5 0 0 0 5 5h10a5 5 0 0 0 5-5 5 5 0 0 0-5-5M7 15a3 3 0 0 1-3-3 3 3 0 0 1 3-3 3 3 0 0 1 3 3 3 3 0 0 1-3 3"/></svg>
      </label>
    
  
</form>
      
    
    
      <script>var palette=__md_get("__palette");if(palette&&palette.color){if("(prefers-color-scheme)"===palette.color.media){var media=matchMedia("(prefers-color-scheme: light)"),input=document.querySelector(media.matches?"[data-md-color-media='(prefers-color-scheme: light)']":"[data-md-color-media='(prefers-color-scheme: dark)']");palette.color.media=input.getAttribute("data-md-color-media"),palette.color.scheme=input.getAttribute("data-md-color-scheme"),palette.color.primary=input.getAttribute("data-md-color-primary"),palette.color.accent=input.getAttribute("data-md-color-accent")}for(var[key,value]of Object.entries(palette.color))document.body.setAttribute("data-md-color-"+key,value)}</script>
    
    
    
      <label class="md-header__button md-icon" for="__search">
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg>
      </label>
      <div class="md-search" data-md-component="search" role="dialog">
  <label class="md-search__overlay" for="__search"></label>
  <div class="md-search__inner" role="search">
    <form class="md-search__form" name="search">
      <input type="text" class="md-search__input" name="query" aria-label="Search" placeholder="Search" autocapitalize="off" autocorrect="off" autocomplete="off" spellcheck="false" data-md-component="search-query" required>
      <label class="md-search__icon md-icon" for="__search">
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg>
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11z"/></svg>
      </label>
      <nav class="md-search__options" aria-label="Search">
        
        <button type="reset" class="md-search__icon md-icon" title="Clear" aria-label="Clear" tabindex="-1">
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"/></svg>
        </button>
      </nav>
      
    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" tabindex="0" data-md-scrollfix>
        <div class="md-search-result" data-md-component="search-result">
          <div class="md-search-result__meta">
            Initializing search
          </div>
          <ol class="md-search-result__list" role="presentation"></ol>
        </div>
      </div>
    </div>
  </div>
</div>
    
    
      <div class="md-header__source">
        <a href="https://github.com/seqscope/cartloader" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 6.7.2 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2024 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81"/></svg>
  </div>
  <div class="md-source__repository">
    seqscope/cartloader
  </div>
</a>
      </div>
    
  </nav>
  
</header>
    
    <div class="md-container" data-md-component="container">
      
      
        
          
            
<nav class="md-tabs" aria-label="Tabs" data-md-component="tabs">
  <div class="md-grid">
    <ul class="md-tabs__list">
      
        
  
  
  
    
    
      <li class="md-tabs__item">
        <a href="../../.." class="md-tabs__link">
          
  
    
  
  Home

        </a>
      </li>
    
  

      
        
  
  
  
    <li class="md-tabs__item">
      <a href="../../../installation/" class="md-tabs__link">
        
  
    
  
  Installation

      </a>
    </li>
  

      
        
  
  
  
    <li class="md-tabs__item">
      <a href="../../quickstart/quickstart/" class="md-tabs__link">
        
  
    
  
  Quick Start

      </a>
    </li>
  

      
        
  
  
    
  
  
    
    
      <li class="md-tabs__item md-tabs__item--active">
        <a href="../../intro/" class="md-tabs__link">
          
  
    
  
  Vignettes

        </a>
      </li>
    
  

      
        
  
  
  
    
    
      <li class="md-tabs__item">
        <a href="../../../reference/sge_convert/" class="md-tabs__link">
          
  
    
  
  References

        </a>
      </li>
    
  

      
    </ul>
  </div>
</nav>
          
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              
              <div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    


  


<nav class="md-nav md-nav--primary md-nav--lifted" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href="../../.." title="cartloader" class="md-nav__button md-logo" aria-label="cartloader" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54"/></svg>

    </a>
    cartloader
  </label>
  
    <div class="md-nav__source">
      <a href="https://github.com/seqscope/cartloader" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 6.7.2 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2024 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81"/></svg>
  </div>
  <div class="md-source__repository">
    seqscope/cartloader
  </div>
</a>
    </div>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
  
  
  
  
    
    
      
        
          
        
      
    
    
      
      
        
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
          
        
        <input class="md-nav__toggle md-toggle md-toggle--indeterminate" type="checkbox" id="__nav_1" >
        
          
          
          <div class="md-nav__link md-nav__container">
            <a href="../../.." class="md-nav__link ">
              
  
  <span class="md-ellipsis">
    Home
    
  </span>
  

            </a>
            
          </div>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_1_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_1">
            <span class="md-nav__icon md-icon"></span>
            Home
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../installation/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Installation
    
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../../quickstart/quickstart/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Quick Start
    
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
    
  
  
  
    
    
      
        
      
        
      
    
    
      
        
        
      
      
        
      
    
    
    <li class="md-nav__item md-nav__item--active md-nav__item--section md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_4" checked>
        
          
          <label class="md-nav__link" for="__nav_4" id="__nav_4_label" tabindex="">
            
  
  <span class="md-ellipsis">
    Vignettes
    
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_4_label" aria-expanded="true">
          <label class="md-nav__title" for="__nav_4">
            <span class="md-nav__icon md-icon"></span>
            Vignettes
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../intro/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Overview
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
    
  
  
  
    
    
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
    
    
      
      
        
          
          
        
      
    
    
    <li class="md-nav__item md-nav__item--active md-nav__item--section md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_4_2" checked>
        
          
          <label class="md-nav__link" for="__nav_4_2" id="__nav_4_2_label" tabindex="">
            
  
  <span class="md-ellipsis">
    Getting Started with `cartloader`
    
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="2" aria-labelledby="__nav_4_2_label" aria-expanded="true">
          <label class="md-nav__title" for="__nav_4_2">
            <span class="md-nav__icon md-icon"></span>
            Getting Started with `cartloader`
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../seqscope/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    SeqScope Starter Tutorial
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../stereoseq/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    StereoSeq Starter Tutorial
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../visiumhd/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Visium HD Starter Tutorial
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../pixelseq/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Pixel-Seq Starter Tutorial
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../xenium/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Xenium Starter Tutorial
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../merscope/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    MERSCOPE Starter Tutorial
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
    
  
  
  
    <li class="md-nav__item md-nav__item--active">
      
      <input class="md-nav__toggle md-toggle" type="checkbox" id="__toc">
      
      
        
      
      
        <label class="md-nav__link md-nav__link--active" for="__toc">
          
  
  <span class="md-ellipsis">
    CoxMX SMI Starter Tutorial
    
  </span>
  

          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <a href="./" class="md-nav__link md-nav__link--active">
        
  
  <span class="md-ellipsis">
    CoxMX SMI Starter Tutorial
    
  </span>
  

      </a>
      
        

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#input-data" class="md-nav__link">
    <span class="md-ellipsis">
      Input Data
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#set-up-the-environment" class="md-nav__link">
    <span class="md-ellipsis">
      Set Up the Environment
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#sge-format-conversion" class="md-nav__link">
    <span class="md-ellipsis">
      SGE Format Conversion
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#ficture-analysis" class="md-nav__link">
    <span class="md-ellipsis">
      FICTURE analysis
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#cartloader-compilation" class="md-nav__link">
    <span class="md-ellipsis">
      cartloader Compilation
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#upload-to-data-repository" class="md-nav__link">
    <span class="md-ellipsis">
      Upload to Data Repository
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#output-data" class="md-nav__link">
    <span class="md-ellipsis">
      Output Data
    </span>
  </a>
  
</li>
      
    </ul>
  
</nav>
      
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
  
  
    
    
      
        
      
        
      
        
      
        
      
    
    
      
      
        
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
          
        
        <input class="md-nav__toggle md-toggle md-toggle--indeterminate" type="checkbox" id="__nav_5" >
        
          
          <label class="md-nav__link" for="__nav_5" id="__nav_5_label" tabindex="0">
            
  
  <span class="md-ellipsis">
    References
    
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_5_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_5">
            <span class="md-nav__icon md-icon"></span>
            References
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../reference/sge_convert/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    SGE Format Conversion
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../reference/run_ficture2/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    FICTURE Analysis
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../reference/run_cartload2/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    cartloader Compilation
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../../reference/upload_repository/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Data Repository Upload
    
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
              
              <div class="md-sidebar md-sidebar--secondary" data-md-component="sidebar" data-md-type="toc" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#input-data" class="md-nav__link">
    <span class="md-ellipsis">
      Input Data
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#set-up-the-environment" class="md-nav__link">
    <span class="md-ellipsis">
      Set Up the Environment
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#sge-format-conversion" class="md-nav__link">
    <span class="md-ellipsis">
      SGE Format Conversion
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#ficture-analysis" class="md-nav__link">
    <span class="md-ellipsis">
      FICTURE analysis
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#cartloader-compilation" class="md-nav__link">
    <span class="md-ellipsis">
      cartloader Compilation
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#upload-to-data-repository" class="md-nav__link">
    <span class="md-ellipsis">
      Upload to Data Repository
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#output-data" class="md-nav__link">
    <span class="md-ellipsis">
      Output Data
    </span>
  </a>
  
</li>
      
    </ul>
  
</nav>
                  </div>
                </div>
              </div>
            
          
          
            <div class="md-content" data-md-component="content">
              <article class="md-content__inner md-typeset">
                
                  


  
  


<h1 id="cosmx-smi-starter-tutorial">CosMX SMI Starter Tutorial<a class="headerlink" href="#cosmx-smi-starter-tutorial" title="Permanent link">&para;</a></h1>
<h2 id="input-data">Input Data<a class="headerlink" href="#input-data" title="Permanent link">&para;</a></h2>
<p>The input data is from an adult mouse hippocampus, extracted by masking a coronal brain section. The original full-section </p>
<p><strong>File Format</strong></p>
<p>The CosMx SMI by NanoString generates high-resolution spatial transcriptomics data with single-molecule resolution with a comma-separated values (CSV) table.</p>
<div class="admonition info">
<p class="admonition-title">CSV File Format</p>
<div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span>
<span class="normal">2</span>
<span class="normal">3</span>
<span class="normal">4</span></pre></div></td><td class="code"><div><pre><span></span><code>&quot;fov&quot;,&quot;cell_ID&quot;,&quot;x_global_px&quot;,&quot;y_global_px&quot;,&quot;x_local_px&quot;,&quot;y_local_px&quot;,&quot;z&quot;,&quot;target&quot;,&quot;CellComp&quot;
64,0,-473043,7954.533,4015.3,4246.2,1,&quot;Gfap&quot;,&quot;None&quot;
64,0,-473022.9,7902.723,4035.48,4194.39,1,&quot;Fth1&quot;,&quot;None&quot;
64,0,-473132,7836.476,3926.34,4128.143,1,&quot;Ptn&quot;,&quot;None&quot;
</code></pre></div></td></tr></table></div>
<ul>
<li><code>fov</code>: The field of view (FOV) number.</li>
<li><code>cell_ID</code>: Unique identifier for a single cell within a given FOV. 0 if background or unassigned molecules.</li>
<li><code>x_global_px</code>, <code>y_global_px</code>: Global pixel coordinates relative to the tisse.</li>
<li><code>x_local_px</code>, <code>y_local_px</code>: The x or y position (in pixels) relative to the given FOV.</li>
<li><code>z</code>: Z-plane index representing the depth (optical section) where the transcript was detected.</li>
<li><code>target</code>: Name of the target.</li>
<li><code>CellComp</code>: Subcellular location of target.</li>
</ul>
</div>
<p><strong>Data Access</strong></p>
<p>The example data is hosted on Zenedo ().</p>
<p>Follow the commands below to download the example data.</p>
<div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span>
<span class="normal">2</span>
<span class="normal">3</span>
<span class="normal">4</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="nv">work_dir</span><span class="o">=</span>/path/to/work/directory
<span class="nb">cd</span><span class="w"> </span><span class="nv">$work_dir</span>
wget<span class="w">  </span>https://zenodo.org/records/15786632/files/cosmxsmi_starter.raw.tar.gz
tar<span class="w"> </span>--strip-components<span class="o">=</span><span class="m">1</span><span class="w"> </span>-zxvf<span class="w"> </span>cosmxsmi_starter.raw.tar.gz
</code></pre></div></td></tr></table></div>
<h2 id="set-up-the-environment">Set Up the Environment<a class="headerlink" href="#set-up-the-environment" title="Permanent link">&para;</a></h2>
<p>Define paths to all required binaries and resources, and target AWS S3 bucket. Optionally, specify a fixed color map for consistent rendering.</p>
<div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"> 1</span>
<span class="normal"> 2</span>
<span class="normal"> 3</span>
<span class="normal"> 4</span>
<span class="normal"> 5</span>
<span class="normal"> 6</span>
<span class="normal"> 7</span>
<span class="normal"> 8</span>
<span class="normal"> 9</span>
<span class="normal">10</span>
<span class="normal">11</span>
<span class="normal">12</span>
<span class="normal">13</span>
<span class="normal">14</span>
<span class="normal">15</span>
<span class="normal">16</span>
<span class="normal">17</span>
<span class="normal">18</span>
<span class="normal">19</span>
<span class="normal">20</span>
<span class="normal">21</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="c1"># ====</span>
<span class="c1"># Replace each placeholder with the actual path on your system.  </span>
<span class="c1"># ====</span>
<span class="nv">work_dir</span><span class="o">=</span>/path/to/work/directory<span class="w">        </span><span class="c1"># path to work directory that contains the downloaded input data</span>
<span class="nb">cd</span><span class="w"> </span><span class="nv">$work_dir</span>

<span class="c1"># Define paths to required binaries and resources</span>
<span class="nv">spatula</span><span class="o">=</span>/path/to/spatula/binary<span class="w">         </span><span class="c1"># path to spatula executable</span>
<span class="nv">punkst</span><span class="o">=</span>/path/to/punkst/binary<span class="w">           </span><span class="c1"># path to FICTURE2/punkst executable</span>
<span class="nv">tippecanoe</span><span class="o">=</span>/path/to/tippecanoe/binary<span class="w">   </span><span class="c1"># path to tippecanoe executable</span>
<span class="nv">pmtiles</span><span class="o">=</span>/path/to/pmtiles/binary<span class="w">         </span><span class="c1"># path to pmtiles executable</span>
<span class="nv">aws</span><span class="o">=</span>/path/to/aws/cli/binary<span class="w">             </span><span class="c1"># path to AWS CLI binary</span>

<span class="c1"># (Optional) Define path to color map. </span>
<span class="nv">cmap</span><span class="o">=</span>/path/to/color/map<span class="w">                 </span><span class="c1"># Path to the fixed color map for rendering. cartloader provides a fixed color map at cartloader/assets/fixed_color_map_256.tsv.</span>

<span class="c1"># AWS S3 target location for cartostore</span>
<span class="nv">AWS_BUCKET</span><span class="o">=</span><span class="s2">&quot;EXAMPLE_AWS_BUCKET&quot;</span><span class="w">         </span><span class="c1"># replace EXAMPLE_AWS_BUCKET with your actual S3 bucket name</span>

<span class="c1"># Activate the bioconda environment</span>
conda<span class="w"> </span>activate<span class="w"> </span>ENV_NAME<span class="w">              </span><span class="c1"># replace BIOENV_NAME with your bioconda environment name</span>
</code></pre></div></td></tr></table></div>
<p>Define data ID and analysis parameters:</p>
<div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span>
<span class="normal">2</span>
<span class="normal">3</span>
<span class="normal">4</span>
<span class="normal">5</span>
<span class="normal">6</span>
<span class="normal">7</span>
<span class="normal">8</span></pre></div></td><td class="code"><div><pre><span></span><code><span class="c1"># Unique identifier for your dataset</span>
<span class="nv">DATA_ID</span><span class="o">=</span><span class="s2">&quot;cosmxsmi_hippo&quot;</span><span class="w">                </span><span class="c1"># change this to reflect your dataset name</span>
<span class="nv">PLATFORM</span><span class="o">=</span><span class="s2">&quot;cosmx_smi&quot;</span><span class="w">                    </span><span class="c1"># platform information</span>
<span class="nv">SCALE</span><span class="o">=</span><span class="k">$(</span><span class="nb">echo</span><span class="w"> </span><span class="m">1000</span>/120<span class="p">|</span>bc<span class="w"> </span>-l<span class="k">)</span><span class="w">              </span><span class="c1"># scale from coordinate to micrometer</span>

<span class="c1"># LDA parameters</span>
<span class="nv">train_width</span><span class="o">=</span><span class="m">12</span><span class="w">                           </span><span class="c1"># define LDA training hexagon width (comma-separated if multiple widths are applied)</span>
<span class="nv">n_factor</span><span class="o">=</span><span class="m">6</span>,12<span class="w">                            </span><span class="c1"># define number of factors in LDA training (comma-separated if multiple n-factor are applied)</span>
</code></pre></div></td></tr></table></div>
<div class="admonition info">
<p class="admonition-title">How to Define Scaling Factors for CosMX SMI?</p>
<p>According to the README.html provided with the Pixel-seq dataset, each pixel has an edge length of 120 nm. To calculate the number of pixels per micrometer, use the formula: scale = 1000 / 120.</p>
</div>
<h2 id="sge-format-conversion"><abbr title="Spatial Digital Gene Expression">SGE</abbr> Format Conversion<a class="headerlink" href="#sge-format-conversion" title="Permanent link">&para;</a></h2>
<p>Convert the raw input to the unified <abbr title="Spatial Digital Gene Expression">SGE</abbr> format. See more details in <a href="../../../reference/sge_convert/">Reference page</a>.</p>
<div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"> 1</span>
<span class="normal"> 2</span>
<span class="normal"> 3</span>
<span class="normal"> 4</span>
<span class="normal"> 5</span>
<span class="normal"> 6</span>
<span class="normal"> 7</span>
<span class="normal"> 8</span>
<span class="normal"> 9</span>
<span class="normal">10</span></pre></div></td><td class="code"><div><pre><span></span><code>cartloader<span class="w"> </span>sge_convert<span class="w"> </span><span class="se">\</span>
<span class="w">  </span>--makefn<span class="w"> </span>sge_convert.mk<span class="w"> </span><span class="se">\</span>
<span class="w">  </span>--platform<span class="w"> </span><span class="si">${</span><span class="nv">PLATFORM</span><span class="si">}</span><span class="w"> </span><span class="se">\</span>
<span class="w">  </span>--in-csv<span class="w"> </span>./input.tsv.gz<span class="w"> </span><span class="se">\</span>
<span class="w">  </span>--units-per-um<span class="w"> </span><span class="si">${</span><span class="nv">SCALE</span><span class="si">}</span><span class="w"> </span><span class="se">\</span>
<span class="w">  </span>--out-dir<span class="w"> </span>./sge<span class="w"> </span><span class="se">\</span>
<span class="w">  </span>--exclude-feature-regex<span class="w"> </span><span class="s1">&#39;^(BLANK|NegCon|NegPrb)&#39;</span><span class="w"> </span><span class="se">\</span>
<span class="w">  </span>--sge-visual<span class="w"> </span><span class="se">\</span>
<span class="w">  </span>--spatula<span class="w"> </span><span class="si">${</span><span class="nv">spatula</span><span class="si">}</span><span class="w"> </span><span class="se">\</span>
<span class="w">  </span>--n-jobs<span class="w"> </span><span class="m">10</span>
</code></pre></div></td></tr></table></div>
<table>
<thead>
<tr>
<th>Parameter</th>
<th>Required</th>
<th>Type</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr>
<td><code>--platform</code></td>
<td>required</td>
<td>string</td>
<td>Platform (options: "<code>10x_visium_hd</code>", "<code>seqscope</code>", "<code>10x_xenium</code>", "<code>bgi_stereoseq</code>", "<code>cosmx_smi</code>", "<code>vizgen_merscope</code>", "<code>pixel_seq</code>", "<code>generic</code>")</td>
</tr>
<tr>
<td><code>--in-csv</code></td>
<td>required</td>
<td>string</td>
<td>Path to the input TSV/CSV file</td>
</tr>
<tr>
<td><code>--units-per-um</code></td>
<td>required</td>
<td>float</td>
<td>Scale to convert coordinates to microns (default: <code>1.0</code>)</td>
</tr>
<tr>
<td><code>--out-dir</code></td>
<td>required</td>
<td>string</td>
<td>Output directory for the converted <abbr title="Spatial Digital Gene Expression">SGE</abbr> files</td>
</tr>
<tr>
<td><code>--makefn</code></td>
<td></td>
<td>string</td>
<td>File name for the generated Makefile (default: <code>sge_convert.mk</code>)</td>
</tr>
<tr>
<td><code>--exclude-feature-regex</code></td>
<td></td>
<td>regex</td>
<td>Pattern to exclude control features</td>
</tr>
<tr>
<td><code>--sge-visual</code></td>
<td></td>
<td>flag</td>
<td>Enable <abbr title="Spatial Digital Gene Expression">SGE</abbr> visualization step (generates diagnostic image) (default: <code>FALSE</code>)</td>
</tr>
<tr>
<td><code>--spatula</code></td>
<td></td>
<td>string</td>
<td>Path to the spatula binary (default: <code>spatula</code>)</td>
</tr>
<tr>
<td><code>--n-jobs</code></td>
<td></td>
<td>int</td>
<td>Number of parallel jobs for processing (default: <code>1</code>)</td>
</tr>
</tbody>
</table>
<h2 id="ficture-analysis"><code>FICTURE</code> analysis<a class="headerlink" href="#ficture-analysis" title="Permanent link">&para;</a></h2>
<p>Compute spatial factors using <code>punkst</code> (FICTURE2 mode). See more details in <a href="../../../reference/run_ficture2/">Reference page</a>.</p>
<!-- ```bash
cartloader run_ficture2 \
    --makefn run_ficture2.mk \                          # (optional) file name of the output make file
    --main \                                            # run all five steps in `run_ficture2`
    --in-transcript ./sge/transcripts.unsorted.tsv.gz \ # path to input transcript-level SGE file
    --in-feature ./sge/feature.clean.tsv.gz \           # (optional) path to input feature file
    --in-minmax ./sge/coordinate_minmax.tsv \           # (optional) path to input minmax file
    --cmap-file ${cmap} \                               # (optional) path to input color map file
    --exclude-feature-regex '^(mt-.*$|Gm\d+$)' \        # regex pattern to exclude features (removing mitochondrial and predicted genes in the example analysis)
    --out-dir ./ficture2 \                              # path to output directory
    --width ${train_width} \                            # LDA training hexagon width (comma-separated if multiple widths are applied)
    --n-factor ${n_factor} \                            # number of factors in LDA training (comma-separated if multiple n-factor are applied)
    --spatula ${spatula} \                              # (optional) path to the spatula binary
    --ficture2 ${punkst} \                              # (optional) path to the punkst directory
    --n-jobs 10  \                                      # (optional) number of parallel jobs 
    --threads 10                                        # (optional) number of threads per job
``` -->

<div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"> 1</span>
<span class="normal"> 2</span>
<span class="normal"> 3</span>
<span class="normal"> 4</span>
<span class="normal"> 5</span>
<span class="normal"> 6</span>
<span class="normal"> 7</span>
<span class="normal"> 8</span>
<span class="normal"> 9</span>
<span class="normal">10</span>
<span class="normal">11</span>
<span class="normal">12</span>
<span class="normal">13</span>
<span class="normal">14</span>
<span class="normal">15</span></pre></div></td><td class="code"><div><pre><span></span><code>cartloader<span class="w"> </span>run_ficture2<span class="w"> </span><span class="se">\</span>
<span class="w">  </span>--makefn<span class="w"> </span>run_ficture2.mk<span class="w"> </span><span class="se">\</span>
<span class="w">  </span>--main<span class="w"> </span><span class="se">\</span>
<span class="w">  </span>--in-transcript<span class="w"> </span>./sge/transcripts.unsorted.tsv.gz<span class="w"> </span><span class="se">\</span>
<span class="w">  </span>--in-feature<span class="w"> </span>./sge/feature.clean.tsv.gz<span class="w"> </span><span class="se">\</span>
<span class="w">  </span>--in-minmax<span class="w"> </span>./sge/coordinate_minmax.tsv<span class="w"> </span><span class="se">\</span>
<span class="w">  </span>--cmap-file<span class="w"> </span><span class="si">${</span><span class="nv">cmap</span><span class="si">}</span><span class="w"> </span><span class="se">\</span>
<span class="w">  </span>--exclude-feature-regex<span class="w"> </span><span class="s1">&#39;^(mt-.*$|Gm\d+$)&#39;</span><span class="w"> </span><span class="se">\</span>
<span class="w">  </span>--out-dir<span class="w"> </span>./ficture2<span class="w"> </span><span class="se">\</span>
<span class="w">  </span>--width<span class="w"> </span><span class="si">${</span><span class="nv">train_width</span><span class="si">}</span><span class="w"> </span><span class="se">\</span>
<span class="w">  </span>--n-factor<span class="w"> </span><span class="si">${</span><span class="nv">n_factor</span><span class="si">}</span><span class="w"> </span><span class="se">\</span>
<span class="w">  </span>--spatula<span class="w"> </span><span class="si">${</span><span class="nv">spatula</span><span class="si">}</span><span class="w"> </span><span class="se">\</span>
<span class="w">  </span>--ficture2<span class="w"> </span><span class="si">${</span><span class="nv">punkst</span><span class="si">}</span><span class="w"> </span><span class="se">\</span>
<span class="w">  </span>--n-jobs<span class="w"> </span><span class="m">10</span><span class="w"> </span><span class="se">\</span>
<span class="w">  </span>--threads<span class="w"> </span><span class="m">10</span>
</code></pre></div></td></tr></table></div>
<!--parameter-start-->
<table>
<thead>
<tr>
<th>Parameter</th>
<th>Required</th>
<th>Type</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr>
<td><code>--main</code></td>
<td>required <sup>1</sup></td>
<td>flag</td>
<td>Enable <code>cartloader</code> to run <a href="../../../reference/run_ficture2/#actions">all five steps</a></td>
</tr>
<tr>
<td><code>--in-transcript</code></td>
<td>required</td>
<td>string</td>
<td>Path to input transcript-level <abbr title="Spatial Digital Gene Expression">SGE</abbr> file</td>
</tr>
<tr>
<td><code>--out-dir</code></td>
<td>required</td>
<td>string</td>
<td>Path to output directory</td>
</tr>
<tr>
<td><code>--width</code></td>
<td>required</td>
<td>int or comma-separated list</td>
<td><abbr title="Latent Dirichlet Allocation">LDA</abbr> training hexagon width(s)</td>
</tr>
<tr>
<td><code>--n-factor</code></td>
<td>required</td>
<td>int or comma-separated list</td>
<td>Number of <abbr title="Latent Dirichlet Allocation">LDA</abbr> factors</td>
</tr>
<tr>
<td><code>--makefn</code></td>
<td></td>
<td>string</td>
<td>File name for the generated Makefile (default: <code>run_ficture2.mk</code> )</td>
</tr>
<tr>
<td><code>--in-feature</code></td>
<td></td>
<td>string</td>
<td>Path to input feature file</td>
</tr>
<tr>
<td><code>--in-minmax</code></td>
<td></td>
<td>string</td>
<td>Path to input coordinate min/max file</td>
</tr>
<tr>
<td><code>--cmap-file</code></td>
<td></td>
<td>string</td>
<td>Path to color map file</td>
</tr>
<tr>
<td><code>--exclude-feature-regex</code></td>
<td></td>
<td>regex</td>
<td>Pattern to exclude features</td>
</tr>
<tr>
<td><code>--spatula</code></td>
<td></td>
<td>string</td>
<td>Path to the <code>spatula</code> binary (default: <code>spatula</code>)</td>
</tr>
<tr>
<td><code>--ficture2</code></td>
<td></td>
<td>string</td>
<td>Path to the <code>punkst</code> directory (defaults to <code>punkst</code> repository within <code>submodules</code> directory of  <code>cartloader</code>)</td>
</tr>
<tr>
<td><code>--n-jobs</code></td>
<td></td>
<td>int</td>
<td>Number of parallel jobs (default: <code>1</code>)</td>
</tr>
<tr>
<td><code>--threads</code></td>
<td></td>
<td>int</td>
<td>Number of threads per job (default: <code>1</code>)</td>
</tr>
</tbody>
</table>
<p><sub><sup>1</sup>: <code>cartloader</code> requires the user to specify <strong>at least one action</strong>. Available actions includes: <code>--tile</code> to run tiling step; <code>--segment</code> to run segmentation step; <code>--init-lda</code> to run <abbr title="Latent Dirichlet Allocation">LDA</abbr> training step; <code>--decode</code> to run decoding step; <code>--summary</code> to run summarization step; <code>--main</code> to run all above five actions.</sub></p>
<!--parameter-end-->

<h2 id="cartloader-compilation"><code>cartloader</code> Compilation<a class="headerlink" href="#cartloader-compilation" title="Permanent link">&para;</a></h2>
<p>Generate pmtiles and web-compatible tile directories. See more details in <a href="../../../reference/run_cartload2/">Reference page</a>.</p>
<!-- ```bash
cartloader run_cartload2 \
    --makefn run_cartload2.mk \         # (optional) file name of the output make file
    --fic-dir ./ficture2 \              # path to input directory containing FICTURE2 output
    --out-dir ./cartload2 \             # path to output directory for PMTiles and web tiles
    --id ${DATA_ID} \                   # dataset ID used for naming outputs and metadata
    --spatula ${spatula} \              # (optional) path to the spatula binary
    --pmtiles ${pmtiles} \              # (optional) path to the pmtiles binary
    --tippecanoe ${tippecanoe} \        # (optional) path to the tippecanoe binary
    --n-jobs 10 \                       # (optional) number of parallel jobs
    --threads 10                        # (optional) number of threads per job
``` -->

<div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"> 1</span>
<span class="normal"> 2</span>
<span class="normal"> 3</span>
<span class="normal"> 4</span>
<span class="normal"> 5</span>
<span class="normal"> 6</span>
<span class="normal"> 7</span>
<span class="normal"> 8</span>
<span class="normal"> 9</span>
<span class="normal">10</span></pre></div></td><td class="code"><div><pre><span></span><code>cartloader<span class="w"> </span>run_cartload2<span class="w"> </span><span class="se">\</span>
<span class="w">  </span>--makefn<span class="w"> </span>run_cartload2.mk<span class="w"> </span><span class="se">\</span>
<span class="w">  </span>--fic-dir<span class="w"> </span>./ficture2<span class="w"> </span><span class="se">\</span>
<span class="w">  </span>--out-dir<span class="w"> </span>./cartload2<span class="w"> </span><span class="se">\</span>
<span class="w">  </span>--id<span class="w"> </span><span class="si">${</span><span class="nv">DATA_ID</span><span class="si">}</span><span class="w"> </span><span class="se">\</span>
<span class="w">  </span>--spatula<span class="w"> </span><span class="si">${</span><span class="nv">spatula</span><span class="si">}</span><span class="w"> </span><span class="se">\</span>
<span class="w">  </span>--pmtiles<span class="w"> </span><span class="si">${</span><span class="nv">pmtiles</span><span class="si">}</span><span class="w"> </span><span class="se">\</span>
<span class="w">  </span>--tippecanoe<span class="w"> </span><span class="si">${</span><span class="nv">tippecanoe</span><span class="si">}</span><span class="w"> </span><span class="se">\</span>
<span class="w">  </span>--n-jobs<span class="w"> </span><span class="m">10</span><span class="w"> </span><span class="se">\</span>
<span class="w">  </span>--threads<span class="w"> </span><span class="m">10</span>
</code></pre></div></td></tr></table></div>
<table>
<thead>
<tr>
<th>Parameter</th>
<th>Required</th>
<th>Type</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr>
<td><code>--fic-dir</code></td>
<td>required</td>
<td>string</td>
<td>Path to the input directory containing FICTURE2 output</td>
</tr>
<tr>
<td><code>--out-dir</code></td>
<td>required</td>
<td>string</td>
<td>Path to the output directory for PMTiles and web tiles</td>
</tr>
<tr>
<td><code>--id</code></td>
<td>required</td>
<td>string</td>
<td>Dataset ID used for naming outputs and metadata</td>
</tr>
<tr>
<td><code>--makefn</code></td>
<td></td>
<td>string</td>
<td>File name for the generated Makefile (default: <code>run_cartload2.mk</code>)</td>
</tr>
<tr>
<td><code>--spatula</code></td>
<td></td>
<td>string</td>
<td>Path to the <code>spatula</code> binary (default: <code>spatula</code>)</td>
</tr>
<tr>
<td><code>--pmtiles</code></td>
<td></td>
<td>string</td>
<td>Path to the <code>pmtiles</code> binary (default: <code>pmtiles</code>)</td>
</tr>
<tr>
<td><code>--tippecanoe</code></td>
<td></td>
<td>string</td>
<td>Path to the <code>tippecanoe</code> binary (default: <code>tippecanoe</code>)</td>
</tr>
<tr>
<td><code>--n-jobs</code></td>
<td></td>
<td>int</td>
<td>Number of parallel jobs (default: <code>1</code>)</td>
</tr>
<tr>
<td><code>--threads</code></td>
<td></td>
<td>int</td>
<td>Number of threads per job (default: <code>1</code>)</td>
</tr>
</tbody>
</table>
<h2 id="upload-to-data-repository">Upload to Data Repository<a class="headerlink" href="#upload-to-data-repository" title="Permanent link">&para;</a></h2>
<div class="admonition info">
<p class="admonition-title">Choose a data repository to host/share your output</p>
<p><code>cartloader</code> supports two upload options (<code>AWS</code> and <code>Zenodo</code>) for storing PMTiles of <abbr title="Spatial Digital Gene Expression">SGE</abbr> and spatial factors in a data repository.</p>
<p><strong>Choose the one that best suits your needs.</strong></p>
</div>
<h3 id="aws-uploads">AWS Uploads<a class="headerlink" href="#aws-uploads" title="Permanent link">&para;</a></h3>
<p>Upload the generated cartloader outputs to your designated AWS S3 directory:</p>
<div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span>
<span class="normal">2</span>
<span class="normal">3</span>
<span class="normal">4</span>
<span class="normal">5</span></pre></div></td><td class="code"><div><pre><span></span><code>cartloader<span class="w"> </span>upload_aws<span class="w"> </span><span class="se">\</span>
<span class="w">  </span>--in-dir<span class="w"> </span>./cartload2<span class="w"> </span><span class="se">\</span>
<span class="w">  </span>--s3-dir<span class="w"> </span><span class="s2">&quot;s3://</span><span class="si">${</span><span class="nv">AWS_BUCKET</span><span class="si">}</span><span class="s2">/</span><span class="si">${</span><span class="nv">DATA_ID</span><span class="si">}</span><span class="s2">&quot;</span><span class="w"> </span><span class="se">\</span>
<span class="w">  </span>--aws<span class="w"> </span><span class="si">${</span><span class="nv">aws</span><span class="si">}</span><span class="w"> </span><span class="se">\</span>
<span class="w">  </span>--n-jobs<span class="w"> </span><span class="m">10</span>
</code></pre></div></td></tr></table></div>
<table>
<thead>
<tr>
<th>Parameter</th>
<th>Required</th>
<th>Type</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr>
<td><code>--in-dir</code></td>
<td>required</td>
<td>string</td>
<td>Path to the input directory containing the cartloader compilation output</td>
</tr>
<tr>
<td><code>--s3-dir</code></td>
<td>required</td>
<td>string</td>
<td>Path to the target S3 directory for uploading</td>
</tr>
<tr>
<td><code>--aws</code></td>
<td></td>
<td>string</td>
<td>Path to the AWS CLI binary</td>
</tr>
<tr>
<td><code>--n-jobs</code></td>
<td></td>
<td>int</td>
<td>Number of parallel jobs</td>
</tr>
</tbody>
</table>
<h3 id="zenodo-uploads">Zenodo Uploads<a class="headerlink" href="#zenodo-uploads" title="Permanent link">&para;</a></h3>
<p>Upload the generated cartloader outputs to your designated Zenodo deposition or a new deposition.</p>
<div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span>
<span class="normal">2</span>
<span class="normal">3</span>
<span class="normal">4</span>
<span class="normal">5</span>
<span class="normal">6</span>
<span class="normal">7</span>
<span class="normal">8</span></pre></div></td><td class="code"><div><pre><span></span><code>cartloader<span class="w"> </span>upload_zenodo<span class="w"> </span><span class="se">\</span>
<span class="w">  </span>--in-dir<span class="w"> </span>./cartload2<span class="w"> </span><span class="se">\</span>
<span class="w">  </span>--upload-method<span class="w"> </span>catalog<span class="w"> </span><span class="se">\</span>
<span class="w">  </span>--zenodo-token<span class="w"> </span>/path/to/zenodo/token/file<span class="w"> </span><span class="se">\</span>
<span class="w">  </span>--create-new-deposition<span class="w"> </span><span class="se">\</span>
<span class="w">  </span>--title<span class="w">  </span><span class="s2">&quot;Yur Title&quot;</span><span class="w"> </span><span class="se">\</span>
<span class="w">  </span>--creators<span class="w"> </span><span class="s2">&quot;Your Name&quot;</span><span class="w"> </span><span class="se">\</span>
<span class="w">  </span>--description<span class="w"> </span><span class="s2">&quot;This is an example description&quot;</span>
</code></pre></div></td></tr></table></div>
<table>
<thead>
<tr>
<th>Parameter</th>
<th>Required</th>
<th>Type</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr>
<td><code>--in-dir</code></td>
<td>required</td>
<td>string</td>
<td>Path to the input directory containing the cartloader compilation output</td>
</tr>
<tr>
<td><code>--upload-method</code></td>
<td>required</td>
<td>string</td>
<td>Method to determine which files to upload. Options: <code>all</code> to upload all files in <code>--in-dir</code>; <code>catalog</code> to upload files listed in a catalog YAML file, <code>user_list</code> to upload files explicitly listed via <code>--in-list</code></td>
</tr>
<tr>
<td><code>--catalog-yaml</code></td>
<td></td>
<td>string</td>
<td>Required if <code>--upload-method catalog</code>.  Path to the catalog.yaml file generated in <code>run_cartload2</code>. If absent, will use the catalog.yaml in the input directory specified by <code>--in-dir</code>.</td>
</tr>
<tr>
<td><code>--zenodo-token</code></td>
<td>required</td>
<td>string</td>
<td>Path to your Zenodo access file</td>
</tr>
<tr>
<td><code>--create-new-deposition</code></td>
<td></td>
<td>flag</td>
<td>a new Zenodo deposition will be created.</td>
</tr>
<tr>
<td><code>--title</code></td>
<td>required</td>
<td>string</td>
<td>Required if <code>--create-new-deposition</code>. Title for the new Zenodo deposition.</td>
</tr>
<tr>
<td><code>--creators</code></td>
<td>required</td>
<td>list of str</td>
<td>List of creators in "Lastname, Firstname" format.</td>
</tr>
</tbody>
</table>
<h2 id="output-data">Output Data<a class="headerlink" href="#output-data" title="Permanent link">&para;</a></h2>
<p>See more details of output at the Reference pages for <a href="../../../reference/run_ficture2/">run_ficture2</a> and <a href="../../../reference/run_cartload2/">run_cartload2</a>.</p>
<h3 id="spatial-factor-inference-from-ficture">Spatial Factor Inference from <code>FICTURE</code><a class="headerlink" href="#spatial-factor-inference-from-ficture" title="Permanent link">&para;</a></h3>
<p>Below is an example of spatial factor inference results produced by <code>FICTURE</code> using a training width of 12, 12 factors, a fit width of 12, and an anchor resolution of 6.</p>
<p><img alt="FICTURE" src="../../../images/starter_vignettes/cosmxsmi.t12_f12_p12_a6.png" title="Spatial factor map" /></p>
<div id="color-legend-container"></div>
<script>
  document.addEventListener("DOMContentLoaded", function() {
    loadColorLegend("../../../tabs/colormap/cosmxsmi_starter.t12-f12-rgb.tsv", "color-legend-container");
  });
</script>

<table>
<thead>
<tr>
<th style="text-align: right;">Factor</th>
<th style="text-align: left;">RGB</th>
<th style="text-align: right;">Weight</th>
<th style="text-align: right;">PostUMI</th>
<th style="text-align: left;">TopGene_pval</th>
<th style="text-align: left;">TopGene_fc</th>
<th style="text-align: left;">TopGene_weight</th>
</tr>
</thead>
<tbody>
<tr>
<td style="text-align: right;">1</td>
<td style="text-align: left;">237,238,0</td>
<td style="text-align: right;">0.17207</td>
<td style="text-align: right;">548469625</td>
<td style="text-align: left;">Tmsb4x,Calm2,Snap25,Calm1,Ywhaz,Slc17a7,Cck,Dnm1,Ywhag,Aldoa,Hsp90ab1,Calm3,Atp2b1,Nell2,Rtn3,Snca,Gpm6a,Prkcb,Prkca,Cyfip2</td>
<td style="text-align: left;">Cck,Camk4,Prkca,Camk1d,Slc17a7,Ywhag,Epha6,Atp2b1,Calm2,Sv2b,Snca,Syn2,Prkcb,Lpl,Nell2,Snap25,Stmn2,Pde1a,Kcnq5,Cnksr2</td>
<td style="text-align: left;">Tmsb4x,Snap25,Calm2,Calm1,Ywhaz,Atp1b1,Aldoa,Hsp90ab1,Dnm1,Slc17a7,Rtn3,Calm3,Camk2a,Ywhag,Gpm6a,Cck,Ppp3ca,Rtn1,Nrgn,Cpe</td>
</tr>
<tr>
<td style="text-align: right;">0</td>
<td style="text-align: left;">255,101,101</td>
<td style="text-align: right;">0.16</td>
<td style="text-align: right;">510000383</td>
<td style="text-align: left;">Camk2b,Adcy1,Ppp3ca,Kalrn,Nrgn,Olfm1,Ppfia2,Ctxn1,Gria1,Gria2,Prkce,Ryr2,Camk2a,Grin2a,Gabrb3,Zbtb20,Wasf1,Tiam1,Ntrk3,Mapk1</td>
<td style="text-align: left;">Ppfia2,Adcy1,Sema5a,Camk2b,Kalrn,Plxna4,Calb1,Tiam1,Jun,Rasgrf2,Nedd4l,Dapk1,Grin2a,Npy1r,Itga8,Ryr2,Auts2,Ppp3ca,Dgkg,Grm7</td>
<td style="text-align: left;">Malat1,Camk2b,Camk2a,Calm1,Ppp3ca,Tmsb4x,Adcy1,Nrgn,Calm2,Olfm1,Atp1b1,Ndrg4,Gria2,Ywhaz,Kalrn,Gria1,Ctxn1,Prkce,Aldoa,Snap25</td>
</tr>
<tr>
<td style="text-align: right;">3</td>
<td style="text-align: left;">101,254,255</td>
<td style="text-align: right;">0.11601</td>
<td style="text-align: right;">369772892</td>
<td style="text-align: left;">Malat1,Hnrnpa2b1,Tardbp,Tnik,Grk3,Mid1,Ptprs,Nf1,Prkn,Vegfa,Kidins220,Tnrc6a,Dmd,Camk2g,Uggt2,Pou2f1,Meg3,Sema6d,Spag9,Ntsr2</td>
<td style="text-align: left;">Gal,Malat1,Avp,Cd109,Grk3,Lilra5,C2,Prkn,Crhr1,Ngfr,Gcgr,Dmd,Vegfa,C3,Dapk2,Pcsk6,Nrf1,Uggt2,P2rx4,Slc18a2</td>
<td style="text-align: left;">Malat1,Hnrnpa2b1,Hmgb1,Meg3,Gria2,Ptprs,Tardbp,Gnao1,Tnik,Apoe,Kidins220,Slc1a2,Ahcyl1,Nf1,Bcan,Mid1,Spag9,Camk2g,Fus,Tuba1a/b/c</td>
</tr>
<tr>
<td style="text-align: right;">5</td>
<td style="text-align: left;">255,101,254</td>
<td style="text-align: right;">0.09996</td>
<td style="text-align: right;">318610467</td>
<td style="text-align: left;">Nap1l5,Atp1b1,Gad1,Ndrg4,Meg3,Cnr1,Sncb,Gad2,Sst,Snap25,Syt1,Zwint,Tmsb10,Snrpn,Gnas,Npy,Sv2a,Cacnb4,Ldhb,Mdh1</td>
<td style="text-align: left;">Sst,Gad2,Npy,Cnr1,Nap1l5,Gad1,Pvalb,Slc32a1,Vip,Gap43,Adgrl2,Cntn4,Sncb,Atp2b4,Cckbr,Grik1,Grm8,Rims2,Cacnb4,Tmsb10</td>
<td style="text-align: left;">Atp1b1,Malat1,Snap25,Ndrg4,Meg3,Nap1l5,Ckb,Rtn3,Snrpn,Gnas,Calm1,Syt1,Zwint,Mdh1,Dnm1,Calm2,Aldoa,Sncb,Gad1,Dynll2</td>
</tr>
<tr>
<td style="text-align: right;">2</td>
<td style="text-align: left;">101,255,101</td>
<td style="text-align: right;">0.09064</td>
<td style="text-align: right;">288906127</td>
<td style="text-align: left;">Plp1,Ptgds,Scd2,Apod,Mag,Cryab,Ndrg1,Mog,Ugt8a,Bin1,Gpr37,Aspa,Mobp,Fa2h,Gsn,Mbp,Abca2,Cntn2,Slc44a1,Jam3</td>
<td style="text-align: left;">Plp1,Mag,Mog,Ugt8a,Aspa,Ptgds,Cryab,Apod,Fa2h,Myrf,Ndrg1,Gpr37,Gjb1,Pde8a,Cntn2,Gsn,Fgfr2,Scd2,Jam3,Abca2</td>
<td style="text-align: left;">Plp1,Malat1,Ptgds,Scd2,Mbp,Fth1,Glul,Apod,Mobp,Mag,Cryab,Bin1,Ndrg1,Gpm6b,Mog,Ywhaq,Cd81,Gpr37,Tmsb4x,Hmgb1</td>
</tr>
<tr>
<td style="text-align: right;">4</td>
<td style="text-align: left;">101,101,255</td>
<td style="text-align: right;">0.08492</td>
<td style="text-align: right;">270686250</td>
<td style="text-align: left;">Apoe,Clu,Slc1a2,Glul,Aldoc,Atp1a2,Slc1a3,Cst3,Gja1,Plpp3,Ckb,Ndrg2,Sparcl1,Gpr37l1,Gfap,Gstm1,Aqp4,Mt1,Mt3,Mfge8</td>
<td style="text-align: left;">Gja1,Slc1a3,Aldoc,Apoe,Atp1a2,Clu,Gpr37l1,Aqp4,Plpp3,Slc1a2,Gfap,Mt1,Mfge8,Gstm1,Slc6a11,Ndrg2,Glul,Ednrb,Ntsr2,Agt</td>
<td style="text-align: left;">Apoe,Clu,Glul,Slc1a2,Cst3,Aldoc,Malat1,Ckb,Atp1a2,Sparcl1,Slc1a3,Mt3,Cpe,Ndrg2,Plpp3,Scd2,Tspan7,Gja1,Ntrk2,Dbi</td>
</tr>
<tr>
<td style="text-align: right;">6</td>
<td style="text-align: left;">255,178,101</td>
<td style="text-align: right;">0.08061</td>
<td style="text-align: right;">256947664</td>
<td style="text-align: left;">Pcp4,Itm2c,Calb2,Nnat,Pcsk1n,Rtn1,Map1b,Cacna1e,Thy1,Clstn1,Cbln2,Gabbr2,Tcf7l2,Bex1/2,Rit2,Kcnd2,Slc2a13,Apba1,Tac1,Nsg1</td>
<td style="text-align: left;">Calb2,Cbln2,Tac1,Nnat,Oprm1,Tcf7l2,Pcp4,Apba1,Slc5a7,Rit2,Slc2a13,Sstr2,Rgs6,Itm2c,Cartpt,Synpr,Gabbr2,Pcsk1n,Gria4,Fgf1</td>
<td style="text-align: left;">Itm2c,Rtn1,Pcp4,Malat1,Map1b,Pcsk1n,Cpe,Atp1b1,Aldoa,Tuba1a/b/c,Hsp90ab1,Thy1,Psap,Rtn3,Eif4a2,Clstn1,Gnao1,Meg3,Cacna1e,Slc25a4</td>
</tr>
<tr>
<td style="text-align: right;">7</td>
<td style="text-align: left;">178,255,101</td>
<td style="text-align: right;">0.04347</td>
<td style="text-align: right;">138561767</td>
<td style="text-align: left;">Bsg,Slc2a1,Cldn5,Flt1,Itm2a,Rgs5,Pltp,Igfbp7,Serinc3,Esam,Vim,Slc7a5,Acta2,Igf1r,Id1,Fn1,Sptbn1,Srgn,Pecam1,St3gal6</td>
<td style="text-align: left;">Flt1,Cldn5,Itm2a,Rgs5,Pltp,Acta2,Fn1,Pecam1,Esam,Slc2a1,Igfbp7,Myl9,Lsr,Emcn,Slc7a5,Kdr,Tagln,Srgn,St3gal6,Vim</td>
<td style="text-align: left;">Bsg,Malat1,Slc2a1,Tmsb4x,Serinc3,Cldn5,Sptbn1,H3f3b,Flt1,Myl6,Aplp2,Itm2a,Gnb1,Pltp,Rgs5,Cpe,Itm2b,Hmgb1,Calm1,Igfbp7</td>
</tr>
<tr>
<td style="text-align: right;">11</td>
<td style="text-align: left;">255,153,204</td>
<td style="text-align: right;">0.0429</td>
<td style="text-align: right;">136738469</td>
<td style="text-align: left;">Ptprz1,Cspg5,Pdgfra,Olig1,Gpr17,Epn2,Tnr,Pllp,Vcan,Fyn,Serpine2,Cd9,Tuba1a/b/c,Olig2,Pcdh15,S100a16,S100b,Sulf2,Serinc5,Cntn1</td>
<td style="text-align: left;">Vcan,Pdgfra,Gpr17,Pcdh15,Olig2,Tnr,S100b,Olig1,Ptprz1,Cspg5,Fyn,Epn2,Itpr2,Serpine2,Cd9,S100a16,Sox6,Megf11,Pllp,Sulf2</td>
<td style="text-align: left;">Malat1,Tuba1a/b/c,Ptprz1,Cspg5,Olig1,Epn2,Pllp,Camk2a,Cd9,Serpine2,Calm1,Ncam1,Fyn,Bcan,Ckb,Tnr,Pdgfra,Cntn1,Hnrnpa2b1,Hmgb1</td>
</tr>
<tr>
<td style="text-align: right;">10</td>
<td style="text-align: left;">178,101,255</td>
<td style="text-align: right;">0.03812</td>
<td style="text-align: right;">121510243</td>
<td style="text-align: left;">Ttr,Psap,Cab39l,Chchd10,Lamp2,Cox8a,Htr2c,Slc12a2,Ppp1r1b,Timp2,Ctsd,Mdh1,Gpi1,Bsg,Ndufa4,Itpr1,Ftl1,App,Sem1,Itgb8</td>
<td style="text-align: left;">Ttr,Htr2c,Cab39l,Ppp1r1b,Lepr,Slc12a2,Col4a5,Foxj1,Lamp2,Msr1,Dlk1,Timp2,Ctnna1,Chchd10,Maob,Sil1,Itpr1,Tspo,Cd55,Dcn</td>
<td style="text-align: left;">Ttr,Psap,Malat1,Atp1b1,Cox8a,Mdh1,Bsg,Gpi1,Ctsd,Cab39l,Chchd10,App,Ndufa4,Clu,Ckb,Hspa8,Lamp2,Ptgds,Cst3,Slc12a2</td>
</tr>
<tr>
<td style="text-align: right;">9</td>
<td style="text-align: left;">101,178,255</td>
<td style="text-align: right;">0.03663</td>
<td style="text-align: right;">116755577</td>
<td style="text-align: left;">Cst3,Camk2a,C1qc,Hexb,Ctsd,C1qa,C1qb,Ctss,Csf1r,Vtn,P2ry12,Sparc,Selplg,Cx3cr1,Hmgb1,Tyrobp,Itm2b,Rps9,Tmem119,Tgfbr1</td>
<td style="text-align: left;">C1qc,Csf1r,Ctss,C1qb,C1qa,Hexb,Selplg,P2ry12,Tmem119,Cx3cr1,Tyrobp,Vtn,Trem2,Ptprc,Csf3r,Tgfbr1,Lyz1/2,Cd84,Higd1b,Epb41l2</td>
<td style="text-align: left;">Cst3,Camk2a,Tmsb4x,Hmgb1,Ctsd,Itm2b,Malat1,Rps9,Sparc,Hexb,Glul,C1qc,C1qa,Fau,C1qb,Ctsb,Ctss,Csf1r,Vtn,Fth1</td>
</tr>
<tr>
<td style="text-align: right;">8</td>
<td style="text-align: left;">0,223,95</td>
<td style="text-align: right;">0.03466</td>
<td style="text-align: right;">110484904</td>
<td style="text-align: left;">Mbp,Fth1,Mobp,Cpe,Bcas1,Hipk2,Kif5c,Pink1,Dync1li2,Gfap,Rims1,Cd6,Frs2,Lpar1,Ptpn11,Gpm6b,Map4k4,Ndrg1,S100b,Drd1</td>
<td style="text-align: left;">Mbp,Mobp,Bcas1,Fth1,Hipk2,Pink1,Cd6,Cpe,Dync1li2,Kif5c,Gfap,Rims1,Frs2,Lpar1,Ptpn11,S100b,Map4k4,Drd1,Gpm6b,Ndrg1</td>
<td style="text-align: left;">Mbp,Fth1,Mobp,Cpe,Malat1,Kif5c,Clu,Apoe,Glul,Plp1,Bcas1,Tmsb4x,Dync1li2,Hipk2,Pink1,Gfap,Hsp90ab1,Slc1a2,Gpm6b,Calm1</td>
</tr>
</tbody>
</table>
<h3 id="packed-sge-and-spatial-factor-outputs-from-run_cartload2">Packed <abbr title="Spatial Digital Gene Expression">SGE</abbr> and Spatial Factor Outputs from <code>run_cartload2</code><a class="headerlink" href="#packed-sge-and-spatial-factor-outputs-from-run_cartload2" title="Permanent link">&para;</a></h3>
<p>The packed <abbr title="Spatial Digital Gene Expression">SGE</abbr> data and spatial factor inferences generated by <code>FICTURE</code> are available in PMTile format on Zenodo: <a href="https://zenodo.org/records/15824926">DOI:10.5281/zenodo.15824926</a>.</p>
<p>These datasets can also be loaded directly using the following catalog YAML file:<br />
<code>https://zenodo.org/records/15824927/catalog.yaml</code></p>












                
              </article>
            </div>
          
          
<script>var target=document.getElementById(location.hash.slice(1));target&&target.name&&(target.checked=target.name.startsWith("__tabbed_"))</script>
        </div>
        
      </main>
      
        <footer class="md-footer">
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-copyright">
  
  
    Made with
    <a href="https://squidfunk.github.io/mkdocs-material/" target="_blank" rel="noopener">
      Material for MkDocs
    </a>
  
</div>
      
    </div>
  </div>
</footer>
      
    </div>
    <div class="md-dialog" data-md-component="dialog">
      <div class="md-dialog__inner md-typeset"></div>
    </div>
    
    
    <script id="__config" type="application/json">{"base": "../../..", "features": ["navigation.tabs", "navigation.indexes", "navigation.sections", "navigation.expand", "navigation.path", "content.code.annotate", "content.grid", "content.cards", "content.footnote.tooltips"], "search": "../../../assets/javascripts/workers/search.f8cc74c7.min.js", "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version": "Select version"}}</script>
    
    
      <script src="../../../assets/javascripts/bundle.f13b1293.min.js"></script>
      
        <script src="../../../js/load_tsv.js"></script>
      
        <script src="../../../js/open_in_new_tab.js"></script>
      
    
  </body>
</html>