
<!doctype html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
      
      
        <link rel="canonical" href="https://seqscope.github.io/cartloader/reference/sge_convert/">
      
      
        <link rel="prev" href="../../vignettes/subregion_tutorials/cosmx_smi/">
      
      
        <link rel="next" href="../run_ficture2/">
      
      
      <link rel="icon" href="../../assets/images/favicon.png">
      <meta name="generator" content="mkdocs-1.6.1, mkdocs-material-9.6.2">
    
    
      
        <title>SGE Format Conversion - cartloader</title>
      
    
    
      <link rel="stylesheet" href="../../assets/stylesheets/main.d7758b05.min.css">
      
        
        <link rel="stylesheet" href="../../assets/stylesheets/palette.06af60db.min.css">
      
      

  
  
  
  
  <style>:root{--md-annotation-icon:url('data:image/svg+xml;charset=utf-8,<svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M22 12a10 10 0 0 1-10 10A10 10 0 0 1 2 12 10 10 0 0 1 12 2a10 10 0 0 1 10 10M6 13h8l-3.5 3.5 1.42 1.42L17.84 12l-5.92-5.92L10.5 7.5 14 11H6z"/></svg>');}</style>


    
    
      
    
    
      
        
        
        <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&display=fallback">
        <style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style>
      
    
    
      <link rel="stylesheet" href="../../css/extra.css">
    
    <script>__md_scope=new URL("../..",location),__md_hash=e=>[...e].reduce(((e,_)=>(e<<5)-e+_.charCodeAt(0)),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script>
    
      

    
    
    
  </head>
  
  
    
    
      
    
    
    
    
    <body dir="ltr" data-md-color-scheme="default" data-md-color-primary="indigo" data-md-color-accent="indigo">
  
    
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
        
        <a href="#spatial-gene-expression-sge-format-conversion" class="md-skip">
          Skip to content
        </a>
      
    </div>
    <div data-md-component="announce">
      
    </div>
    
    
      

<header class="md-header" data-md-component="header">
  <nav class="md-header__inner md-grid" aria-label="Header">
    <a href="../.." title="cartloader" class="md-header__button md-logo" aria-label="cartloader" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54"/></svg>

    </a>
    <label class="md-header__button md-icon" for="__drawer">
      
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3zm0 5h18v2H3zm0 5h18v2H3z"/></svg>
    </label>
    <div class="md-header__title" data-md-component="header-title">
      <div class="md-header__ellipsis">
        <div class="md-header__topic">
          <span class="md-ellipsis">
            cartloader
          </span>
        </div>
        <div class="md-header__topic" data-md-component="header-topic">
          <span class="md-ellipsis">
            
              SGE Format Conversion
            
          </span>
        </div>
      </div>
    </div>
    
      
        <form class="md-header__option" data-md-component="palette">
  
    
    
    
    <input class="md-option" data-md-color-media="(prefers-color-scheme)" data-md-color-scheme="default" data-md-color-primary="indigo" data-md-color-accent="indigo"  aria-label="Switch to light mode"  type="radio" name="__palette" id="__palette_0">
    
      <label class="md-header__button md-icon" title="Switch to light mode" for="__palette_1" hidden>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3.9 12c0-1.71 1.39-3.1 3.1-3.1h4V7H7a5 5 0 0 0-5 5 5 5 0 0 0 5 5h4v-1.9H7c-1.71 0-3.1-1.39-3.1-3.1M8 13h8v-2H8zm9-6h-4v1.9h4c1.71 0 3.1 1.39 3.1 3.1s-1.39 3.1-3.1 3.1h-4V17h4a5 5 0 0 0 5-5 5 5 0 0 0-5-5"/></svg>
      </label>
    
  
    
    
    
    <input class="md-option" data-md-color-media="(prefers-color-scheme: light)" data-md-color-scheme="default" data-md-color-primary="indigo" data-md-color-accent="indigo"  aria-label="Switch to dark mode"  type="radio" name="__palette" id="__palette_1">
    
      <label class="md-header__button md-icon" title="Switch to dark mode" for="__palette_2" hidden>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M17 7H7a5 5 0 0 0-5 5 5 5 0 0 0 5 5h10a5 5 0 0 0 5-5 5 5 0 0 0-5-5m0 8a3 3 0 0 1-3-3 3 3 0 0 1 3-3 3 3 0 0 1 3 3 3 3 0 0 1-3 3"/></svg>
      </label>
    
  
    
    
    
    <input class="md-option" data-md-color-media="(prefers-color-scheme: dark)" data-md-color-scheme="slate" data-md-color-primary="black" data-md-color-accent="indigo"  aria-label="Switch to system preference"  type="radio" name="__palette" id="__palette_2">
    
      <label class="md-header__button md-icon" title="Switch to system preference" for="__palette_0" hidden>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M17 7H7a5 5 0 0 0-5 5 5 5 0 0 0 5 5h10a5 5 0 0 0 5-5 5 5 0 0 0-5-5M7 15a3 3 0 0 1-3-3 3 3 0 0 1 3-3 3 3 0 0 1 3 3 3 3 0 0 1-3 3"/></svg>
      </label>
    
  
</form>
      
    
    
      <script>var palette=__md_get("__palette");if(palette&&palette.color){if("(prefers-color-scheme)"===palette.color.media){var media=matchMedia("(prefers-color-scheme: light)"),input=document.querySelector(media.matches?"[data-md-color-media='(prefers-color-scheme: light)']":"[data-md-color-media='(prefers-color-scheme: dark)']");palette.color.media=input.getAttribute("data-md-color-media"),palette.color.scheme=input.getAttribute("data-md-color-scheme"),palette.color.primary=input.getAttribute("data-md-color-primary"),palette.color.accent=input.getAttribute("data-md-color-accent")}for(var[key,value]of Object.entries(palette.color))document.body.setAttribute("data-md-color-"+key,value)}</script>
    
    
    
      <label class="md-header__button md-icon" for="__search">
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg>
      </label>
      <div class="md-search" data-md-component="search" role="dialog">
  <label class="md-search__overlay" for="__search"></label>
  <div class="md-search__inner" role="search">
    <form class="md-search__form" name="search">
      <input type="text" class="md-search__input" name="query" aria-label="Search" placeholder="Search" autocapitalize="off" autocorrect="off" autocomplete="off" spellcheck="false" data-md-component="search-query" required>
      <label class="md-search__icon md-icon" for="__search">
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg>
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11z"/></svg>
      </label>
      <nav class="md-search__options" aria-label="Search">
        
        <button type="reset" class="md-search__icon md-icon" title="Clear" aria-label="Clear" tabindex="-1">
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"/></svg>
        </button>
      </nav>
      
    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" tabindex="0" data-md-scrollfix>
        <div class="md-search-result" data-md-component="search-result">
          <div class="md-search-result__meta">
            Initializing search
          </div>
          <ol class="md-search-result__list" role="presentation"></ol>
        </div>
      </div>
    </div>
  </div>
</div>
    
    
      <div class="md-header__source">
        <a href="https://github.com/seqscope/cartloader" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 6.7.2 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2024 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81"/></svg>
  </div>
  <div class="md-source__repository">
    seqscope/cartloader
  </div>
</a>
      </div>
    
  </nav>
  
</header>
    
    <div class="md-container" data-md-component="container">
      
      
        
          
            
<nav class="md-tabs" aria-label="Tabs" data-md-component="tabs">
  <div class="md-grid">
    <ul class="md-tabs__list">
      
        
  
  
  
    
    
      <li class="md-tabs__item">
        <a href="../.." class="md-tabs__link">
          
  
    
  
  Home

        </a>
      </li>
    
  

      
        
  
  
  
    <li class="md-tabs__item">
      <a href="../../installation/" class="md-tabs__link">
        
  
    
  
  Installation

      </a>
    </li>
  

      
        
  
  
  
    <li class="md-tabs__item">
      <a href="../../vignettes/quickstart/quickstart/" class="md-tabs__link">
        
  
    
  
  Quick Start

      </a>
    </li>
  

      
        
  
  
  
    
    
      <li class="md-tabs__item">
        <a href="../../vignettes/intro/" class="md-tabs__link">
          
  
    
  
  Vignettes

        </a>
      </li>
    
  

      
        
  
  
    
  
  
    
    
      <li class="md-tabs__item md-tabs__item--active">
        <a href="./" class="md-tabs__link">
          
  
    
  
  References

        </a>
      </li>
    
  

      
    </ul>
  </div>
</nav>
          
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              
              <div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    


  


<nav class="md-nav md-nav--primary md-nav--lifted" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href="../.." title="cartloader" class="md-nav__button md-logo" aria-label="cartloader" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M12 8a3 3 0 0 0 3-3 3 3 0 0 0-3-3 3 3 0 0 0-3 3 3 3 0 0 0 3 3m0 3.54C9.64 9.35 6.5 8 3 8v11c3.5 0 6.64 1.35 9 3.54 2.36-2.19 5.5-3.54 9-3.54V8c-3.5 0-6.64 1.35-9 3.54"/></svg>

    </a>
    cartloader
  </label>
  
    <div class="md-nav__source">
      <a href="https://github.com/seqscope/cartloader" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 6.7.2 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2024 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81"/></svg>
  </div>
  <div class="md-source__repository">
    seqscope/cartloader
  </div>
</a>
    </div>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
  
  
  
  
    
    
      
        
          
        
      
    
    
      
      
        
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
          
        
        <input class="md-nav__toggle md-toggle md-toggle--indeterminate" type="checkbox" id="__nav_1" >
        
          
          
          <div class="md-nav__link md-nav__container">
            <a href="../.." class="md-nav__link ">
              
  
  <span class="md-ellipsis">
    Home
    
  </span>
  

            </a>
            
          </div>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_1_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_1">
            <span class="md-nav__icon md-icon"></span>
            Home
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../../installation/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Installation
    
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../../vignettes/quickstart/quickstart/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Quick Start
    
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    
    
      
        
      
        
      
    
    
      
      
        
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
          
        
        <input class="md-nav__toggle md-toggle md-toggle--indeterminate" type="checkbox" id="__nav_4" >
        
          
          <label class="md-nav__link" for="__nav_4" id="__nav_4_label" tabindex="0">
            
  
  <span class="md-ellipsis">
    Vignettes
    
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_4_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_4">
            <span class="md-nav__icon md-icon"></span>
            Vignettes
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../vignettes/intro/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Overview
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    
    
      
        
      
        
      
        
      
        
      
        
      
        
      
        
      
    
    
      
      
        
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
          
        
        <input class="md-nav__toggle md-toggle md-toggle--indeterminate" type="checkbox" id="__nav_4_2" >
        
          
          <label class="md-nav__link" for="__nav_4_2" id="__nav_4_2_label" tabindex="0">
            
  
  <span class="md-ellipsis">
    Getting Started with `cartloader`
    
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="2" aria-labelledby="__nav_4_2_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_4_2">
            <span class="md-nav__icon md-icon"></span>
            Getting Started with `cartloader`
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../vignettes/subregion_tutorials/seqscope/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    SeqScope Starter Tutorial
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../vignettes/subregion_tutorials/stereoseq/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    StereoSeq Starter Tutorial
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../vignettes/subregion_tutorials/visiumhd/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Visium HD Starter Tutorial
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../vignettes/subregion_tutorials/pixelseq/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Pixel-Seq Starter Tutorial
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../vignettes/subregion_tutorials/xenium/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Xenium Starter Tutorial
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../vignettes/subregion_tutorials/merscope/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    MERSCOPE Starter Tutorial
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../../vignettes/subregion_tutorials/cosmx_smi/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    CoxMX SMI Starter Tutorial
    
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
    
  
  
  
    
    
      
        
      
        
      
        
      
        
      
    
    
      
        
        
      
      
        
      
    
    
    <li class="md-nav__item md-nav__item--active md-nav__item--section md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_5" checked>
        
          
          <label class="md-nav__link" for="__nav_5" id="__nav_5_label" tabindex="">
            
  
  <span class="md-ellipsis">
    References
    
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_5_label" aria-expanded="true">
          <label class="md-nav__title" for="__nav_5">
            <span class="md-nav__icon md-icon"></span>
            References
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
    
  
  
  
    <li class="md-nav__item md-nav__item--active">
      
      <input class="md-nav__toggle md-toggle" type="checkbox" id="__toc">
      
      
        
      
      
        <label class="md-nav__link md-nav__link--active" for="__toc">
          
  
  <span class="md-ellipsis">
    SGE Format Conversion
    
  </span>
  

          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <a href="./" class="md-nav__link md-nav__link--active">
        
  
  <span class="md-ellipsis">
    SGE Format Conversion
    
  </span>
  

      </a>
      
        

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#overview" class="md-nav__link">
    <span class="md-ellipsis">
      Overview
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#requirements" class="md-nav__link">
    <span class="md-ellipsis">
      Requirements
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#example-usages" class="md-nav__link">
    <span class="md-ellipsis">
      Example Usages
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#actions" class="md-nav__link">
    <span class="md-ellipsis">
      Actions
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#parameters" class="md-nav__link">
    <span class="md-ellipsis">
      Parameters
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#output" class="md-nav__link">
    <span class="md-ellipsis">
      Output
    </span>
  </a>
  
</li>
      
    </ul>
  
</nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../run_ficture2/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    FICTURE Analysis
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../run_cartload2/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    cartloader Compilation
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../upload_repository/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Data Repository Upload
    
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
              
              <div class="md-sidebar md-sidebar--secondary" data-md-component="sidebar" data-md-type="toc" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#overview" class="md-nav__link">
    <span class="md-ellipsis">
      Overview
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#requirements" class="md-nav__link">
    <span class="md-ellipsis">
      Requirements
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#example-usages" class="md-nav__link">
    <span class="md-ellipsis">
      Example Usages
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#actions" class="md-nav__link">
    <span class="md-ellipsis">
      Actions
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#parameters" class="md-nav__link">
    <span class="md-ellipsis">
      Parameters
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#output" class="md-nav__link">
    <span class="md-ellipsis">
      Output
    </span>
  </a>
  
</li>
      
    </ul>
  
</nav>
                  </div>
                </div>
              </div>
            
          
          
            <div class="md-content" data-md-component="content">
              <article class="md-content__inner md-typeset">
                
                  


  
  


<h1 id="spatial-gene-expression-sge-format-conversion">Spatial Gene Expression (<abbr title="Spatial Digital Gene Expression">SGE</abbr>) Format Conversion<a class="headerlink" href="#spatial-gene-expression-sge-format-conversion" title="Permanent link">&para;</a></h1>
<h2 id="overview">Overview<a class="headerlink" href="#overview" title="Permanent link">&para;</a></h2>
<p><abbr title="Spatial Digital Gene Expression">SGE</abbr> datasets vary widely in format and resolution across platforms. Since <code>FICTURE</code> requires <abbr title="Spatial Digital Gene Expression">SGE</abbr> in a specific format, <code>cartloader</code> toolkit provides this <code>sge_convert</code> module to standardize the raw, platform-specific <abbr title="Spatial Digital Gene Expression">SGE</abbr> into a FICTURE-compatible format.</p>
<h2 id="requirements">Requirements<a class="headerlink" href="#requirements" title="Permanent link">&para;</a></h2>
<div class="admonition info">
<p class="admonition-title">Input Data Requirement:</p>
<p>Please make sure the input data (raw, platform-specific <abbr title="Spatial Digital Gene Expression">SGE</abbr>) contains at least the following information. should be a transcript-indexed <abbr title="Spatial Digital Gene Expression">SGE</abbr> containing at least:</p>
<ul>
<li>Spatial coordinates (X coordinates, Y coordinates)</li>
<li>Feature metadata (such as gene symbols)</li>
<li>Expression Counts</li>
</ul>
</div>
<div class="admonition info">
<p class="admonition-title">Platform Compatibility</p>
<p>The current <code>sge_convert</code> supports standarizing <abbr title="Spatial Digital Gene Expression">SGE</abbr> from the following platforms:</p>
<table>
<thead>
<tr>
<th>Source</th>
<th><code>--platform</code> Option</th>
<th>Required Input Files</th>
</tr>
</thead>
<tbody>
<tr>
<td><a href="https://www.10xgenomics.com/platforms/visium">10x Visium HD</a></td>
<td><code>10x_visium_hd</code></td>
<td><code>--in-mex</code>, <code>--in-parquet</code>, <code>--scale-json</code></td>
</tr>
<tr>
<td><a href="https://www.nature.com/articles/s41596-024-01065-0">Seq-Scope</a></td>
<td><code>seqscope</code></td>
<td><code>--in-mex</code></td>
</tr>
<tr>
<td><a href="https://www.10xgenomics.com/platforms/xenium">10x Xenium</a></td>
<td><code>10x_xenium</code></td>
<td><code>--in-csv</code></td>
</tr>
<tr>
<td><a href="https://www.bgi.com/global/service/spatial-transcriptome-stereo-seq">Stereo-seq</a></td>
<td><code>bgi_stereoseq</code></td>
<td><code>--in-csv</code></td>
</tr>
<tr>
<td><a href="https://nanostring.com/products/cosmx-spatial-molecular-imager">CosMx SMI</a></td>
<td><code>cosmx_smi</code></td>
<td><code>--in-csv</code></td>
</tr>
<tr>
<td><a href="https://vizgen.com/merscope-ultra/">Vizgen MERSCOPE</a></td>
<td><code>vizgen_merscope</code></td>
<td><code>--in-csv</code></td>
</tr>
<tr>
<td><a href="https://www.cell.com/cell/fulltext/S0092-8674(22)01367-8">Pixel-seq</a></td>
<td><code>pixel_seq</code></td>
<td><code>--in-csv</code></td>
</tr>
<tr>
<td>Generic CSV/TSV input <sup>1</sup></td>
<td><code>generic</code></td>
<td><code>--in-csv</code></td>
</tr>
</tbody>
</table>
<p><sub><sup>1</sup>: For <abbr title="Spatial Digital Gene Expression">SGE</abbr> from platforms not yet explicitly supported by <code>cartloader</code>, or from custom/preprocessed sources, <code>sge_convert</code> provides a <code>generic</code> option that accepts CSV/TSV files with basic required fields (e.g., gene, spatial coordinates, expression count) for standardization and processing.</p>
</div>
<h2 id="example-usages">Example Usages<a class="headerlink" href="#example-usages" title="Permanent link">&para;</a></h2>
<h3 id="input-sge-in-mex-format">Input <abbr title="Spatial Digital Gene Expression">SGE</abbr> in <abbr title="Market Exchange">MEX</abbr> Format<a class="headerlink" href="#input-sge-in-mex-format" title="Permanent link">&para;</a></h3>
<h4 id="seq-scope"><code>Seq-Scope</code><a class="headerlink" href="#seq-scope" title="Permanent link">&para;</a></h4>
<div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"> 1</span>
<span class="normal"> 2</span>
<span class="normal"> 3</span>
<span class="normal"> 4</span>
<span class="normal"> 5</span>
<span class="normal"> 6</span>
<span class="normal"> 7</span>
<span class="normal"> 8</span>
<span class="normal"> 9</span>
<span class="normal">10</span>
<span class="normal">11</span>
<span class="normal">12</span>
<span class="normal">13</span></pre></div></td><td class="code"><div><pre><span></span><code>cartloader<span class="w"> </span>sge_convert<span class="w"> </span><span class="se">\</span>
<span class="w">    </span>--platform<span class="w"> </span>seqscope<span class="w"> </span><span class="se">\</span>
<span class="w">    </span>--in-mex<span class="w"> </span>/path/to/input/dir/of/mex<span class="w"> </span><span class="se">\ </span><span class="w">  </span>
<span class="w">    </span>--units-per-um<span class="w"> </span><span class="m">1000</span><span class="w"> </span><span class="se">\</span>
<span class="w">    </span>--icols-mtx<span class="w"> </span><span class="m">1</span><span class="w"> </span><span class="se">\</span>
<span class="w">    </span>--out-dir<span class="w"> </span>/path/to/output/dir<span class="w"> </span><span class="se">\</span>
<span class="w">    </span>--colnames-count<span class="w"> </span>count<span class="w"> </span><span class="se">\</span>
<span class="w">    </span>--filter-by-density<span class="w"> </span><span class="se">\</span>
<span class="w">    </span>--out-filtered-prefix<span class="w"> </span>filtered<span class="w"> </span><span class="se">\</span>
<span class="w">    </span>--genomic-feature<span class="w"> </span>count<span class="w"> </span><span class="se">\</span>
<span class="w">    </span>--sge-visual<span class="w"> </span><span class="se">\</span>
<span class="w">    </span>--north-up<span class="w"> </span><span class="se">\</span>
<span class="w">    </span>--spatula<span class="w"> </span>/path/to/spatul/binary<span class="w"> </span>
</code></pre></div></td></tr></table></div>
<h4 id="10x-visium-hd"><code>10X Visium HD</code><a class="headerlink" href="#10x-visium-hd" title="Permanent link">&para;</a></h4>
<div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span>
<span class="normal">2</span>
<span class="normal">3</span>
<span class="normal">4</span>
<span class="normal">5</span>
<span class="normal">6</span>
<span class="normal">7</span>
<span class="normal">8</span>
<span class="normal">9</span></pre></div></td><td class="code"><div><pre><span></span><code>cartloader<span class="w"> </span>sge_convert<span class="w"> </span><span class="se">\</span>
<span class="w">    </span>--platform<span class="w"> </span>10x_visium_hd<span class="w"> </span><span class="se">\</span>
<span class="w">    </span>--in-mex<span class="w"> </span>/path/to/input/dir/of/mex<span class="w"> </span><span class="se">\ </span><span class="w">  </span>
<span class="w">    </span>--in-parquet<span class="w"> </span>/path/to/input/parquet/file<span class="w"> </span><span class="se">\</span>
<span class="w">    </span>--scale-json<span class="w"> </span>/path/to/input/json/file<span class="w"> </span><span class="se">\</span>
<span class="w">    </span>--exclude-feature-regex<span class="w"> </span><span class="s1">&#39;^(BLANK.*$|NegCon.*$|NegPrb.*$)&#39;</span><span class="w"> </span><span class="se">\</span>
<span class="w">    </span>--out-dir<span class="w"> </span>/path/to/output/dir<span class="w"> </span><span class="se">\</span>
<span class="w">    </span>--spatula<span class="w"> </span>/path/to/spatul/binary<span class="w"> </span><span class="se">\</span>
<span class="w">    </span>--sge-visual<span class="w"> </span>
</code></pre></div></td></tr></table></div>
<h3 id="input-sge-in-tsvcsv-format">Input <abbr title="Spatial Digital Gene Expression">SGE</abbr> in TSV/CSV Format<a class="headerlink" href="#input-sge-in-tsvcsv-format" title="Permanent link">&para;</a></h3>
<p>This applies to input <abbr title="Spatial Digital Gene Expression">SGE</abbr> in TSV/CSV format from platforms including 10X Xenium, StereoSeq, Cosmx SMI, MERSCOPE, Pixel-seq. To simplify preprocessing, <code>sge_convert</code> automatically applies platform-specific defaults for common CSV/TSV parameters.</p>
<p>Below is an example converting <abbr title="Spatial Digital Gene Expression">SGE</abbr> from StereoSeq.</p>
<div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal"> 1</span>
<span class="normal"> 2</span>
<span class="normal"> 3</span>
<span class="normal"> 4</span>
<span class="normal"> 5</span>
<span class="normal"> 6</span>
<span class="normal"> 7</span>
<span class="normal"> 8</span>
<span class="normal"> 9</span>
<span class="normal">10</span>
<span class="normal">11</span>
<span class="normal">12</span>
<span class="normal">13</span></pre></div></td><td class="code"><div><pre><span></span><code>cartloader<span class="w"> </span>sge_convert<span class="w"> </span><span class="se">\</span>
<span class="w">    </span>--platform<span class="w"> </span>bgi_stereoseq<span class="w"> </span><span class="se">\</span>
<span class="w">    </span>--in-csv<span class="w"> </span>/path/to/input/csv/file<span class="w"> </span><span class="se">\</span>
<span class="w">    </span>--units-per-um<span class="w"> </span><span class="m">28</span>.75<span class="w"> </span><span class="se">\</span>
<span class="w">    </span>--out-dir<span class="w"> </span>/path/to/output/dir<span class="w"> </span><span class="se">\</span>
<span class="w">    </span>--colnames-count<span class="w"> </span>count<span class="w"> </span><span class="se">\</span>
<span class="w">    </span>--exclude-feature-regex<span class="w"> </span><span class="s1">&#39;^(BLANK.*$|NegCon.*$|NegPrb.*$)&#39;</span><span class="w"> </span><span class="se">\</span>
<span class="w">    </span>--filter-by-density<span class="w"> </span><span class="se">\</span>
<span class="w">    </span>--out-filtered-prefix<span class="w"> </span>filtered<span class="w"> </span><span class="se">\</span>
<span class="w">    </span>--genomic-feature<span class="w"> </span>count<span class="w"> </span><span class="se">\</span>
<span class="w">    </span>--sge-visual<span class="w"> </span><span class="se">\</span>
<span class="w">    </span>--north-up<span class="w"> </span><span class="se">\</span>
<span class="w">    </span>--spatula<span class="w"> </span>/path/to/spatul/binary<span class="w"> </span>
</code></pre></div></td></tr></table></div>
<div class="admonition warning">
<p class="admonition-title">Verify your input structure</p>
<p>Always double-check the column names and metadata format of your input files. If they differ from the expected defaults, override them using <code>--csv-*</code> and <code>--min-phred-score</code> options.</p>
</div>
<details>
<summary>Click to view platform-specific default settings</summary>
<p>To streamline the process, <code>sge_convert</code> automatically applies platform-dependent defaults for CSV/TSV parsing based on known file formats and column conventions.
Below summarizes the default values for key parameters per supported platform:</p>
<table>
<thead>
<tr>
<th>Platform</th>
<th><code>--csv-comment</code><sup>1</sup></th>
<th><code>--csv-delim</code></th>
<th><code>--csv-colname-x</code></th>
<th><code>--csv-colname-y</code></th>
<th><code>--csv-colnames-count</code></th>
<th><code>--csv-colname-feature-name</code></th>
</tr>
</thead>
<tbody>
<tr>
<td>10X Xenium<sup>2</sup></td>
<td><code>False</code></td>
<td><code>,</code></td>
<td><code>x_location</code></td>
<td><code>y_location</code></td>
<td>-</td>
<td><code>feature_name</code></td>
</tr>
<tr>
<td>StereoSeq</td>
<td><code>False</code></td>
<td><code>\t</code></td>
<td><code>x</code></td>
<td><code>y</code></td>
<td><code>MIDCounts</code></td>
<td><code>geneID</code></td>
</tr>
<tr>
<td>CosMx SMI</td>
<td><code>False</code></td>
<td><code>,</code></td>
<td><code>x_local_px</code></td>
<td><code>y_local_px</code></td>
<td>-</td>
<td><code>target</code></td>
</tr>
<tr>
<td>MERSCOPE</td>
<td><code>False</code></td>
<td><code>,</code></td>
<td><code>global_x</code></td>
<td><code>global_y</code></td>
<td>-</td>
<td><code>gene</code></td>
</tr>
<tr>
<td>Pixel-seq</td>
<td><code>False</code></td>
<td><code>\t</code></td>
<td><code>xcoord</code></td>
<td><code>ycoord</code></td>
<td>-</td>
<td><code>geneName</code></td>
</tr>
</tbody>
</table>
<p><sub><sup>1</sup> <code>--csv-comment</code>: If <code>True</code>, the lines starts with <code>#</code> will be treated as comments and will be skipped.</p>
<p><sub><sup>2</sup> 10X Xenium: Besides the above default settings, for 10X Xenium data, <code>sge_convert</code> also applies <code>--csv-colname-phredscore qv</code> and <code>--min-phred-score 20</code>.</p>
</details>
<hr />
<h2 id="actions">Actions<a class="headerlink" href="#actions" title="Permanent link">&para;</a></h2>
<h3 id="sge-conversion"><abbr title="Spatial Digital Gene Expression">SGE</abbr> Conversion<a class="headerlink" href="#sge-conversion" title="Permanent link">&para;</a></h3>
<p>Converting <abbr title="Spatial Digital Gene Expression">SGE</abbr> into a FICTURE-compatible TSV format. During conversion, <abbr title="Spatial Digital Gene Expression">SGE</abbr> coordinates are rescaled to micrometer units based on the pixel resolution specified in the input. It's also available to apply feature (typically, genes) filtering.</p>
<h3 id="optional-density-based-filtering">(Optional) Density-based Filtering<a class="headerlink" href="#optional-density-based-filtering" title="Permanent link">&para;</a></h3>
<p>Automatically identify and retain high-quality tissue regions based on transcript density and spatial structure. This step takes the format-standardized <abbr title="Spatial Digital Gene Expression">SGE</abbr> as input and generate a density-based filtered <abbr title="Spatial Digital Gene Expression">SGE</abbr>.</p>
<h3 id="optional-sge-visualization">(Optional) <abbr title="Spatial Digital Gene Expression">SGE</abbr> Visualization<a class="headerlink" href="#optional-sge-visualization" title="Permanent link">&para;</a></h3>
<p>Draws an image of 2D points provided as an input. In this step, it is optional to enable the <code>--north-up</code> option to ensuring correct spatial orientation (i.e., Y-axis increases upward/north and X-axis increases to the right/east).</p>
<hr />
<h2 id="parameters">Parameters<a class="headerlink" href="#parameters" title="Permanent link">&para;</a></h2>
<p>The following outlines the <strong>minimum required parameters</strong>.</p>
<p>For most auxiliary parameters, the <strong>default values are recommended</strong> and could be modified when they do not suit your use case. See more details in the collapsible sections below or by running:
<div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span></pre></div></td><td class="code"><div><pre><span></span><code>cartloader<span class="w"> </span>sge_convert<span class="w"> </span>--help
</code></pre></div></td></tr></table></div></p>
<h3 id="sge-conversion_1"><abbr title="Spatial Digital Gene Expression">SGE</abbr> Conversion<a class="headerlink" href="#sge-conversion_1" title="Permanent link">&para;</a></h3>
<ul>
<li><code>--platform</code> (str): Source platform to infer the input file format and default setting (options: "<code>10x_visium_hd</code>", "<code>seqscope</code>", "<code>10x_xenium</code>", "<code>bgi_stereoseq</code>", "<code>cosmx_smi</code>", "<code>vizgen_merscope</code>", "<code>pixel_seq</code>", "<code>generic</code>").</li>
<li><code>--in-mex</code> (str): Path to the input <abbr title="Spatial Digital Gene Expression">SGE</abbr> directory in <abbr title="Market Exchange">MEX</abbr> format. Required for <abbr title="Market Exchange">MEX</abbr>-formatted data (e.g., 10X Visium HD, SeqScope).</li>
<li><code>--in-csv</code> (str): Path to the input <abbr title="Spatial Digital Gene Expression">SGE</abbr> file in CSV or TSV format. Required for CSV/TSV-formatted data (e.g., 10X Xenium, StereoSeq, CosMX SMI, MERSCOPE, Pixel-seq).</li>
<li><code>--in-parquet</code> (str) : Path to the input parquet file with spatial coordinates, if available. Typically named <code>tissue_positions.parquet</code> (10X Visium HD).</li>
<li><code>--scale-json</code> (str): Path to the scale JSON file to compute <code>--units-per-um</code>, if available. Typically named <code>scalefactors_json.json</code> (10X Visium HD).</li>
<li><code>--units-per-um</code> (float): Coordinate units per micrometer (default: 1.00). Skip if <code>--scale-json</code> is provided.</li>
<li><code>--out-dir</code> (str) : Path to the output directory.</li>
<li><code>--include-feature-regex</code> (regex): (Optional) A regex pattern of feature/gene names to be included.</li>
<li><code>--exclude-feature-regex</code> (regex): (Optional) A regex pattern of feature/gene names to be excluded.</li>
</ul>
<!-- * `--include-feature-list` and `--exclude-feature-list`: A file containing a list of input genes to be included or excluded. 
* `--include-feature-type-regex`: A regex pattern of feature/gene type to be included.
* `--csv-colname-feature-type`: If `--include-feature-type-regex`, column name in the input TSV/CSV that contains gene type information.
* `--feature-type-ref`, `--feature-type-ref-delim`, `--feature-type-ref-colidx-name`, and `--feature-type-ref-colidx-type`: If `--include-feature-type-regex`, define the file path, delimiter, and column indices for gene names and types of an additional reference file to provide gene type information. -->
<details class="note">
<summary>Auxiliary <abbr title="Spatial Digital Gene Expression">SGE</abbr> Conversion Paramaters</summary>
<p><strong>Auxiliary Input <abbr title="Market Exchange">MEX</abbr> Parameters</strong>:</p>
<ul>
<li><code>--icols-mtx</code> (int or comma-spearated list): Comma-separated, 1-based indices of the target genomic features among the count columns in the input matrix file. (Default: 1)</li>
<li><code>--colnames-count</code> (string or comma-spearated list): Comma-separated output column names for the specified genomic features. (Default: count). The number of names specified by <code>--colnames-count</code> must match the number of indices provided in <code>--icols-mtx</code>.</li>
</ul>
<p><strong>Auxiliary Input CSV/TSV Parameters</strong>:</p>
<ul>
<li><code>--csv-comment</code> (flag): If enabled, lines starts with <code>#</code> will be skipped (default: <code>False</code> for 10X Xenium, StereoSeq, CosMx SMI, MERSCOPE, and Pixel-seq).</li>
<li><code>--csv-delim</code> (str): Delimiter for the input file (default: <code>","</code> for 10X Xenium, CosMx SMI, and MERSCOPE; <code>"\t"</code> for StereoSeq, Pixel-seq).</li>
<li><code>--csv-colname-x</code> (str): Column name for X coordinates (default: <code>x_location</code> for 10X Xenium; <code>x</code> for StereoSeq; <code>x_local_px</code> for CosMx SMI; <code>global_x</code> for MERSCOPE; <code>xcoord</code> for Pixel-seq).</li>
<li><code>--csv-colname-y</code> (str): Column name for Y coordinates (default: <code>y_location</code> for 10X Xenium; <code>y</code> for StereoSeq; <code>y_local_px</code> for CosMx SMI; <code>global_y</code> for MERSCOPE; <code>ycoord</code> for Pixel-seq).</li>
<li><code>--csv-colnames-count</code> (str): Comma-separated column names for expression count. If not provided, a count of 1 per transcript (default: <code>MIDCounts</code> for StereoSeq).</li>
<li><code>--csv-colname-feature-name</code> (str): Column name for gene name (default: <code>feature_name</code> for 10X Xenium; <code>geneID</code> for StereoSeq; <code>target</code> for CosMx SMI; <code>gene</code> for MERSCOPE; <code>geneName</code> for Pixel-seq).</li>
<li><code>--csv-colnames-others</code> (str): Columns names to keep.</li>
<li><code>--csv-colname-phredscore</code> (str): Column name for Phred-scaled quality value estimating the probability of incorrect calls (default: <code>qv</code> for 10X Xenium).</li>
<li><code>--min-phred-score</code> (int): Phred-scaled quality score cutoff (default: <code>20</code> for 10X Xenium).</li>
</ul>
<p><strong>Auxiliary Output Parameters</strong>: </p>
<ul>
<li><code>--out-transcript</code> (str): File name for output compressed transcript-indexed <abbr title="Spatial Digital Gene Expression">SGE</abbr> file in TSV format (default: <code>transcripts.unsorted.tsv.gz</code>).</li>
<li><code>--out-minmax</code> (str): File name for coordinate min-max values in TSV format (default: <code>coordinate_minmax.tsv</code>).</li>
<li><code>--out-feature</code> (str): File name for compressed <abbr title="Unique Molecular Identifier">UMI</abbr> count per gene in TSV format (default: <code>feature.clean.tsv.gz</code>).</li>
<li><code>--precision-um</code> (int): Decimal precision for transcript coordinates; set to <code>0</code> to round to integers (default: 2).</li>
<li><code>--colname-x</code> (str): Column name for the X-coordinate in the output <abbr title="Spatial Digital Gene Expression">SGE</abbr> (default: X).</li>
<li><code>--colname-y</code> (str): Column name for the Y-coordinate in the output <abbr title="Spatial Digital Gene Expression">SGE</abbr> (default: Y).</li>
<li><code>--colnames-count</code> (str): Comma-separated column names for expression count in the output <abbr title="Spatial Digital Gene Expression">SGE</abbr> (default: count).</li>
<li><code>--colname-feature-name</code> (str): Column name for the gene name in the output <abbr title="Spatial Digital Gene Expression">SGE</abbr>(default: gene).</li>
</ul>
<p><strong>Auxiliary Environment Parameters</strong><br />
If the binaries are already available in your system's <code>PATH</code>, you may omit these options.</p>
<ul>
<li><code>--gzip</code> (str): Path to <code>gzip</code> binary; consider <code>pigz -p 4</code> for faster processing. (Default: <code>gzip</code>)</li>
<li><code>--spatula</code> (str): Path to <code>spatula</code> binary. (Default: <code>spatula</code>)</li>
<li><code>--parquet-tools</code> (str): Required if <code>--in-parquet</code> is used; path to <code>parquet-tools</code> binary. (Default: <code>parquet-tools</code>)</li>
</ul>
</details>
<h3 id="optional-density-based-filtering_1">(Optional) Density-based Filtering<a class="headerlink" href="#optional-density-based-filtering_1" title="Permanent link">&para;</a></h3>
<ul>
<li><code>--filter-by-density</code> (flag): Enable filtering of <abbr title="Spatial Digital Gene Expression">SGE</abbr> by density.</li>
<li><code>--out-filtered-prefix</code> (str): Prefix for output filtered <abbr title="Spatial Digital Gene Expression">SGE</abbr> files (default: filtered).</li>
<li><code>--genomic-feature</code> (str): Genomic feature to be used for density-based filtering. Defaults to the value of <code>--colnames-count</code> if only one column name is provided.</li>
</ul>
<details class="note">
<summary>Auxiliary Density-based Filtering Paramaters</summary>
<ul>
<li><code>--mu-scale</code> (float): Scale factor for the polygon area calculation (default: 1.0).</li>
<li><code>--radius</code> (int): Radius for the polygon area calculation (default: 15).</li>
<li><code>--quartile</code> (int): Quartile for the polygon area calculation (default: 2).</li>
<li><code>--hex-n-move</code> (int): Sliding step (default: 1).</li>
<li><code>--polygon-min-size</code> (int): The minimum polygon size (default: 500).</li>
</ul>
</details>
<h3 id="optional-sge-visualization_1">(Optional) <abbr title="Spatial Digital Gene Expression">SGE</abbr> Visualization<a class="headerlink" href="#optional-sge-visualization_1" title="Permanent link">&para;</a></h3>
<ul>
<li><code>sge-visual</code> (flag): Enable <abbr title="Spatial Digital Gene Expression">SGE</abbr> visualization.</li>
<li><code>--north-up</code> (flag): Enable the north-up orientation for the <abbr title="Spatial Digital Gene Expression">SGE</abbr> visualization.</li>
</ul>
<details class="note">
<summary>Auxiliary <abbr title="Spatial Digital Gene Expression">SGE</abbr> Visualization Paramaters</summary>
<ul>
<li><code>--out-xy</code> (str): File name for output <abbr title="Spatial Digital Gene Expression">SGE</abbr> visualization image (default: <code>xy.png</code>).</li>
<li><code>--out-northup-tif</code> (str): File name for output north-up orientated image (default: <code>xy_northup.tif</code>).</li>
<li><code>--srs</code> (str): If <code>--north-up</code>, define the spatial reference system (default: EPSG:3857).</li>
<li><code>--resample</code> (str): If <code>--north-up</code>, Define the resampling method (default: cubic). Options: near, bilinear, cubic, etc.</li>
<li><code>--gdal_translate</code> (str): Required if <code>--north-up</code>; path to <code>gdal_translate</code> binary. (Default: <code>gdal_translate</code>)</li>
<li><code>--gdalwarp</code> (str): Required if <code>--north-up</code>; path to <code>gdalwarp</code> binary. (Default: <code>gdalwarp</code>)</li>
</ul>
</details>
<hr />
<h2 id="output">Output<a class="headerlink" href="#output" title="Permanent link">&para;</a></h2>
<p><code>cartloader</code> generates the following harmonized outputs:</p>
<h3 id="unified-sge-matrix">Unified <abbr title="Spatial Digital Gene Expression">SGE</abbr> matrix<a class="headerlink" href="#unified-sge-matrix" title="Permanent link">&para;</a></h3>
<p>Both <abbr title="Spatial Digital Gene Expression">SGE</abbr> conversion and density-based filtering generate a unified <abbr title="Spatial Digital Gene Expression">SGE</abbr> matrix, consisting of:</p>
<details>
<summary><code>transcripts.unsorted.tsv.gz</code>: transcript-indexed <abbr title="Spatial Digital Gene Expression">SGE</abbr> in TSV</summary>
<div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span>
<span class="normal">2</span>
<span class="normal">3</span>
<span class="normal">4</span></pre></div></td><td class="code"><div><pre><span></span><code>X        Y        gene     count
295.29   1422.35  Myo3a    0
1745.54  1110.72  Med14    1
1745.54  1110.72  Ntpcr    1
</code></pre></div></td></tr></table></div>
<ul>
<li><code>X</code>: X coordinates in um</li>
<li><code>Y</code>: Y coordinates in um</li>
<li><code>gene</code>: gene symbols</li>
<li><code>count</code>: expression count per pixel per gene</li>
</ul>
</details>
<details>
<summary><code>feature.clean.tsv.gz</code>: <abbr title="Unique Molecular Identifier">UMI</abbr> counts on a per-gene basis in TSV</summary>
<p><div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span>
<span class="normal">2</span>
<span class="normal">3</span>
<span class="normal">4</span></pre></div></td><td class="code"><div><pre><span></span><code>gene           gene_id             count
Gm29155        ENSMUSG00000100764  1
Pcmtd1         ENSMUSG00000051285  431
Gm26901        ENSMUSG00000097797  1
</code></pre></div></td></tr></table></div>
* <code>gene</code>: gene symbols
* <code>gene_id</code>: gene IDs
* <code>count</code>: expression count per gene</p>
</details>
<details>
<summary><code>coordinate_minmax.tsv</code>: X Y min/max includemd_vigenettes_sgeformat.mdcoordinates</summary>
<div class="highlight"><table class="highlighttable"><tr><td class="linenos"><div class="linenodiv"><pre><span></span><span class="normal">1</span>
<span class="normal">2</span>
<span class="normal">3</span>
<span class="normal">4</span></pre></div></td><td class="code"><div><pre><span></span><code>xmin    0.14
xmax    2359.90
ymin    0.23
ymax    1439.95
</code></pre></div></td></tr></table></div>
<ul>
<li><code>xmin</code> <code>xmax</code>: min and max X coordinates in um</li>
<li><code>ymin</code> <code>ymax</code>: min and max Y coordinates in um</li>
</ul>
</details>
<h3 id="sge-images"><abbr title="Spatial Digital Gene Expression">SGE</abbr> Images<a class="headerlink" href="#sge-images" title="Permanent link">&para;</a></h3>
<ul>
<li>When <code>--sge-visual</code> is enabled, a monochrome PNG image is generated to visualize the <abbr title="Spatial Digital Gene Expression">SGE</abbr> data.</li>
<li>When <code>--north-up</code> is enabled, a georeferenced TIFF image is produced with a north-up orientation.</li>
</ul>












                
              </article>
            </div>
          
          
<script>var target=document.getElementById(location.hash.slice(1));target&&target.name&&(target.checked=target.name.startsWith("__tabbed_"))</script>
        </div>
        
      </main>
      
        <footer class="md-footer">
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-copyright">
  
  
    Made with
    <a href="https://squidfunk.github.io/mkdocs-material/" target="_blank" rel="noopener">
      Material for MkDocs
    </a>
  
</div>
      
    </div>
  </div>
</footer>
      
    </div>
    <div class="md-dialog" data-md-component="dialog">
      <div class="md-dialog__inner md-typeset"></div>
    </div>
    
    
    <script id="__config" type="application/json">{"base": "../..", "features": ["navigation.tabs", "navigation.indexes", "navigation.sections", "navigation.expand", "navigation.path", "content.code.annotate", "content.grid", "content.cards", "content.footnote.tooltips"], "search": "../../assets/javascripts/workers/search.f8cc74c7.min.js", "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version": "Select version"}}</script>
    
    
      <script src="../../assets/javascripts/bundle.f13b1293.min.js"></script>
      
        <script src="../../js/open_in_new_tab.js"></script>
      
    
  </body>
</html>